// @microsoft/omnichannel-chat-components/lib/cjs/components/prechatsurveypane/common/defaultProps/defaultStyles/defaultPreChatSurveyPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultPreChatSurveyPaneStyles = void 0;

var _defaultPreChatSurveyPaneACContainerStyles = require("./defaultPreChatSurveyPaneACContainerStyles");

var _defaultPreChatSurveyPaneButtonStyles = require("./defaultPreChatSurveyPaneButtonStyles");

var _defaultPreChatSurveyPaneGeneralStyles = require("./defaultPreChatSurveyPaneGeneralStyles");

var _defaultPreChatSurveyPaneMultilineTextInputStyles = require("./defaultPreChatSurveyPaneMultilineTextInputStyles");

var _defaultPreChatSurveyPaneTextInputStyles = require("./defaultPreChatSurveyPaneTextInputStyles");

var _defaultPreChatSurveyPaneMultichoiceInputStyles = require("./defaultPreChatSurveyPaneMultichoiceInputStyles");

const defaultPreChatSurveyPaneStyles = {
  generalStyleProps: _defaultPreChatSurveyPaneGeneralStyles.defaultPreChatSurveyPaneGeneralStyles,
  customButtonStyleProps: _defaultPreChatSurveyPaneButtonStyles.defaultPreChatSurveyPaneButtonStyles,
  adaptiveCardContainerStyleProps: _defaultPreChatSurveyPaneACContainerStyles.defaultPreChatSurveyPaneACContainerStyles,
  customTextInputStyleProps: _defaultPreChatSurveyPaneTextInputStyles.defaultPreChatSurveyPaneTextInputStyles,
  customMultilineTextInputStyleProps: _defaultPreChatSurveyPaneMultilineTextInputStyles.defaultPreChatSurveyPaneMultilineTextInputStyles,
  customMultichoiceInputStyleProps: _defaultPreChatSurveyPaneMultichoiceInputStyles.defaultPreChatSurveyPaneMultichoiceInputStyles
};
exports.defaultPreChatSurveyPaneStyles = defaultPreChatSurveyPaneStyles;
