// @microsoft/omnichannel-chat-components/lib/cjs/components/reconnectchatpane/common/default/defaultStyles/defaultReconnectChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultReconnectChatPaneStyles = void 0;

var _defaultReconnectChatPaneSubtitleStyles = require("./defaultReconnectChatPaneSubtitleStyles");

var _defaultReconnectChatPaneButtonGroupStyles = require("./defaultReconnectChatPaneButtonGroupStyles");

var _defaultReconnectChatPaneContinueChatButtonStyles = require("./defaultReconnectChatPaneContinueChatButtonStyles");

var _defaultReconnectChatPaneGeneralStyles = require("./defaultReconnectChatPaneGeneralStyles");

var _defaultReconnectChatPaneTitleStyles = require("./defaultReconnectChatPaneTitleStyles");

var _defaultReconnectChatPaneIconStyles = require("./defaultReconnectChatPaneIconStyles");

var _defaultReconnectChatPaneWrapperStyles = require("./defaultReconnectChatPaneWrapperStyles");

var _defaultReconnectChatPaneStartNewChatButtonStyles = require("./defaultReconnectChatPaneStartNewChatButtonStyles");

var _defaultReconnectChatPaneContinueChatButtonHoveredStyles = require("./defaultReconnectChatPaneContinueChatButtonHoveredStyles");

var _defaultReconnectChatPaneStartNewChatButtonHoveredStyles = require("./defaultReconnectChatPaneStartNewChatButtonHoveredStyles");

const defaultReconnectChatPaneStyles = {
  generalStyleProps: _defaultReconnectChatPaneGeneralStyles.defaultReconnectChatPaneGeneralStyles,
  wrapperStyleProps: _defaultReconnectChatPaneWrapperStyles.defaultReconnectChatPaneWrapperStyles,
  titleStyleProps: _defaultReconnectChatPaneTitleStyles.defaultReconnectChatPaneTitleStyles,
  subtitleStyleProps: _defaultReconnectChatPaneSubtitleStyles.defaultReconnectChatPaneSubtitleStyles,
  iconStyleProps: _defaultReconnectChatPaneIconStyles.defaultReconnectChatPaneIconStyles,
  buttonGroupStyleProps: _defaultReconnectChatPaneButtonGroupStyles.defaultReconnectChatPaneButtonGroupStyles,
  continueChatButtonStyleProps: _defaultReconnectChatPaneContinueChatButtonStyles.defaultReconnectChatPaneContinueChatButtonStyles,
  continueChatButtonHoveredStyleProps: _defaultReconnectChatPaneContinueChatButtonHoveredStyles.defaultReconnectChatPaneContinueChatButtonHoveredStyles,
  startNewChatButtonStyleProps: _defaultReconnectChatPaneStartNewChatButtonStyles.defaultReconnectChatPaneStartNewChatButtonStyles,
  startNewChatButtonHoveredStyleProps: _defaultReconnectChatPaneStartNewChatButtonHoveredStyles.defaultReconnectChatPaneStartNewChatButtonHoveredStyles
};
exports.defaultReconnectChatPaneStyles = defaultReconnectChatPaneStyles;
