// @microsoft/omnichannel-chat-components/lib/cjs/components/inputvalidationpane/InputValidationPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Button = require("@fluentui/react/lib/Button");

var _react = require("@fluentui/react");

var _react2 = _interopRequireWildcard(require("react"));

var _BroadcastService = require("../../services/BroadcastService");

var _Constants = require("../../common/Constants");

var _decodeComponentString = require("../../common/decodeComponentString");

var _defaultInputValidationPaneButtonGroupStyles = require("./common/default/defaultStyles/defaultInputValidationPaneButtonGroupStyles");

var _defaultInputValidationPaneCancelButtonHoveredStyles = require("./common/default/defaultStyles/defaultInputValidationPaneCancelButtonHoveredStyles");

var _defaultInputValidationPaneCancelButtonStyles = require("./common/default/defaultStyles/defaultInputValidationPaneCancelButtonStyles");

var _defaultInputValidationPaneControlProps = require("./common/default/defaultProps/defaultInputValidationPaneControlProps");

var _defaultInputValidationPaneGeneralStyles = require("./common/default/defaultStyles/defaultInputValidationPaneGeneralStyles");

var _defaultInputValidationPaneHeaderGroupStyles = require("./common/default/defaultStyles/defaultInputValidationPaneHeaderGroupStyles");

var _defaultInputValidationPaneInputStyles = require("./common/default/defaultStyles/defaultInputValidationPaneInputStyles");

var _defaultInputValidationPaneInvalidInputErrorMessageStyles = require("./common/default/defaultStyles/defaultInputValidationPaneInvalidInputErrorMessageStyles");

var _defaultInputValidationPaneSendButtonHoveredStyles = require("./common/default/defaultStyles/defaultInputValidationPaneSendButtonHoveredStyles");

var _defaultInputValidationPaneSendButtonStyles = require("./common/default/defaultStyles/defaultInputValidationPaneSendButtonStyles");

var _defaultInputValidationPaneSubtitleStyles = require("./common/default/defaultStyles/defaultInputValidationPaneSubtitleStyles");

var _defaultInputValidationPaneTitleStyles = require("./common/default/defaultStyles/defaultInputValidationPaneTitleStyles");

var _utils = require("../../common/utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function InputValidationPane(props) {
  var _props$controlProps, _props$styleProps, _props$styleProps2, _props$styleProps3, _props$styleProps4, _props$controlProps14, _props$styleProps5, _props$styleProps6, _props$styleProps7, _props$styleProps8, _props$styleProps9, _props$styleProps10, _props$styleProps11, _props$styleProps12, _props$styleProps13, _props$styleProps14, _props$styleProps15, _props$styleProps16, _props$controlProps15, _props$styleProps17, _props$styleProps17$c, _props$controlProps16, _props$controlProps17, _props$styleProps18, _props$styleProps18$c, _props$controlProps18, _props$componentOverr, _props$styleProps19, _props$styleProps19$c, _props$controlProps19, _props$controlProps20, _props$componentOverr2, _props$styleProps20, _props$styleProps20$c, _props$controlProps21, _props$controlProps22, _props$componentOverr3, _props$styleProps21, _props$styleProps21$c, _props$controlProps23, _props$controlProps24, _props$controlProps25, _props$componentOverr4, _props$styleProps22, _props$styleProps22$c, _props$controlProps26, _props$controlProps27, _props$controlProps28, _props$styleProps23, _props$styleProps23$c, _props$controlProps29, _props$componentOverr5, _props$styleProps24, _props$styleProps24$c, _props$controlProps30, _props$controlProps31, _props$controlProps32, _props$controlProps33, _props$componentOverr6, _props$styleProps25, _props$styleProps25$c, _props$controlProps34, _props$controlProps35, _props$controlProps36;

  const elementId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.id;
  const [inputValue, setInputValue] = (0, _react2.useState)("");
  const [isInitialRendering, setIsInitialRendering] = (0, _react2.useState)(true);
  const [isInvalidInput, setIsInvalidInput] = (0, _react2.useState)(false);
  const [isSendButtonEnabled, setIsSendButtonEnabled] = (0, _react2.useState)(false);
  const isValidInput = (0, _react2.useCallback)(() => {
    var _props$controlProps2, _props$controlProps3;

    if (!((_props$controlProps2 = props.controlProps) !== null && _props$controlProps2 !== void 0 && _props$controlProps2.checkInput)) {
      return true;
    }

    if (!inputValue) {
      return false;
    }

    return (_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.checkInput(inputValue);
  }, [inputValue]); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const handleInputChange = (0, _react2.useCallback)(e => {
    var _props$controlProps4, _props$controlProps5;

    setInputValue(e.target.value);
    e.target.value ? setIsSendButtonEnabled(((_props$controlProps4 = props.controlProps) === null || _props$controlProps4 === void 0 ? void 0 : _props$controlProps4.enableSendButton) || e.target.value !== "") : setIsSendButtonEnabled(((_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.enableSendButton) ?? false);
    setIsInvalidInput(false);
  }, []);
  const send = (0, _react2.useCallback)((controlId, suffix) => {
    var _props$controlProps6;

    if ((_props$controlProps6 = props.controlProps) !== null && _props$controlProps6 !== void 0 && _props$controlProps6.onSend) {
      if (isValidInput()) {
        var _props$controlProps7;

        const eventName = (0, _utils.generateEventName)(controlId, "on", suffix);
        const customEvent = {
          eventName: eventName
        };

        _BroadcastService.BroadcastService.postMessage(customEvent);

        (_props$controlProps7 = props.controlProps) === null || _props$controlProps7 === void 0 ? void 0 : _props$controlProps7.onSend(inputValue);
      } else {
        setIsInvalidInput(true);
      }
    }
  }, [inputValue]); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const handleInputKeyUp = (0, _react2.useCallback)(e => {
    if (e.code === _Constants.KeyCodes.ENTER) {
      send(elementId + "-textField", "KeyUp");
    }
  }, [inputValue]);
  const handleSendClick = (0, _react2.useCallback)(() => {
    send(elementId + "-sendbutton", "Click");
  }, [inputValue]);
  const cancel = (0, _react2.useCallback)((controlId, suffix) => {
    var _props$controlProps8;

    if ((_props$controlProps8 = props.controlProps) !== null && _props$controlProps8 !== void 0 && _props$controlProps8.onCancel) {
      var _props$controlProps9, _props$controlProps10;

      setInputValue("");
      setIsInvalidInput(false);
      setIsSendButtonEnabled(((_props$controlProps9 = props.controlProps) === null || _props$controlProps9 === void 0 ? void 0 : _props$controlProps9.enableSendButton) ?? false);
      const eventName = (0, _utils.generateEventName)(controlId, "on", suffix);
      const customEvent = {
        eventName: eventName
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps10 = props.controlProps) === null || _props$controlProps10 === void 0 ? void 0 : _props$controlProps10.onCancel();
    }
  }, []); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const handleEscKeyDown = (0, _react2.useCallback)(e => {
    if (e.code === _Constants.KeyCodes.ESCAPE) {
      cancel(elementId, "KeyDown");
    }
  }, []);
  const handleCancelClick = (0, _react2.useCallback)(() => {
    cancel(elementId + "-cancelbutton", "Click");
  }, []);
  (0, _react2.useEffect)(() => {
    var _props$controlProps11, _props$controlProps12, _props$controlProps13;

    setInputValue(((_props$controlProps11 = props.controlProps) === null || _props$controlProps11 === void 0 ? void 0 : _props$controlProps11.inputInitialText) ?? "");
    const isInputEmpty = (_props$controlProps12 = props.controlProps) !== null && _props$controlProps12 !== void 0 && _props$controlProps12.inputInitialText ? false : true;
    setIsSendButtonEnabled(((_props$controlProps13 = props.controlProps) === null || _props$controlProps13 === void 0 ? void 0 : _props$controlProps13.enableSendButton) || !isInputEmpty);

    if (isInitialRendering) {
      setIsInitialRendering(!isInitialRendering);
    }
  }, [isInitialRendering]);
  const containerStyles = {
    root: Object.assign({}, _defaultInputValidationPaneGeneralStyles.defaultInputValidationPaneGeneralStyles, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const headerGroupStyles = {
    root: Object.assign({}, _defaultInputValidationPaneHeaderGroupStyles.defaultInputValidationPaneHeaderGroupStyles, (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.headerGroupStyleProps)
  };
  const titleStyles = {
    root: Object.assign({}, _defaultInputValidationPaneTitleStyles.defaultInputValidationPaneTitleStyles, (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.titleStyleProps)
  };
  const subtitleStyles = {
    root: Object.assign({}, _defaultInputValidationPaneSubtitleStyles.defaultInputValidationPaneSubtitleStyles, (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.subtitleStyleProps)
  };
  const redBorderStyles = {
    borderColor: ((_props$controlProps14 = props.controlProps) === null || _props$controlProps14 === void 0 ? void 0 : _props$controlProps14.inputWithErrorMessageBorderColor) ?? _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputWithErrorMessageBorderColor,
    borderRadius: "1px",
    borderStyle: "solid",
    borderWidth: "1px"
  };
  const inputStyles = {
    root: Object.assign({}, _defaultInputValidationPaneInputStyles.defaultInputValidationPaneInputStyles, (_props$styleProps5 = props.styleProps) === null || _props$styleProps5 === void 0 ? void 0 : _props$styleProps5.inputStyleProps),
    field: Object.assign({}, _defaultInputValidationPaneInputStyles.defaultInputValidationPaneInputStyles, (_props$styleProps6 = props.styleProps) === null || _props$styleProps6 === void 0 ? void 0 : _props$styleProps6.inputStyleProps),
    fieldGroup: undefined,
    prefix: undefined,
    suffix: undefined,
    icon: undefined,
    description: undefined,
    wrapper: undefined,
    errorMessage: undefined,
    subComponentStyles: {
      label: {}
    },
    revealButton: undefined,
    revealSpan: undefined,
    revealIcon: undefined
  };
  const inputWithErrorMessageStyles = {
    root: Object.assign({}, _defaultInputValidationPaneInputStyles.defaultInputValidationPaneInputStyles, redBorderStyles, (_props$styleProps7 = props.styleProps) === null || _props$styleProps7 === void 0 ? void 0 : _props$styleProps7.inputStyleProps),
    field: Object.assign({}, _defaultInputValidationPaneInputStyles.defaultInputValidationPaneInputStyles, redBorderStyles, (_props$styleProps8 = props.styleProps) === null || _props$styleProps8 === void 0 ? void 0 : _props$styleProps8.inputStyleProps),
    fieldGroup: undefined,
    prefix: undefined,
    suffix: undefined,
    icon: undefined,
    description: undefined,
    wrapper: undefined,
    errorMessage: undefined,
    subComponentStyles: {
      label: {}
    },
    revealButton: undefined,
    revealSpan: undefined,
    revealIcon: undefined
  };
  const invalidInputErrorMessageStyles = {
    root: Object.assign({}, _defaultInputValidationPaneInvalidInputErrorMessageStyles.defaultInputValidationPaneInvalidInputErrorMessageStyles, (_props$styleProps9 = props.styleProps) === null || _props$styleProps9 === void 0 ? void 0 : _props$styleProps9.invalidInputErrorMessageStyleProps)
  };
  const buttonGroupStyles = {
    root: Object.assign({}, _defaultInputValidationPaneButtonGroupStyles.defaultInputValidationPaneButtonGroupStyles, (_props$styleProps10 = props.styleProps) === null || _props$styleProps10 === void 0 ? void 0 : _props$styleProps10.buttonGroupStyleProps)
  };
  const sendButtonStyles = {
    root: Object.assign({}, _defaultInputValidationPaneSendButtonStyles.defaultInputValidationPaneSendButtonStyles, (_props$styleProps11 = props.styleProps) === null || _props$styleProps11 === void 0 ? void 0 : _props$styleProps11.sendButtonStyleProps),
    rootHovered: Object.assign({}, _defaultInputValidationPaneSendButtonHoveredStyles.defaultInputValidationPaneSendButtonHoveredStyles, (_props$styleProps12 = props.styleProps) === null || _props$styleProps12 === void 0 ? void 0 : _props$styleProps12.sendButtonHoveredStyleProps),
    rootPressed: Object.assign({}, _defaultInputValidationPaneSendButtonHoveredStyles.defaultInputValidationPaneSendButtonHoveredStyles, (_props$styleProps13 = props.styleProps) === null || _props$styleProps13 === void 0 ? void 0 : _props$styleProps13.sendButtonHoveredStyleProps)
  };
  const cancelButtonStyles = {
    root: Object.assign({}, _defaultInputValidationPaneCancelButtonStyles.defaultInputValidationPaneCancelButtonStyles, (_props$styleProps14 = props.styleProps) === null || _props$styleProps14 === void 0 ? void 0 : _props$styleProps14.cancelButtonStyleProps),
    rootHovered: Object.assign({}, _defaultInputValidationPaneCancelButtonHoveredStyles.defaultInputValidationPaneCancelButtonHoveredStyles, (_props$styleProps15 = props.styleProps) === null || _props$styleProps15 === void 0 ? void 0 : _props$styleProps15.cancelButtonHoveredStyleProps),
    rootPressed: Object.assign({}, _defaultInputValidationPaneCancelButtonHoveredStyles.defaultInputValidationPaneCancelButtonHoveredStyles, (_props$styleProps16 = props.styleProps) === null || _props$styleProps16 === void 0 ? void 0 : _props$styleProps16.cancelButtonHoveredStyleProps)
  };
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, !((_props$controlProps15 = props.controlProps) !== null && _props$controlProps15 !== void 0 && _props$controlProps15.hideInputValidationPane) && /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps17 = props.styleProps) === null || _props$styleProps17 === void 0 ? void 0 : (_props$styleProps17$c = _props$styleProps17.classNames) === null || _props$styleProps17$c === void 0 ? void 0 : _props$styleProps17$c.containerClassName,
    id: elementId,
    tabIndex: -1,
    onKeyDown: handleEscKeyDown,
    dir: ((_props$controlProps16 = props.controlProps) === null || _props$controlProps16 === void 0 ? void 0 : _props$controlProps16.dir) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.dir,
    "aria-label": ((_props$controlProps17 = props.controlProps) === null || _props$controlProps17 === void 0 ? void 0 : _props$controlProps17.inputValidationPaneAriaLabel) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputValidationPaneAriaLabel,
    styles: containerStyles
  }, /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps18 = props.styleProps) === null || _props$styleProps18 === void 0 ? void 0 : (_props$styleProps18$c = _props$styleProps18.classNames) === null || _props$styleProps18$c === void 0 ? void 0 : _props$styleProps18$c.headerGroupClassName,
    styles: headerGroupStyles,
    tabIndex: -1,
    id: elementId + "-headergroup"
  }, !((_props$controlProps18 = props.controlProps) !== null && _props$controlProps18 !== void 0 && _props$controlProps18.hideTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr = props.componentOverrides) === null || _props$componentOverr === void 0 ? void 0 : _props$componentOverr.title) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps19 = props.styleProps) === null || _props$styleProps19 === void 0 ? void 0 : (_props$styleProps19$c = _props$styleProps19.classNames) === null || _props$styleProps19$c === void 0 ? void 0 : _props$styleProps19$c.titleClassName,
    styles: titleStyles,
    tabIndex: -1,
    id: elementId + "-title"
  }, ((_props$controlProps19 = props.controlProps) === null || _props$controlProps19 === void 0 ? void 0 : _props$controlProps19.titleText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.titleText)), !((_props$controlProps20 = props.controlProps) !== null && _props$controlProps20 !== void 0 && _props$controlProps20.hideSubtitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr2 = props.componentOverrides) === null || _props$componentOverr2 === void 0 ? void 0 : _props$componentOverr2.subtitle) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps20 = props.styleProps) === null || _props$styleProps20 === void 0 ? void 0 : (_props$styleProps20$c = _props$styleProps20.classNames) === null || _props$styleProps20$c === void 0 ? void 0 : _props$styleProps20$c.subtitleClassName,
    styles: subtitleStyles,
    tabIndex: -1,
    id: elementId + "-subtitle"
  }, ((_props$controlProps21 = props.controlProps) === null || _props$controlProps21 === void 0 ? void 0 : _props$controlProps21.subtitleText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.subtitleText)), !((_props$controlProps22 = props.controlProps) !== null && _props$controlProps22 !== void 0 && _props$controlProps22.hideInput) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr3 = props.componentOverrides) === null || _props$componentOverr3 === void 0 ? void 0 : _props$componentOverr3.input) || /*#__PURE__*/_react2.default.createElement(_react.TextField, {
    className: (_props$styleProps21 = props.styleProps) === null || _props$styleProps21 === void 0 ? void 0 : (_props$styleProps21$c = _props$styleProps21.classNames) === null || _props$styleProps21$c === void 0 ? void 0 : _props$styleProps21$c.inputClassName,
    styles: isInvalidInput ? inputWithErrorMessageStyles : inputStyles,
    tabIndex: 0,
    value: inputValue ?? _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputInitialText,
    id: ((_props$controlProps23 = props.controlProps) === null || _props$controlProps23 === void 0 ? void 0 : _props$controlProps23.inputId) ?? _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputId,
    ariaLabel: ((_props$controlProps24 = props.controlProps) === null || _props$controlProps24 === void 0 ? void 0 : _props$controlProps24.inputAriaLabel) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputAriaLabel,
    borderless: isInvalidInput,
    onChange: handleInputChange,
    onKeyUp: handleInputKeyUp,
    placeholder: ((_props$controlProps25 = props.controlProps) === null || _props$controlProps25 === void 0 ? void 0 : _props$controlProps25.inputPlaceHolder) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.inputPlaceHolder
  })), isInvalidInput && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr4 = props.componentOverrides) === null || _props$componentOverr4 === void 0 ? void 0 : _props$componentOverr4.invalidInputErrorMessage) || /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps22 = props.styleProps) === null || _props$styleProps22 === void 0 ? void 0 : (_props$styleProps22$c = _props$styleProps22.classNames) === null || _props$styleProps22$c === void 0 ? void 0 : _props$styleProps22$c.invalidInputErrorMessageClassName,
    styles: invalidInputErrorMessageStyles,
    "aria-label": ((_props$controlProps26 = props.controlProps) === null || _props$controlProps26 === void 0 ? void 0 : _props$controlProps26.invalidInputErrorMessageText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.invalidInputErrorMessageText,
    tabIndex: -1,
    role: "alert",
    id: elementId + "-invalidinputerrormessage"
  }, ((_props$controlProps27 = props.controlProps) === null || _props$controlProps27 === void 0 ? void 0 : _props$controlProps27.invalidInputErrorMessageText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.invalidInputErrorMessageText))), /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    horizontal: ((_props$controlProps28 = props.controlProps) === null || _props$controlProps28 === void 0 ? void 0 : _props$controlProps28.isButtonGroupHorizontal) ?? _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.isButtonGroupHorizontal,
    className: (_props$styleProps23 = props.styleProps) === null || _props$styleProps23 === void 0 ? void 0 : (_props$styleProps23$c = _props$styleProps23.classNames) === null || _props$styleProps23$c === void 0 ? void 0 : _props$styleProps23$c.buttonGroupClassName,
    styles: buttonGroupStyles,
    tabIndex: -1,
    id: elementId + "-buttongroup"
  }, !((_props$controlProps29 = props.controlProps) !== null && _props$controlProps29 !== void 0 && _props$controlProps29.hideSendButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr5 = props.componentOverrides) === null || _props$componentOverr5 === void 0 ? void 0 : _props$componentOverr5.sendButton) || /*#__PURE__*/_react2.default.createElement(_Button.PrimaryButton, {
    className: (_props$styleProps24 = props.styleProps) === null || _props$styleProps24 === void 0 ? void 0 : (_props$styleProps24$c = _props$styleProps24.classNames) === null || _props$styleProps24$c === void 0 ? void 0 : _props$styleProps24$c.sendButtonClassName,
    styles: sendButtonStyles,
    title: ((_props$controlProps30 = props.controlProps) === null || _props$controlProps30 === void 0 ? void 0 : _props$controlProps30.sendButtonText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.sendButtonText,
    tabIndex: 0,
    disabled: !isSendButtonEnabled,
    text: ((_props$controlProps31 = props.controlProps) === null || _props$controlProps31 === void 0 ? void 0 : _props$controlProps31.sendButtonText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.sendButtonText,
    onClick: handleSendClick,
    id: elementId + "-sendbutton",
    ariaLabel: ((_props$controlProps32 = props.controlProps) === null || _props$controlProps32 === void 0 ? void 0 : _props$controlProps32.sendButtonAriaLabel) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.sendButtonAriaLabel
  })), !((_props$controlProps33 = props.controlProps) !== null && _props$controlProps33 !== void 0 && _props$controlProps33.hideCancelButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr6 = props.componentOverrides) === null || _props$componentOverr6 === void 0 ? void 0 : _props$componentOverr6.cancelButton) || /*#__PURE__*/_react2.default.createElement(_Button.DefaultButton, {
    className: (_props$styleProps25 = props.styleProps) === null || _props$styleProps25 === void 0 ? void 0 : (_props$styleProps25$c = _props$styleProps25.classNames) === null || _props$styleProps25$c === void 0 ? void 0 : _props$styleProps25$c.cancelButtonClassName,
    styles: cancelButtonStyles,
    title: ((_props$controlProps34 = props.controlProps) === null || _props$controlProps34 === void 0 ? void 0 : _props$controlProps34.cancelButtonText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.cancelButtonText,
    tabIndex: 0,
    text: ((_props$controlProps35 = props.controlProps) === null || _props$controlProps35 === void 0 ? void 0 : _props$controlProps35.cancelButtonText) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.cancelButtonText,
    onClick: handleCancelClick,
    id: elementId + "-cancelbutton",
    ariaLabel: ((_props$controlProps36 = props.controlProps) === null || _props$controlProps36 === void 0 ? void 0 : _props$controlProps36.cancelButtonAriaLabel) || _defaultInputValidationPaneControlProps.defaultInputValidationPaneControlProps.cancelButtonAriaLabel
  })))));
}

var _default = InputValidationPane;
exports.default = _default;
