// @microsoft/omnichannel-chat-widget/lib/cjs/components/webchatcontainerstateful/common/defaultProps/defaultWebChatContainerStatefulProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultWebChatContainerStatefulProps = void 0;
var _mockadapter = _interopRequireDefault(require("../mockadapter"));
var _defaultWebChatStatefulContainerStyles = require("../defaultStyles/defaultWebChatStatefulContainerStyles");
var _defaultWebChatStatefulProps = require("./defaultWebChatStatefulProps");
var _defaultWebChatStyles = require("../defaultStyles/defaultWebChatStyles");
var _defaultAdaptiveCardStyles = require("../defaultStyles/defaultAdaptiveCardStyles");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const defaultWebChatContainerStatefulProps = {
  webChatStyles: _defaultWebChatStyles.defaultWebChatStyles,
  webChatProps: _defaultWebChatStatefulProps.defaultWebChatStatefulProps,
  containerStyles: _defaultWebChatStatefulContainerStyles.defaultWebChatStatefulContainerStyles,
  disableNewLineMarkdownSupport: false,
  disableMarkdownMessageFormatting: false,
  hyperlinkTextOverride: false,
  directLine: new _mockadapter.default(),
  adaptiveCardStyles: _defaultAdaptiveCardStyles.defaultAdaptiveCardStyles
};
exports.defaultWebChatContainerStatefulProps = defaultWebChatContainerStatefulProps;
