// @microsoft/omnichannel-chat-components/lib/cjs/components/reconnectchatpane/common/default/defaultProps/defaultReconnectChatPaneControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultReconnectChatPaneControlProps = void 0;

var _Constants = require("../../../../../common/Constants");

const defaultReconnectChatPaneControlProps = {
  id: _Constants.Ids.DefaultReconnectChatPaneId,
  dir: "ltr",
  hideReconnectChatPane: false,
  reconnectChatPaneAriaLabel: _Constants.AriaLabels.ReconnectChatPane,
  hideTitle: false,
  titleText: _Constants.Texts.ReconnectChatPaneTitleText,
  hideSubtitle: false,
  subtitleText: _Constants.Texts.ReconnectChatPaneSubtitleText,
  hideIcon: false,
  iconAriaLabel: _Constants.AriaLabels.ReconnectChatPaneIcon,
  isButtonGroupHorizontal: false,
  hideContinueChatButton: false,
  continueChatButtonText: _Constants.Texts.ReconnectChatPaneContinueChatButtonText,
  continueChatButtonAriaLabel: _Constants.AriaLabels.ContinueConversation,
  hideStartNewChatButton: false,
  startNewChatButtonText: _Constants.Texts.ReconnectChatPaneStartNewChatButtonText,
  startNewChatButtonAriaLabel: _Constants.AriaLabels.StartNewConversation,
  onContinueChat: function () {
    console.log("on continue conversation");
  },
  onStartNewChat: function () {
    console.log("on start new conversation");
  },
  onMinimize: function () {
    console.log("on minimize");
  }
};
exports.defaultReconnectChatPaneControlProps = defaultReconnectChatPaneControlProps;
