// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/ProactiveChatPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("@fluentui/react");

var _react2 = _interopRequireWildcard(require("react"));

var _BroadcastService = require("../../services/BroadcastService");

var _Button = require("@fluentui/react/lib/Button");

var _CloseButton = _interopRequireDefault(require("../common/subcomponents/CloseButton"));

var _Constants = require("../../common/Constants");

var _decodeComponentString = require("../../common/decodeComponentString");

var _defaultProactiveChatPaneBodyContainerStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneBodyContainerStyles");

var _defaultProactiveChatPaneBodyTitleStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneBodyTitleStyles");

var _defaultProactiveChatPaneControlProps = require("./common/default/defaultProps/defaultProactiveChatPaneControlProps");

var _defaultProactiveChatPaneGeneralStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneGeneralStyles");

var _defaultProactiveChatPaneHeaderContainerStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneHeaderContainerStyles");

var _defaultProactiveChatPaneStartButtonHoveredStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneStartButtonHoveredStyles");

var _defaultProactiveChatPaneStartButtonStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneStartButtonStyles");

var _defaultProactiveChatPaneSubtitleStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneSubtitleStyles");

var _defaultProactiveChatPaneTextContainerStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneTextContainerStyles");

var _defaultProactiveChatPaneTitleStyles = require("./common/default/defaultStyles/defaultProactiveChatPaneTitleStyles");

var _utils = require("../../common/utils");

var _defaultProactiveChatPaneProps = require("./common/default/defaultProps/defaultProactiveChatPaneProps");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ProactiveChatPane(props) {
  var _props$controlProps, _defaultProactiveChat, _props$controlProps6, _props$styleProps, _props$styleProps2, _props$styleProps3, _props$styleProps4, _props$styleProps5, _defaultProactiveChat2, _props$styleProps6, _defaultProactiveChat3, _props$styleProps7, _props$styleProps8, _props$styleProps9, _props$styleProps10, _props$styleProps11, _props$styleProps12, _props$controlProps7, _props$styleProps13, _props$styleProps13$c, _props$controlProps8, _props$controlProps9, _props$styleProps14, _props$styleProps14$c, _props$styleProps15, _props$styleProps15$c, _props$controlProps10, _props$componentOverr, _props$styleProps16, _props$styleProps16$c, _props$controlProps11, _props$controlProps12, _props$componentOverr2, _props$styleProps17, _props$styleProps17$c, _props$controlProps13, _props$controlProps14, _props$componentOverr3, _props$styleProps18, _props$styleProps18$c, _props$controlProps15, _props$styleProps19, _props$styleProps19$c, _props$controlProps16, _props$componentOverr4, _props$styleProps20, _props$styleProps20$c, _props$controlProps17, _props$controlProps18, _props$componentOverr5, _props$styleProps21, _props$styleProps21$c, _props$controlProps19, _props$controlProps20;

  const elementId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.id;

  const close = (controlId, suffix) => {
    var _props$controlProps2;

    if ((_props$controlProps2 = props.controlProps) !== null && _props$controlProps2 !== void 0 && _props$controlProps2.onClose) {
      var _props$controlProps3;

      const performance = window.performance;
      const startTime = performance.now();
      (_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.onClose();
      const endTime = performance.now();
      const payload = {
        elapsedTimeInMilliseconds: endTime - startTime
      };
      const eventName = (0, _utils.generateEventName)(controlId, "on", suffix);
      const customEvent = {
        eventName: eventName + " - Proactive Chat Rejected",
        payload: payload
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);
    }
  }; // eslint-disable-next-line @typescript-eslint/no-explicit-any


  const handleEscKeyDown = (0, _react2.useCallback)(e => {
    if (e.code === _Constants.KeyCodes.ESCAPE) {
      close(elementId, "KeyDown");
    }
  }, []);
  const handleCloseClick = (0, _react2.useCallback)(() => {
    close(elementId + "-closebutton", "Click");
  }, []);
  const handleStartClick = (0, _react2.useCallback)(() => {
    var _props$controlProps4;

    if ((_props$controlProps4 = props.controlProps) !== null && _props$controlProps4 !== void 0 && _props$controlProps4.onStart) {
      var _props$controlProps5;

      const onClickEventName = (0, _utils.generateEventName)(elementId + "-startbutton", "on", "Click");
      const onClickEvent = {
        eventName: onClickEventName + " - Proactive Chat Accepted"
      };

      _BroadcastService.BroadcastService.postMessage(onClickEvent);

      (_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.onStart();
    }
  }, []);
  const closeButtonProps = Object.assign({}, (_defaultProactiveChat = _defaultProactiveChatPaneProps.defaultProactiveChatPaneProps.controlProps) === null || _defaultProactiveChat === void 0 ? void 0 : _defaultProactiveChat.closeButtonProps, (_props$controlProps6 = props.controlProps) === null || _props$controlProps6 === void 0 ? void 0 : _props$controlProps6.closeButtonProps);
  const containerStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneGeneralStyles.defaultProactiveChatPaneGeneralStyles, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const headerContainerStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneHeaderContainerStyles.defaultProactiveChatPaneHeaderContainerStyles, (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.headerContainerStyleProps)
  };
  const textContainerStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneTextContainerStyles.defaultProactiveChatPaneTextContainerStyles, (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.textContainerStyleProps)
  };
  const titleStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneTitleStyles.defaultProactiveChatPaneTitleStyles, (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.titleStyleProps)
  };
  const subtitleStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneSubtitleStyles.defaultProactiveChatPaneSubtitleStyles, (_props$styleProps5 = props.styleProps) === null || _props$styleProps5 === void 0 ? void 0 : _props$styleProps5.subtitleStyleProps)
  };
  const closeButtonStyles = Object.assign({}, (_defaultProactiveChat2 = _defaultProactiveChatPaneProps.defaultProactiveChatPaneProps.styleProps) === null || _defaultProactiveChat2 === void 0 ? void 0 : _defaultProactiveChat2.closeButtonStyleProps, (_props$styleProps6 = props.styleProps) === null || _props$styleProps6 === void 0 ? void 0 : _props$styleProps6.closeButtonStyleProps);
  const closeButtonHoverStyles = Object.assign({}, (_defaultProactiveChat3 = _defaultProactiveChatPaneProps.defaultProactiveChatPaneProps.styleProps) === null || _defaultProactiveChat3 === void 0 ? void 0 : _defaultProactiveChat3.closeButtonHoveredStyleProps, (_props$styleProps7 = props.styleProps) === null || _props$styleProps7 === void 0 ? void 0 : _props$styleProps7.closeButtonHoveredStyleProps);
  const bodyContainerStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneBodyContainerStyles.defaultProactiveChatPaneBodyContainerStyles, (_props$styleProps8 = props.styleProps) === null || _props$styleProps8 === void 0 ? void 0 : _props$styleProps8.bodyContainerStyleProps)
  };
  const bodyTitleStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneBodyTitleStyles.defaultProactiveChatPaneBodyTitleStyles, (_props$styleProps9 = props.styleProps) === null || _props$styleProps9 === void 0 ? void 0 : _props$styleProps9.bodyTitleStyleProps)
  };
  const startButtonStyles = {
    root: Object.assign({}, _defaultProactiveChatPaneStartButtonStyles.defaultProactiveChatPaneStartButtonStyles, (_props$styleProps10 = props.styleProps) === null || _props$styleProps10 === void 0 ? void 0 : _props$styleProps10.startButtonStyleProps),
    rootHovered: Object.assign({}, _defaultProactiveChatPaneStartButtonHoveredStyles.defaultProactiveChatPaneStartButtonHoveredStyles, (_props$styleProps11 = props.styleProps) === null || _props$styleProps11 === void 0 ? void 0 : _props$styleProps11.startButtonHoveredStyleProps),
    rootPressed: Object.assign({}, _defaultProactiveChatPaneStartButtonHoveredStyles.defaultProactiveChatPaneStartButtonHoveredStyles, (_props$styleProps12 = props.styleProps) === null || _props$styleProps12 === void 0 ? void 0 : _props$styleProps12.startButtonHoveredStyleProps)
  };
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, !((_props$controlProps7 = props.controlProps) !== null && _props$controlProps7 !== void 0 && _props$controlProps7.hideProactiveChatPane) && /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps13 = props.styleProps) === null || _props$styleProps13 === void 0 ? void 0 : (_props$styleProps13$c = _props$styleProps13.classNames) === null || _props$styleProps13$c === void 0 ? void 0 : _props$styleProps13$c.containerClassName,
    id: elementId,
    tabIndex: -1,
    onKeyDown: handleEscKeyDown,
    dir: ((_props$controlProps8 = props.controlProps) === null || _props$controlProps8 === void 0 ? void 0 : _props$controlProps8.dir) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.dir,
    "aria-label": ((_props$controlProps9 = props.controlProps) === null || _props$controlProps9 === void 0 ? void 0 : _props$controlProps9.proactiveChatPaneAriaLabel) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.proactiveChatPaneAriaLabel,
    styles: containerStyles
  }, /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    horizontal: true,
    className: (_props$styleProps14 = props.styleProps) === null || _props$styleProps14 === void 0 ? void 0 : (_props$styleProps14$c = _props$styleProps14.classNames) === null || _props$styleProps14$c === void 0 ? void 0 : _props$styleProps14$c.headerContainerClassName,
    styles: headerContainerStyles,
    tabIndex: -1,
    id: elementId + "-headercontainer"
  }, /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps15 = props.styleProps) === null || _props$styleProps15 === void 0 ? void 0 : (_props$styleProps15$c = _props$styleProps15.classNames) === null || _props$styleProps15$c === void 0 ? void 0 : _props$styleProps15$c.textContainerClassName,
    styles: textContainerStyles,
    tabIndex: -1,
    id: elementId + "-textcontainer"
  }, !((_props$controlProps10 = props.controlProps) !== null && _props$controlProps10 !== void 0 && _props$controlProps10.hideTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr = props.componentOverrides) === null || _props$componentOverr === void 0 ? void 0 : _props$componentOverr.title) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps16 = props.styleProps) === null || _props$styleProps16 === void 0 ? void 0 : (_props$styleProps16$c = _props$styleProps16.classNames) === null || _props$styleProps16$c === void 0 ? void 0 : _props$styleProps16$c.titleClassName,
    styles: titleStyles,
    tabIndex: -1,
    id: elementId + "-title"
  }, ((_props$controlProps11 = props.controlProps) === null || _props$controlProps11 === void 0 ? void 0 : _props$controlProps11.titleText) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.titleText)), !((_props$controlProps12 = props.controlProps) !== null && _props$controlProps12 !== void 0 && _props$controlProps12.hideSubtitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr2 = props.componentOverrides) === null || _props$componentOverr2 === void 0 ? void 0 : _props$componentOverr2.subtitle) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps17 = props.styleProps) === null || _props$styleProps17 === void 0 ? void 0 : (_props$styleProps17$c = _props$styleProps17.classNames) === null || _props$styleProps17$c === void 0 ? void 0 : _props$styleProps17$c.subtitleClassName,
    styles: subtitleStyles,
    tabIndex: -1,
    id: elementId + "-subtitle"
  }, ((_props$controlProps13 = props.controlProps) === null || _props$controlProps13 === void 0 ? void 0 : _props$controlProps13.subtitleText) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.subtitleText))), !((_props$controlProps14 = props.controlProps) !== null && _props$controlProps14 !== void 0 && _props$controlProps14.hideCloseButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr3 = props.componentOverrides) === null || _props$componentOverr3 === void 0 ? void 0 : _props$componentOverr3.closeButton) || /*#__PURE__*/_react2.default.createElement(_CloseButton.default, _extends({}, closeButtonProps, {
    className: (_props$styleProps18 = props.styleProps) === null || _props$styleProps18 === void 0 ? void 0 : (_props$styleProps18$c = _props$styleProps18.classNames) === null || _props$styleProps18$c === void 0 ? void 0 : _props$styleProps18$c.closeButtonClassName,
    onClick: handleCloseClick,
    styles: closeButtonStyles,
    hoverStyles: closeButtonHoverStyles,
    id: elementId + "-closebutton"
  })))), /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    horizontal: ((_props$controlProps15 = props.controlProps) === null || _props$controlProps15 === void 0 ? void 0 : _props$controlProps15.isBodyContainerHorizantal) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.isBodyContainerHorizantal,
    className: (_props$styleProps19 = props.styleProps) === null || _props$styleProps19 === void 0 ? void 0 : (_props$styleProps19$c = _props$styleProps19.classNames) === null || _props$styleProps19$c === void 0 ? void 0 : _props$styleProps19$c.bodyContainerClassName,
    styles: bodyContainerStyles,
    tabIndex: -1,
    id: elementId + "-bodycontainer"
  }, !((_props$controlProps16 = props.controlProps) !== null && _props$controlProps16 !== void 0 && _props$controlProps16.hideBodyTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr4 = props.componentOverrides) === null || _props$componentOverr4 === void 0 ? void 0 : _props$componentOverr4.bodyTitle) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps20 = props.styleProps) === null || _props$styleProps20 === void 0 ? void 0 : (_props$styleProps20$c = _props$styleProps20.classNames) === null || _props$styleProps20$c === void 0 ? void 0 : _props$styleProps20$c.bodyTitleClassName,
    styles: bodyTitleStyles,
    tabIndex: -1,
    id: elementId + "-bodytitle"
  }, ((_props$controlProps17 = props.controlProps) === null || _props$controlProps17 === void 0 ? void 0 : _props$controlProps17.bodyTitleText) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.bodyTitleText)), !((_props$controlProps18 = props.controlProps) !== null && _props$controlProps18 !== void 0 && _props$controlProps18.hideStartButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr5 = props.componentOverrides) === null || _props$componentOverr5 === void 0 ? void 0 : _props$componentOverr5.startButton) || /*#__PURE__*/_react2.default.createElement(_Button.DefaultButton, {
    className: (_props$styleProps21 = props.styleProps) === null || _props$styleProps21 === void 0 ? void 0 : (_props$styleProps21$c = _props$styleProps21.classNames) === null || _props$styleProps21$c === void 0 ? void 0 : _props$styleProps21$c.startButtonClassName,
    styles: startButtonStyles,
    tabIndex: 0,
    text: ((_props$controlProps19 = props.controlProps) === null || _props$controlProps19 === void 0 ? void 0 : _props$controlProps19.startButtonText) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.startButtonText,
    onClick: handleStartClick,
    id: elementId + "-startbutton",
    "aria-label": ((_props$controlProps20 = props.controlProps) === null || _props$controlProps20 === void 0 ? void 0 : _props$controlProps20.startButtonAriaLabel) || _defaultProactiveChatPaneControlProps.defaultProactiveChatPaneControlProps.startButtonAriaLabel
  })))));
}

var _default = ProactiveChatPane;
exports.default = _default;
