// @microsoft/omnichannel-chat-components/lib/cjs/components/confirmationpane/ConfirmationPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Button = require("@fluentui/react/lib/Button");

var _react = require("@fluentui/react");

var _react2 = _interopRequireWildcard(require("react"));

var _BroadcastService = require("../../services/BroadcastService");

var _Constants = require("../../common/Constants");

var _decodeComponentString = require("../../common/decodeComponentString");

var _defaultConfirmationPaneButtonGroupStyles = require("./common/defaultStyles/defaultConfirmationPaneButtonGroupStyles");

var _defaultConfirmationPaneCancelButtonHoveredStyles = require("./common/defaultStyles/defaultConfirmationPaneCancelButtonHoveredStyles");

var _defaultConfirmationPaneCancelButtonStyles = require("./common/defaultStyles/defaultConfirmationPaneCancelButtonStyles");

var _defaultConfirmationPaneConfirmButtonHoveredStyles = require("./common/defaultStyles/defaultConfirmationPaneConfirmButtonHoveredStyles");

var _defaultConfirmationPaneConfirmButtonStyles = require("./common/defaultStyles/defaultConfirmationPaneConfirmButtonStyles");

var _defaultConfirmationPaneControlProps = require("./common/defaultProps/defaultConfirmationPaneControlProps");

var _defaultConfirmationPaneGeneralStyles = require("./common/defaultStyles/defaultConfirmationPaneGeneralStyles");

var _defaultConfirmationPaneSubtitleStyles = require("./common/defaultStyles/defaultConfirmationPaneSubtitleStyles");

var _defaultConfirmationPaneTitleStyles = require("./common/defaultStyles/defaultConfirmationPaneTitleStyles");

var _defaultConfirmationPaneConfirmButtonFocusedStyles = require("./common/defaultStyles/defaultConfirmationPaneConfirmButtonFocusedStyles");

var _defaultConfirmationPaneCancelButtonFocusedStyles = require("./common/defaultStyles/defaultConfirmationPaneCancelButtonFocusedStyles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ConfirmationPane(props) {
  var _props$controlProps, _props$styleProps, _props$styleProps2, _props$styleProps3, _props$styleProps4, _props$styleProps5, _props$styleProps6, _props$styleProps7, _props$styleProps8, _props$styleProps9, _props$styleProps10, _props$styleProps11, _props$styleProps12, _props$controlProps6, _props$controlProps7, _props$controlProps8, _props$componentOverr, _props$styleProps13, _props$styleProps13$c, _props$controlProps9, _props$controlProps10, _props$componentOverr2, _props$styleProps14, _props$styleProps14$c, _props$controlProps11, _props$styleProps15, _props$styleProps15$c, _props$controlProps12, _props$componentOverr3, _props$styleProps16, _props$styleProps16$c, _props$controlProps13, _props$controlProps14, _props$controlProps15, _props$componentOverr4, _props$styleProps17, _props$styleProps17$c, _props$controlProps16, _props$controlProps17;

  const elementId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.id;
  const handleConfirmClick = (0, _react2.useCallback)(() => {
    var _props$controlProps2;

    if ((_props$controlProps2 = props.controlProps) !== null && _props$controlProps2 !== void 0 && _props$controlProps2.onConfirm) {
      var _props$controlProps3;

      const customEvent = {
        elementType: _Constants.ElementType.ConfirmationPaneConfirmButton,
        elementId: elementId + "-confirmbutton",
        eventName: _Constants.EventNames.OnClick
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.onConfirm();
    }
  }, []);
  const handleCancelClick = (0, _react2.useCallback)(() => {
    var _props$controlProps4;

    if ((_props$controlProps4 = props.controlProps) !== null && _props$controlProps4 !== void 0 && _props$controlProps4.onCancel) {
      var _props$controlProps5;

      const customEvent = {
        elementType: _Constants.ElementType.ConfirmationPaneCancelButton,
        elementId: elementId + "-cancelbutton",
        eventName: _Constants.EventNames.OnClick
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.onCancel();
    }
  }, []); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const handleEscKeyDown = (0, _react2.useCallback)(e => {
    if (e.code === _Constants.KeyCodes.ESCAPE) {
      handleCancelClick();
    }
  }, []);
  const containerStyles = {
    root: Object.assign({}, _defaultConfirmationPaneGeneralStyles.defaultConfirmationPaneGeneralStyles, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const titleStyles = {
    root: Object.assign({}, _defaultConfirmationPaneTitleStyles.defaultConfirmationPaneTitleStyles, (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.titleStyleProps)
  };
  const subtitleStyles = {
    root: Object.assign({}, _defaultConfirmationPaneSubtitleStyles.defaultConfirmationPaneSubtitleStyles, (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.subtitleStyleProps)
  };
  const buttonGroupStyles = {
    root: Object.assign({}, _defaultConfirmationPaneButtonGroupStyles.defaultConfirmationPaneButtonGroupStyles, (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.buttonGroupStyleProps)
  };
  const confirmButtonStyles = {
    root: Object.assign({}, _defaultConfirmationPaneConfirmButtonStyles.defaultConfirmationPaneConfirmButtonStyles, (_props$styleProps5 = props.styleProps) === null || _props$styleProps5 === void 0 ? void 0 : _props$styleProps5.confirmButtonStyleProps),
    rootHovered: Object.assign({}, _defaultConfirmationPaneConfirmButtonHoveredStyles.defaultConfirmationPaneConfirmButtonHoveredStyles, (_props$styleProps6 = props.styleProps) === null || _props$styleProps6 === void 0 ? void 0 : _props$styleProps6.confirmButtonHoveredStyleProps),
    rootFocused: Object.assign({}, _defaultConfirmationPaneConfirmButtonFocusedStyles.defaultConfirmationPaneConfirmButtonFocusedStyles, (_props$styleProps7 = props.styleProps) === null || _props$styleProps7 === void 0 ? void 0 : _props$styleProps7.confirmButtonFocusedStyleProps),
    rootPressed: Object.assign({}, _defaultConfirmationPaneConfirmButtonHoveredStyles.defaultConfirmationPaneConfirmButtonHoveredStyles, (_props$styleProps8 = props.styleProps) === null || _props$styleProps8 === void 0 ? void 0 : _props$styleProps8.confirmButtonHoveredStyleProps)
  };
  const cancelButtonStyles = {
    root: Object.assign({}, _defaultConfirmationPaneCancelButtonStyles.defaultConfirmationPaneCancelButtonStyles, (_props$styleProps9 = props.styleProps) === null || _props$styleProps9 === void 0 ? void 0 : _props$styleProps9.cancelButtonStyleProps),
    rootHovered: Object.assign({}, _defaultConfirmationPaneCancelButtonHoveredStyles.defaultConfirmationPaneCancelButtonHoveredStyles, (_props$styleProps10 = props.styleProps) === null || _props$styleProps10 === void 0 ? void 0 : _props$styleProps10.cancelButtonHoveredStyleProps),
    rootFocused: Object.assign({}, _defaultConfirmationPaneCancelButtonFocusedStyles.defaultConfirmationPaneCancelButtonFocusedStyles, (_props$styleProps11 = props.styleProps) === null || _props$styleProps11 === void 0 ? void 0 : _props$styleProps11.cancelButtonFocusedStyleProps),
    rootPressed: Object.assign({}, _defaultConfirmationPaneCancelButtonHoveredStyles.defaultConfirmationPaneCancelButtonHoveredStyles, (_props$styleProps12 = props.styleProps) === null || _props$styleProps12 === void 0 ? void 0 : _props$styleProps12.cancelButtonHoveredStyleProps)
  };
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, !((_props$controlProps6 = props.controlProps) !== null && _props$controlProps6 !== void 0 && _props$controlProps6.hideConfirmationPane) && /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    id: elementId,
    onKeyDown: handleEscKeyDown,
    tabIndex: -1,
    dir: ((_props$controlProps7 = props.controlProps) === null || _props$controlProps7 === void 0 ? void 0 : _props$controlProps7.dir) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.dir,
    styles: containerStyles,
    role: "dialog",
    "aria-labelledby": elementId + "-title",
    "aria-describedby": elementId + "-subtitle"
  }, !((_props$controlProps8 = props.controlProps) !== null && _props$controlProps8 !== void 0 && _props$controlProps8.hideTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr = props.componentOverrides) === null || _props$componentOverr === void 0 ? void 0 : _props$componentOverr.title) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps13 = props.styleProps) === null || _props$styleProps13 === void 0 ? void 0 : (_props$styleProps13$c = _props$styleProps13.classNames) === null || _props$styleProps13$c === void 0 ? void 0 : _props$styleProps13$c.titleClassName,
    styles: titleStyles,
    tabIndex: -1,
    id: elementId + "-title"
  }, ((_props$controlProps9 = props.controlProps) === null || _props$controlProps9 === void 0 ? void 0 : _props$controlProps9.titleText) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.titleText)), !((_props$controlProps10 = props.controlProps) !== null && _props$controlProps10 !== void 0 && _props$controlProps10.hideSubtitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr2 = props.componentOverrides) === null || _props$componentOverr2 === void 0 ? void 0 : _props$componentOverr2.subtitle) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps14 = props.styleProps) === null || _props$styleProps14 === void 0 ? void 0 : (_props$styleProps14$c = _props$styleProps14.classNames) === null || _props$styleProps14$c === void 0 ? void 0 : _props$styleProps14$c.subtitleClassName,
    styles: subtitleStyles,
    tabIndex: -1,
    id: elementId + "-subtitle"
  }, ((_props$controlProps11 = props.controlProps) === null || _props$controlProps11 === void 0 ? void 0 : _props$controlProps11.subtitleText) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.subtitleText)), /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    horizontal: true,
    className: (_props$styleProps15 = props.styleProps) === null || _props$styleProps15 === void 0 ? void 0 : (_props$styleProps15$c = _props$styleProps15.classNames) === null || _props$styleProps15$c === void 0 ? void 0 : _props$styleProps15$c.buttonGroupClassName,
    styles: buttonGroupStyles,
    id: elementId + "-buttongroup"
  }, !((_props$controlProps12 = props.controlProps) !== null && _props$controlProps12 !== void 0 && _props$controlProps12.hideConfirmButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr3 = props.componentOverrides) === null || _props$componentOverr3 === void 0 ? void 0 : _props$componentOverr3.confirmButton) || /*#__PURE__*/_react2.default.createElement(_Button.PrimaryButton, {
    className: (_props$styleProps16 = props.styleProps) === null || _props$styleProps16 === void 0 ? void 0 : (_props$styleProps16$c = _props$styleProps16.classNames) === null || _props$styleProps16$c === void 0 ? void 0 : _props$styleProps16$c.confirmButtonClassName,
    styles: confirmButtonStyles,
    text: ((_props$controlProps13 = props.controlProps) === null || _props$controlProps13 === void 0 ? void 0 : _props$controlProps13.confirmButtonText) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.confirmButtonText,
    onClick: handleConfirmClick,
    id: elementId + "-confirmbutton",
    ariaLabel: ((_props$controlProps14 = props.controlProps) === null || _props$controlProps14 === void 0 ? void 0 : _props$controlProps14.confirmButtonAriaLabel) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.confirmButtonAriaLabel
  })), !((_props$controlProps15 = props.controlProps) !== null && _props$controlProps15 !== void 0 && _props$controlProps15.hideCancelButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr4 = props.componentOverrides) === null || _props$componentOverr4 === void 0 ? void 0 : _props$componentOverr4.cancelButton) || /*#__PURE__*/_react2.default.createElement(_Button.DefaultButton, {
    className: (_props$styleProps17 = props.styleProps) === null || _props$styleProps17 === void 0 ? void 0 : (_props$styleProps17$c = _props$styleProps17.classNames) === null || _props$styleProps17$c === void 0 ? void 0 : _props$styleProps17$c.cancelButtonClassName,
    styles: cancelButtonStyles,
    text: ((_props$controlProps16 = props.controlProps) === null || _props$controlProps16 === void 0 ? void 0 : _props$controlProps16.cancelButtonText) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.cancelButtonText,
    onClick: handleCancelClick,
    id: elementId + "-cancelbutton",
    ariaLabel: ((_props$controlProps17 = props.controlProps) === null || _props$controlProps17 === void 0 ? void 0 : _props$controlProps17.cancelButtonAriaLabel) || _defaultConfirmationPaneControlProps.defaultConfirmationPaneControlProps.cancelButtonAriaLabel
  })))));
}

var _default = ConfirmationPane;
exports.default = _default;
