// @microsoft/omnichannel-chat-components/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AgentIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.AgentIconBase64;
  }
});
Object.defineProperty(exports, "AudioNotificationOffIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.AudioNotificationOffIconBase64;
  }
});
Object.defineProperty(exports, "AudioNotificationOnIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.AudioNotificationOnIconBase64;
  }
});
Object.defineProperty(exports, "BroadcastService", {
  enumerable: true,
  get: function () {
    return _BroadcastService.BroadcastService;
  }
});
Object.defineProperty(exports, "BroadcastServiceInitialize", {
  enumerable: true,
  get: function () {
    return _BroadcastService.BroadcastServiceInitialize;
  }
});
Object.defineProperty(exports, "CallAcceptButtonBase64", {
  enumerable: true,
  get: function () {
    return _Icons.CallAcceptButtonBase64;
  }
});
Object.defineProperty(exports, "CallRejectButtonBase64", {
  enumerable: true,
  get: function () {
    return _Icons.CallRejectButtonBase64;
  }
});
Object.defineProperty(exports, "CallingContainer", {
  enumerable: true,
  get: function () {
    return _CallingContainer.default;
  }
});
Object.defineProperty(exports, "ChatButton", {
  enumerable: true,
  get: function () {
    return _ChatButton.default;
  }
});
Object.defineProperty(exports, "ChatReconnectIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.ChatReconnectIconBase64;
  }
});
Object.defineProperty(exports, "CloseChatButtonIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.CloseChatButtonIconBase64;
  }
});
Object.defineProperty(exports, "ConfirmationPane", {
  enumerable: true,
  get: function () {
    return _ConfirmationPane.default;
  }
});
Object.defineProperty(exports, "CurrentCall", {
  enumerable: true,
  get: function () {
    return _CurrentCall.default;
  }
});
Object.defineProperty(exports, "CustomChatIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.CustomChatIconBase64;
  }
});
Object.defineProperty(exports, "ElementType", {
  enumerable: true,
  get: function () {
    return _Constants.ElementType;
  }
});
Object.defineProperty(exports, "ErrorIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.ErrorIconBase64;
  }
});
Object.defineProperty(exports, "Footer", {
  enumerable: true,
  get: function () {
    return _Footer.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function () {
    return _Header.default;
  }
});
Object.defineProperty(exports, "IncomingCall", {
  enumerable: true,
  get: function () {
    return _IncomingCall.default;
  }
});
Object.defineProperty(exports, "InputValidationPane", {
  enumerable: true,
  get: function () {
    return _InputValidationPane.default;
  }
});
Object.defineProperty(exports, "LegacyChatIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.LegacyChatIconBase64;
  }
});
Object.defineProperty(exports, "LoadingPane", {
  enumerable: true,
  get: function () {
    return _LoadingPane.default;
  }
});
Object.defineProperty(exports, "LoadingSpinnerBase64", {
  enumerable: true,
  get: function () {
    return _Icons.LoadingSpinnerBase64;
  }
});
Object.defineProperty(exports, "MinimizeChatButtonIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.MinimizeChatButtonIconBase64;
  }
});
Object.defineProperty(exports, "ModernChatIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.ModernChatIconBase64;
  }
});
Object.defineProperty(exports, "OutOfOfficeHoursPane", {
  enumerable: true,
  get: function () {
    return _OOOHPane.default;
  }
});
Object.defineProperty(exports, "PostChatSurveyPane", {
  enumerable: true,
  get: function () {
    return _PostChatSurveyPane.default;
  }
});
Object.defineProperty(exports, "PreChatSurveyPane", {
  enumerable: true,
  get: function () {
    return _PreChatSurveyPane.default;
  }
});
Object.defineProperty(exports, "ProactiveChatBannerBase64", {
  enumerable: true,
  get: function () {
    return _Icons.ProactiveChatBannerBase64;
  }
});
Object.defineProperty(exports, "ProactiveChatPane", {
  enumerable: true,
  get: function () {
    return _ProactiveChatPane.default;
  }
});
Object.defineProperty(exports, "ReconnectChatPane", {
  enumerable: true,
  get: function () {
    return _ReconnectChatPane.default;
  }
});
Object.defineProperty(exports, "Timer", {
  enumerable: true,
  get: function () {
    return _Timer.default;
  }
});
Object.defineProperty(exports, "TranscriptDownloadIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.TranscriptDownloadIconBase64;
  }
});
Object.defineProperty(exports, "TranscriptEmailIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.TranscriptEmailIconBase64;
  }
});
Object.defineProperty(exports, "VideoCallAcceptButtonIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.VideoCallAcceptButtonIconBase64;
  }
});
Object.defineProperty(exports, "VideoOffIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.VideoOffIconBase64;
  }
});
Object.defineProperty(exports, "VideoOnIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.VideoOnIconBase64;
  }
});
Object.defineProperty(exports, "VoiceOffIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.VoiceOffIconBase64;
  }
});
Object.defineProperty(exports, "VoiceOnIconBase64", {
  enumerable: true,
  get: function () {
    return _Icons.VoiceOnIconBase64;
  }
});
Object.defineProperty(exports, "decodeComponentString", {
  enumerable: true,
  get: function () {
    return _decodeComponentString.decodeComponentString;
  }
});
Object.defineProperty(exports, "encodeComponentString", {
  enumerable: true,
  get: function () {
    return _encodeComponentString.encodeComponentString;
  }
});

var _Header = _interopRequireDefault(require("./components/header/Header"));

var _Footer = _interopRequireDefault(require("./components/footer/Footer"));

var _ConfirmationPane = _interopRequireDefault(require("./components/confirmationpane/ConfirmationPane"));

var _ChatButton = _interopRequireDefault(require("./components/chatbutton/ChatButton"));

var _InputValidationPane = _interopRequireDefault(require("./components/inputvalidationpane/InputValidationPane"));

var _ProactiveChatPane = _interopRequireDefault(require("./components/proactivechatpane/ProactiveChatPane"));

var _ReconnectChatPane = _interopRequireDefault(require("./components/reconnectchatpane/ReconnectChatPane"));

var _LoadingPane = _interopRequireDefault(require("./components/loadingpane/LoadingPane"));

var _OOOHPane = _interopRequireDefault(require("./components/outofofficehourspane/OOOHPane"));

var _PreChatSurveyPane = _interopRequireDefault(require("./components/prechatsurveypane/PreChatSurveyPane"));

var _PostChatSurveyPane = _interopRequireDefault(require("./components/postchatsurveypane/PostChatSurveyPane"));

var _encodeComponentString = require("./common/encodeComponentString");

var _decodeComponentString = require("./common/decodeComponentString");

var _BroadcastService = require("./services/BroadcastService");

var _Constants = require("./common/Constants");

var _CallingContainer = _interopRequireDefault(require("./components/callingcontainer/CallingContainer"));

var _CurrentCall = _interopRequireDefault(require("./components/callingcontainer/subcomponents/CurrentCall/CurrentCall"));

var _IncomingCall = _interopRequireDefault(require("./components/callingcontainer/subcomponents/IncomingCall/IncomingCall"));

var _Timer = _interopRequireDefault(require("./components/callingcontainer/subcomponents/Timer/Timer"));

var _Icons = require("./assets/Icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
