// botframework-webchat/lib/createDirectLineSpeechAdapters.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createDirectLineSpeechAdapters;

var _botframeworkDirectlinespeechSdk = require("botframework-directlinespeech-sdk");

var _createMicrophoneAudioConfigAndAudioContext = _interopRequireDefault(require("./speech/createMicrophoneAudioConfigAndAudioContext"));

var DEFAULT_LANGUAGE = 'en-US'; // TODO: When using DLSpeech via bundle, we will add our own MicrophoneAudioConfig.

function createDirectLineSpeechAdapters(_ref) {
  var _window, _window$navigator;

  var audioConfig = _ref.audioConfig,
      audioContext = _ref.audioContext,
      audioInputDeviceId = _ref.audioInputDeviceId,
      enableInternalHTTPSupport = _ref.enableInternalHTTPSupport,
      enableTelemetry = _ref.enableTelemetry,
      fetchCredentials = _ref.fetchCredentials,
      speechRecognitionEndpointId = _ref.speechRecognitionEndpointId,
      _ref$speechRecognitio = _ref.speechRecognitionLanguage,
      speechRecognitionLanguage = _ref$speechRecognitio === void 0 ? ((_window = window) === null || _window === void 0 ? void 0 : (_window$navigator = _window.navigator) === null || _window$navigator === void 0 ? void 0 : _window$navigator.language) || DEFAULT_LANGUAGE : _ref$speechRecognitio,
      speechSynthesisDeploymentId = _ref.speechSynthesisDeploymentId,
      speechSynthesisOutputFormat = _ref.speechSynthesisOutputFormat,
      textNormalization = _ref.textNormalization,
      userID = _ref.userID,
      username = _ref.username;

  if (audioConfig) {
    audioInputDeviceId && console.warn('botframework-webchat: "audioConfig" and "audioInputDeviceId" cannot be set at the same time; ignoring "audioInputDeviceId".');
    audioContext && console.warn('botframework-webchat: "audioConfig" and "audioContext" cannot be set at the same time; ignoring "audioContext" for speech recognition.');
  } else if (!window.navigator.mediaDevices) {
    // If the browser does not support or allow microphone access, we will continue to create Direct Line Speech adapter without custom "audioConfig" and "audioContext".
    // In Direct Line Speech SDK, it will disable speech functionality, only leaving text chat available via the protocol.
    console.warn('botframework-webchat: Your browser does not support or allow microphone access or the page is not loaded via HTTPS or localhost. Speech is disabled for Direct Line Speech. However, you may pass a custom "audioConfig" to enable speech in this environment.');
  } else {
    var _createMicrophoneAudi = (0, _createMicrophoneAudioConfigAndAudioContext.default)({
      audioContext: audioContext,
      audioInputDeviceId: audioInputDeviceId,
      enableTelemetry: enableTelemetry
    });

    audioConfig = _createMicrophoneAudi.audioConfig;
    audioContext = _createMicrophoneAudi.audioContext;
  }

  return (0, _botframeworkDirectlinespeechSdk.createAdapters)({
    audioConfig: audioConfig,
    audioContext: audioContext,
    enableInternalHTTPSupport: enableInternalHTTPSupport,
    enableTelemetry: enableTelemetry,
    fetchCredentials: fetchCredentials,
    speechRecognitionEndpointId: speechRecognitionEndpointId,
    speechRecognitionLanguage: speechRecognitionLanguage,
    speechSynthesisDeploymentId: speechSynthesisDeploymentId,
    speechSynthesisOutputFormat: speechSynthesisOutputFormat,
    textNormalization: textNormalization,
    userID: userID,
    username: username
  });
}
