// botframework-webchat-core/lib/reducers/createNotificationsReducer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createNotificationsReducer;

var _simpleUpdateIn = _interopRequireDefault(require("simple-update-in"));

var _dismissNotification = require("../actions/dismissNotification");

var _sagaError = require("../actions/sagaError");

var _setNotification = require("../actions/setNotification");

var _isForbiddenPropertyName = _interopRequireDefault(require("../utils/isForbiddenPropertyName"));

var DEFAULT_STATE = {};

function createNotificationsReducer(_ref) {
  var Date = _ref.Date;
  return function notifications() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : DEFAULT_STATE;
    var action = arguments.length > 1 ? arguments[1] : undefined;
    var type = action.type;
    var now = Date.now();

    if (type === _dismissNotification.DISMISS_NOTIFICATION) {
      state = (0, _simpleUpdateIn["default"])(state, [action.payload.id]);
    } else if (type === _sagaError.SAGA_ERROR) {
      state = (0, _simpleUpdateIn["default"])(state, ['connectivitystatus', 'message'], function () {
        return 'javascripterror';
      });
    } else if (type === _setNotification.SET_NOTIFICATION) {
      var _action$payload = action.payload,
          alt = _action$payload.alt,
          data = _action$payload.data,
          id = _action$payload.id,
          level = _action$payload.level,
          message = _action$payload.message;

      if (!(0, _isForbiddenPropertyName["default"])(id)) {
        // Mitigated through denylisting.
        // eslint-disable-next-line security/detect-object-injection
        var notification = state[id];

        if (!notification || alt !== notification.alt || !Object.is(data, notification.data) || level !== notification.level || message !== notification.message) {
          state = (0, _simpleUpdateIn["default"])(state, [id], function () {
            return {
              alt: alt,
              data: data,
              id: id,
              level: level,
              message: message,
              timestamp: now
            };
          });
        }
      }
    }

    return state;
  };
}
