// botframework-webchat-core/lib/reducers/createActivitiesReducer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createActivitiesReducer;

var _simpleUpdateIn = _interopRequireDefault(require("simple-update-in"));

var _deleteActivity = require("../actions/deleteActivity");

var _incomingActivity = require("../actions/incomingActivity");

var _markActivity = require("../actions/markActivity");

var _postActivity = require("../actions/postActivity");

var _SendStatus = require("../types/internal/SendStatus");

/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, -1] }] */
var DEFAULT_STATE = [];
var DIRECT_LINE_PLACEHOLDER_URL = 'https://docs.botframework.com/static/devportal/client/images/bot-framework-default-placeholder.png';

function getClientActivityID(activity) {
  var _activity$channelData;

  return (_activity$channelData = activity.channelData) === null || _activity$channelData === void 0 ? void 0 : _activity$channelData.clientActivityID;
}

function findByClientActivityID(clientActivityID) {
  return function (activity) {
    return getClientActivityID(activity) === clientActivityID;
  };
}

function patchActivity(activity, lastActivity, _ref) {
  var Date = _ref.Date;
  // Direct Line channel will return a placeholder image for the user-uploaded image.
  // As observed, the URL for the placeholder image is https://docs.botframework.com/static/devportal/client/images/bot-framework-default-placeholder.png.
  // To make our code simpler, we are removing the value if "contentUrl" is pointing to a placeholder image.
  // TODO: [P2] #2869 This "contentURL" removal code should be moved to DirectLineJS adapter.
  // Also, if the "contentURL" starts with "blob:", this means the user is uploading a file (the URL is constructed by URL.createObjectURL)
  // Although the copy/reference of the file is temporary in-memory, to make the UX consistent across page refresh, we do not allow the user to re-download the file either.
  activity = (0, _simpleUpdateIn["default"])(activity, ['attachments', function () {
    return true;
  }, 'contentUrl'], function (contentUrl) {
    if (contentUrl !== DIRECT_LINE_PLACEHOLDER_URL && !/^blob:/i.test(contentUrl)) {
      return contentUrl;
    }
  }); // If the message does not have sequence ID, use these fallback values:
  // 1. "timestamp" field
  //    - outgoing activity will not have "timestamp" field
  // 2. last activity sequence ID (or 0) + 0.001
  //    - best effort to put this message the last one in the chat history

  activity = (0, _simpleUpdateIn["default"])(activity, ['channelData', 'webchat:sequence-id'], function (sequenceId) {
    var _lastActivity$channel;

    return typeof sequenceId === 'number' ? sequenceId : typeof activity.timestamp !== 'undefined' ? +new Date(activity.timestamp) : // We assume there will be no more than 1,000 messages sent before receiving server response.
    // If there are more than 1,000 messages, some messages will get reordered and appear jumpy after receiving server response.
    // eslint-disable-next-line no-magic-numbers
    ((lastActivity === null || lastActivity === void 0 ? void 0 : (_lastActivity$channel = lastActivity.channelData) === null || _lastActivity$channel === void 0 ? void 0 : _lastActivity$channel['webchat:sequence-id']) || 0) + 0.001;
  }); // TODO: [P1] #3953 We should move this patching logic to a DLJS wrapper for simplicity.

  activity = (0, _simpleUpdateIn["default"])(activity, ['channelData', 'webchat:sequence-id'], function (sequenceId) {
    return typeof sequenceId === 'number' ? sequenceId : +new Date(activity.timestamp || 0) || 0;
  });
  return activity;
}

function upsertActivityWithSort(activities, nextActivity, ponyfill) {
  nextActivity = patchActivity(nextActivity, activities[activities.length - 1], ponyfill);
  var _nextActivity = nextActivity,
      _nextActivity$channel = _nextActivity.channelData;
  _nextActivity$channel = _nextActivity$channel === void 0 ? {} : _nextActivity$channel;
  var nextClientActivityID = _nextActivity$channel.clientActivityID,
      nextSequenceId = _nextActivity$channel['webchat:sequence-id'];
  var nextActivities = activities.filter(function (_ref2) {
    var _ref2$channelData = _ref2.channelData;
    _ref2$channelData = _ref2$channelData === void 0 ? {} : _ref2$channelData;
    var clientActivityID = _ref2$channelData.clientActivityID,
        id = _ref2.id;
    return (// We will remove all "sending messages" activities and activities with same ID
      // "clientActivityID" is unique and used to track if the message has been sent and echoed back from the server
      !(nextClientActivityID && clientActivityID === nextClientActivityID) && !(id && id === nextActivity.id)
    );
  }); // Then, find the right (sorted) place to insert the new activity at, based on sequence ID.

  var indexToInsert = nextActivities.findIndex(function (_ref3) {
    var _ref3$channelData = _ref3.channelData;
    _ref3$channelData = _ref3$channelData === void 0 ? {} : _ref3$channelData;
    var sequenceId = _ref3$channelData['webchat:sequence-id'];
    return (sequenceId || 0) > (nextSequenceId || 0);
  }); // If no right place are found, append it

  nextActivities.splice(~indexToInsert ? indexToInsert : nextActivities.length, 0, nextActivity);
  return nextActivities;
}

function createActivitiesReducer(ponyfill) {
  return function activities() {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : DEFAULT_STATE;
    var action = arguments.length > 1 ? arguments[1] : undefined;

    switch (action.type) {
      case _deleteActivity.DELETE_ACTIVITY:
        state = (0, _simpleUpdateIn["default"])(state, [function (_ref4) {
          var id = _ref4.id;
          return id === action.payload.activityID;
        }]);
        break;

      case _markActivity.MARK_ACTIVITY:
        {
          var payload = action.payload;
          state = (0, _simpleUpdateIn["default"])(state, [function (_ref5) {
            var id = _ref5.id;
            return id === payload.activityID;
          }, 'channelData', payload.name], function () {
            return payload.value;
          });
        }
        break;

      case _postActivity.POST_ACTIVITY_PENDING:
        {
          var _activity = action.payload.activity; // `channelData.state` is being deprecated in favor of `channelData['webchat:send-status']`.
          // Please refer to #4362 for details. Remove on or after 2024-07-31.

          _activity = (0, _simpleUpdateIn["default"])(_activity, ['channelData', 'state'], function () {
            return _SendStatus.SENDING;
          });
          _activity = (0, _simpleUpdateIn["default"])(_activity, ['channelData', 'webchat:send-status'], function () {
            return _SendStatus.SENDING;
          });
          state = upsertActivityWithSort(state, _activity, ponyfill);
        }
        break;

      case _postActivity.POST_ACTIVITY_IMPEDED:
        state = (0, _simpleUpdateIn["default"])(state, // `channelData.state` is being deprecated in favor of `channelData['webchat:send-status']`.
        // Please refer to #4362 for details. Remove on or after 2024-07-31.
        [findByClientActivityID(action.meta.clientActivityID), 'channelData', 'state'], function () {
          return _SendStatus.SEND_FAILED;
        });
        break;

      case _postActivity.POST_ACTIVITY_REJECTED:
        state = (0, _simpleUpdateIn["default"])(state, [findByClientActivityID(action.meta.clientActivityID)], function (activity) {
          activity = (0, _simpleUpdateIn["default"])(activity, ['channelData', 'state'], function () {
            return _SendStatus.SEND_FAILED;
          });
          return (0, _simpleUpdateIn["default"])(activity, ['channelData', 'webchat:send-status'], function () {
            return _SendStatus.SEND_FAILED;
          });
        });
        break;

      case _postActivity.POST_ACTIVITY_FULFILLED:
        state = (0, _simpleUpdateIn["default"])(state, [findByClientActivityID(action.meta.clientActivityID)], function () {
          // We will replace the activity with the version from the server
          var activity = (0, _simpleUpdateIn["default"])(patchActivity(action.payload.activity, state[state.length - 1], ponyfill), // `channelData.state` is being deprecated in favor of `channelData['webchat:send-status']`.
          // Please refer to #4362 for details. Remove on or after 2024-07-31.
          ['channelData', 'state'], function () {
            return _SendStatus.SENT;
          });
          return (0, _simpleUpdateIn["default"])(activity, ['channelData', 'webchat:send-status'], function () {
            return _SendStatus.SENT;
          });
        });
        break;

      case _incomingActivity.INCOMING_ACTIVITY:
        {
          var _activity2 = action.payload.activity; // If the incoming activity is an echo back, we should keep the existing `channelData['webchat:send-status']` field.
          //
          // Otherwise, it will fail following scenario:
          //
          // 1. Send an activity to the service
          // 2. Service echoed back the activity
          // 3. Service did NOT return `postActivity` call
          // -  EXPECT: `channelData['webchat:send-status']` should be "sending".
          // -  ACTUAL: `channelData['webchat:send-status']` is `undefined` because the activity get overwritten by the echo back activity.
          //            The echo back activity contains no `channelData['webchat:send-status']`.
          //
          // While we are looking out for the scenario above, we should also look at the following scenarios:
          //
          // 1. Service restore chat history, including activities sent from the user. These activities has the following characteristics:
          //    - They do not have `channelData['webchat:send-status']`;
          //    - They do not have an ongoing `postActivitySaga`;
          //    - They should not previously appear in the chat history.
          // 2. We need to mark these activities as "sent".
          //
          // In the future, when we revamp our object model, we could use a different signal so we don't need the code below, for example:
          //
          // -  If `activity.id` is set, it is "sent", because the chat service assigned an ID to the activity;
          // -  If `activity.id` is not set, it is either "sending" or "send failed";
          //    - If `activity.channelData['webchat:send-failed-reason']` is set, it is "send failed" with the reason, otherwise;
          //    - It is sending.

          if (_activity2.from.role === 'user') {
            var _activity3 = _activity2,
                id = _activity3.id;
            var clientActivityID = getClientActivityID(_activity2);
            var existingActivity = state.find(function (activity) {
              return clientActivityID && getClientActivityID(activity) === clientActivityID || id && activity.id === id;
            });

            if (existingActivity) {
              var sendStatus = existingActivity.channelData['webchat:send-status'];

              if (sendStatus === _SendStatus.SENDING || sendStatus === _SendStatus.SEND_FAILED || sendStatus === _SendStatus.SENT) {
                _activity2 = (0, _simpleUpdateIn["default"])(_activity2, ['channelData', 'webchat:send-status'], function () {
                  return sendStatus;
                });
              }
            } else {
              // If there are no existing activity, probably this activity is restored from chat history.
              // All outgoing activities restored from service means they arrived at the service successfully.
              // Thus, we are marking them as "sent".
              _activity2 = (0, _simpleUpdateIn["default"])(_activity2, ['channelData', 'webchat:send-status'], function () {
                return _SendStatus.SENT;
              });
            }
          }

          state = upsertActivityWithSort(state, _activity2, ponyfill);
        }
        break;

      default:
        break;
    }

    return state;
  };
}
