// botframework-webchat/lib/adaptiveCards/Attachment/AudioCardContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _botframeworkWebchatComponent = require("botframework-webchat-component");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _CommonCard = _interopRequireDefault(require("./CommonCard"));

/* eslint react/no-array-index-key: "off" */
var AudioContent = _botframeworkWebchatComponent.Components.AudioContent;
var useStyleSet = _botframeworkWebchatComponent.hooks.useStyleSet;

var AudioCardContent = function AudioCardContent(_ref) {
  var actionPerformedClassName = _ref.actionPerformedClassName,
      content = _ref.content,
      disabled = _ref.disabled;

  var _useStyleSet = useStyleSet(),
      _useStyleSet2 = (0, _slicedToArray2.default)(_useStyleSet, 1),
      audioCardAttachmentStyleSet = _useStyleSet2[0].audioCardAttachment;

  var _content$autostart = content.autostart,
      autostart = _content$autostart === void 0 ? false : _content$autostart,
      _content$autoloop = content.autoloop,
      autoloop = _content$autoloop === void 0 ? false : _content$autoloop,
      _content$image = content.image;
  _content$image = _content$image === void 0 ? {} : _content$image;
  var _content$image$url = _content$image.url,
      imageURL = _content$image$url === void 0 ? '' : _content$image$url,
      _content$media = content.media,
      media = _content$media === void 0 ? [] : _content$media;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: audioCardAttachmentStyleSet
  }, /*#__PURE__*/_react.default.createElement("ul", {
    className: "media-list"
  }, media.map(function (_ref2, index) {
    var url = _ref2.url;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: index
    }, /*#__PURE__*/_react.default.createElement(AudioContent, {
      autoPlay: autostart,
      loop: autoloop,
      poster: imageURL,
      src: url
    }));
  })), /*#__PURE__*/_react.default.createElement(_CommonCard.default, {
    actionPerformedClassName: actionPerformedClassName,
    content: content,
    disabled: disabled
  }));
};

AudioCardContent.defaultProps = {
  actionPerformedClassName: '',
  disabled: undefined
};
AudioCardContent.propTypes = {
  actionPerformedClassName: _propTypes.default.string,
  // PropTypes cannot fully capture TypeScript types.
  // @ts-ignore
  content: _propTypes.default.shape({
    autostart: _propTypes.default.bool,
    autoloop: _propTypes.default.bool,
    image: _propTypes.default.shape({
      url: _propTypes.default.string.isRequired
    }),
    media: _propTypes.default.arrayOf(_propTypes.default.shape({
      url: _propTypes.default.string.isRequired
    }).isRequired).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = AudioCardContent;
exports.default = _default;
