// botframework-webchat/lib/adaptiveCards/Attachment/ThumbnailCardAttachment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ThumbnailCardContent = _interopRequireDefault(require("./ThumbnailCardContent"));

var ThumbnailCardAttachment = function ThumbnailCardAttachment(_ref) {
  var content = _ref.attachment.content,
      disabled = _ref.disabled;
  return /*#__PURE__*/_react.default.createElement(_ThumbnailCardContent.default, {
    content: content,
    disabled: disabled
  });
};

ThumbnailCardAttachment.defaultProps = {
  disabled: undefined
};
ThumbnailCardAttachment.propTypes = {
  attachment: _propTypes.default.shape({
    content: _propTypes.default.shape({
      buttons: _propTypes.default.array,
      images: _propTypes.default.arrayOf(_propTypes.default.shape({
        alt: _propTypes.default.string.isRequired,
        tap: _propTypes.default.any,
        url: _propTypes.default.string.isRequired
      })),
      subtitle: _propTypes.default.string,
      tap: _propTypes.default.any,
      text: _propTypes.default.string,
      title: _propTypes.default.string
    }).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = ThumbnailCardAttachment;
exports.default = _default;
