// @microsoft/omnichannel-chat-components/lib/cjs/components/loadingpane/common/presetThreeProps/presetThreeStyles/presetThreeLoadingPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetThreeLoadingPaneStyles = void 0;

var _presetThreeLoadingPaneGeneralStyles = require("./presetThreeLoadingPaneGeneralStyles");

var _presetThreeLoadingPaneIconStyles = require("./presetThreeLoadingPaneIconStyles");

var _presetThreeLoadingPaneIconImageProps = require("./presetThreeLoadingPaneIconImageProps");

var _presetThreeLoadingPaneSubtitleStyles = require("./presetThreeLoadingPaneSubtitleStyles");

var _presetThreeLoadingPaneTitleStyles = require("./presetThreeLoadingPaneTitleStyles");

var _presetThreeLoadingPaneSpinnerStyles = require("./presetThreeLoadingPaneSpinnerStyles");

var _presetThreeLoadingPaneSpinnerTextStyles = require("./presetThreeLoadingPaneSpinnerTextStyles");

const presetThreeLoadingPaneStyles = {
  generalStyleProps: _presetThreeLoadingPaneGeneralStyles.presetThreeLoadingPaneGeneralStyles,
  titleStyleProps: _presetThreeLoadingPaneTitleStyles.presetThreeLoadingPaneTitleStyles,
  subtitleStyleProps: _presetThreeLoadingPaneSubtitleStyles.presetThreeLoadingPaneSubtitleStyles,
  iconStyleProps: _presetThreeLoadingPaneIconStyles.presetThreeLoadingPaneIconStyles,
  iconImageProps: _presetThreeLoadingPaneIconImageProps.presetThreeLoadingPaneIconImageProps,
  spinnerStyleProps: _presetThreeLoadingPaneSpinnerStyles.presetThreeLoadingPaneSpinnerStyles,
  spinnerTextStyleProps: _presetThreeLoadingPaneSpinnerTextStyles.presetThreeLoadingPaneSpinnerTextStyles
};
exports.presetThreeLoadingPaneStyles = presetThreeLoadingPaneStyles;
