// @microsoft/omnichannel-ic3core/lib/model/ConversationGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConversationGenerator = /** @class */ (function () {
    function ConversationGenerator() {
        this.messagesObservers = [];
        this.fileMessagesObservers = [];
    }
    ConversationGenerator.prototype.generateMessage = function (conversationId, message) {
        var conversation = {
            id: conversationId
        };
        this.messagesObservers.forEach(function (messagesObserver) {
            messagesObserver(conversation, message);
        });
    };
    ConversationGenerator.prototype.generateFileMessage = function (conversationId, message, file) {
        var conversation = {
            id: conversationId
        };
        this.fileMessagesObservers.forEach(function (fileMessagesObserver) {
            fileMessagesObserver(conversation, message, file);
        });
    };
    ConversationGenerator.prototype.subscribeToMessages = function (subscriber) {
        this.messagesObservers.push(subscriber);
    };
    ConversationGenerator.prototype.subscribeToFileMessages = function (subscriber) {
        this.fileMessagesObservers.push(subscriber);
    };
    return ConversationGenerator;
}());
exports.default = ConversationGenerator;
