// @microsoft/omnichannel-chat-components/lib/cjs/components/footer/subcomponents/DownloadTranscriptButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Constants = require("../../../common/Constants");

var _CommandButton = _interopRequireDefault(require("../../common/commandbutton/CommandButton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DownloadTranscriptButton(props) {
  const {
    type
  } = props;
  const customEvent = {
    elementType: _Constants.ElementType.FooterDownloadTranscriptButton,
    elementId: props === null || props === void 0 ? void 0 : props.id,
    eventName: _Constants.EventNames.OnClick
  };
  return /*#__PURE__*/_react.default.createElement(_CommandButton.default, {
    id: props === null || props === void 0 ? void 0 : props.id,
    type: type,
    styles: props.styles,
    hoverStyles: props.hoverStyles,
    focusStyles: props.focusStyles,
    iconName: props.iconName ?? _Constants.IconNames.Download,
    imageIconProps: props.imageIconProps,
    onClick: props.onClick,
    ariaLabel: props.ariaLabel ?? _Constants.AriaLabels.DownloadChatTranscript,
    className: props.className,
    customEvent: customEvent,
    hideButtonTitle: props.hideButtonTitle
  });
}

var _default = DownloadTranscriptButton;
exports.default = _default;
