// @microsoft/omnichannel-chat-components/lib/cjs/components/inputvalidationpane/common/presetTwo/presetTwoProps/presetTwoInputValidationPaneControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetTwoInputValidationPaneControlProps = void 0;

var _Constants = require("../../../../../common/Constants");

const presetTwoInputValidationPaneControlProps = {
  id: _Constants.Ids.DefaultInputValidationPaneId,
  dir: "ltr",
  hideInputValidationPane: false,
  inputValidationPaneAriaLabel: _Constants.AriaLabels.EmailChatTranscriptPane,
  hideTitle: false,
  titleText: _Constants.Texts.InputValidationPaneTitleText,
  hideSubtitle: false,
  subtitleText: _Constants.Texts.InputValidationPaneSubtitleText,
  inputId: _Constants.Ids.DefaultInputValidationPaneInputId,
  hideInput: false,
  inputAriaLabel: _Constants.AriaLabels.InputValidationPaneInput,
  inputWithErrorMessageBorderColor: "rgb(164, 38, 44)",
  inputPlaceHolder: _Constants.Texts.EmailPlaceHolderText,
  invalidInputErrorMessageText: _Constants.Texts.InvalidInputErrorMessageText,
  isButtonGroupHorizontal: false,
  hideSendButton: false,
  enableSendButton: true,
  sendButtonText: _Constants.Texts.SendButtonText,
  sendButtonAriaLabel: _Constants.AriaLabels.Send,
  hideCancelButton: false,
  cancelButtonText: _Constants.Texts.CancelButtonText,
  cancelButtonAriaLabel: _Constants.AriaLabels.Cancel,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onSend: function (input) {
    console.log("on send");
  },
  onCancel: function () {
    console.log("on cancel");
  },
  checkInput: function (input) {
    return new RegExp(_Constants.Regex.EmailRegex).test(input);
  }
};
exports.presetTwoInputValidationPaneControlProps = presetTwoInputValidationPaneControlProps;
