// @microsoft/omnichannel-chat-components/lib/esm/components/callingcontainer/subcomponents/IncomingCall/common/defaultProps/componentOverrideProps.js
import * as React from "react";
import { Image, Label, Stack } from "@fluentui/react";
import { AgentIconBase64 } from "../../../../../../assets/Icons";
const agentImageStyles = {
  root: {
    height: "50px",
    width: "50px"
  },
  image: {
    height: "50px",
    width: "50px"
  }
};
const label1Styles = {
  root: {
    color: "white",
    fontSize: 13,
    fontWeight: "bold",
    padding: 0
  }
};
const label2Styles = {
  root: {
    color: "white",
    fontSize: 11,
    padding: 0
  }
};
const stackStyles = {
  root: {
    paddingLeft: "10px"
  }
};
export const componentOverrideProps = {
  incomingCallTitle: /*#__PURE__*/React.createElement(Stack, {
    horizontal: true
  }, /*#__PURE__*/React.createElement(Image, {
    src: AgentIconBase64,
    styles: agentImageStyles
  }), /*#__PURE__*/React.createElement(Stack, {
    verticalAlign: "center",
    horizontalAlign: "baseline",
    styles: stackStyles
  }, /*#__PURE__*/React.createElement(Label, {
    styles: label1Styles
  }, "Omnichannel Agent 007"), /*#__PURE__*/React.createElement(Label, {
    styles: label2Styles
  }, "Incoming call")))
};
