// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/presetTwo/presetTwoStyles/presetTwoProactiveChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetTwoProactiveChatPaneStyles = void 0;

var _presetTwoProactiveChatPaneBodyContainerStyles = require("./presetTwoProactiveChatPaneBodyContainerStyles");

var _presetTwoProactiveChatPaneGeneralStyles = require("./presetTwoProactiveChatPaneGeneralStyles");

var _presetTwoProactiveChatPaneHeaderContainerStyles = require("./presetTwoProactiveChatPaneHeaderContainerStyles");

var _presetTwoProactiveChatPaneTextContainerStyles = require("./presetTwoProactiveChatPaneTextContainerStyles");

var _presetTwoProactiveChatPaneStartButtonStyles = require("./presetTwoProactiveChatPaneStartButtonStyles");

var _presetTwoProactiveChatPaneStartButtonHoveredStyles = require("./presetTwoProactiveChatPaneStartButtonHoveredStyles");

const presetTwoProactiveChatPaneStyles = {
  generalStyleProps: _presetTwoProactiveChatPaneGeneralStyles.presetTwoProactiveChatPaneGeneralStyles,
  headerContainerStyleProps: _presetTwoProactiveChatPaneHeaderContainerStyles.presetTwoProactiveChatPaneHeaderContainerStyles,
  textContainerStyleProps: _presetTwoProactiveChatPaneTextContainerStyles.presetTwoProactiveChatPaneTextContainerStyles,
  bodyContainerStyleProps: _presetTwoProactiveChatPaneBodyContainerStyles.presetTwoProactiveChatPaneBodyContainerStyles,
  startButtonStyleProps: _presetTwoProactiveChatPaneStartButtonStyles.presetTwoProactiveChatPaneStartButtonStyles,
  startButtonHoveredStyleProps: _presetTwoProactiveChatPaneStartButtonHoveredStyles.presetTwoProactiveChatPaneStartButtonHoveredStyles
};
exports.presetTwoProactiveChatPaneStyles = presetTwoProactiveChatPaneStyles;
