// botframework-webchat-component/lib/Activity/CarouselLayout.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _botframeworkWebchatApi = require("botframework-webchat-api");

var _reactFilm = require("react-film");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _CarouselFilmStrip = _interopRequireDefault(require("./CarouselFilmStrip"));

var _useNonce3 = _interopRequireDefault(require("../hooks/internal/useNonce"));

var _useStyleSet3 = _interopRequireDefault(require("../hooks/useStyleSet"));

var _useStyleToEmotionObject = _interopRequireDefault(require("../hooks/internal/useStyleToEmotionObject"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useDirection = _botframeworkWebchatApi.hooks.useDirection,
    useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer;
var ROOT_STYLE = {
  '&.webchat__carousel-layout': {
    overflow: 'hidden',
    position: 'relative'
  }
};

var CarouselLayoutCore = function CarouselLayoutCore(_ref) {
  var activity = _ref.activity,
      hideTimestamp = _ref.hideTimestamp,
      renderActivityStatus = _ref.renderActivityStatus,
      renderAttachment = _ref.renderAttachment,
      renderAvatar = _ref.renderAvatar,
      showCallout = _ref.showCallout;

  var _useStyleSet = (0, _useStyleSet3.default)(),
      _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
      carouselFlipperStyleSet = _useStyleSet2[0].carouselFlipper;

  var _useReactFilmStyleSet = (0, _reactFilm.useStyleSetClassNames)(),
      _useReactFilmStyleSet2 = _slicedToArray(_useReactFilmStyleSet, 1),
      filmRootClassName = _useReactFilmStyleSet2[0].root;

  var _useDirection = useDirection(),
      _useDirection2 = _slicedToArray(_useDirection, 1),
      direction = _useDirection2[0];

  var _useScrollBarWidth = (0, _reactFilm.useScrollBarWidth)(),
      _useScrollBarWidth2 = _slicedToArray(_useScrollBarWidth, 1),
      scrollBarWidth = _useScrollBarWidth2[0];

  var _useScrolling = (0, _reactFilm.useScrolling)(),
      _useScrolling2 = _slicedToArray(_useScrolling, 1),
      scrolling = _useScrolling2[0];

  var leftSideFlipper = direction === 'rtl' ? '>' : '<';
  var localize = useLocalizer();
  var rightSideFlipper = direction === 'rtl' ? '<' : '>';
  var rootClassName = (0, _useStyleToEmotionObject.default)()(ROOT_STYLE) + '';
  var nextAlt = localize('CAROUSEL_FLIPPER_NEXT_ALT');
  var previousAlt = localize('CAROUSEL_FLIPPER_PREVIOUS_ALT');
  var leftFlipperAriaLabel = direction === 'rtl' ? nextAlt : previousAlt;
  var rightFlipperAriaLabel = direction === 'rtl' ? previousAlt : nextAlt;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('webchat__carousel-layout', rootClassName, carouselFlipperStyleSet + '', filmRootClassName)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('react-film__main', {
      'react-film__main--scrolling': scrolling
    })
  }, /*#__PURE__*/_react.default.createElement(_CarouselFilmStrip.default, {
    activity: activity,
    hideTimestamp: hideTimestamp,
    renderActivityStatus: renderActivityStatus,
    renderAttachment: renderAttachment,
    renderAvatar: renderAvatar,
    showCallout: showCallout
  }), scrollBarWidth !== '100%' && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactFilm.Flipper, {
    "aria-label": leftFlipperAriaLabel,
    blurFocusOnClick: true,
    mode: "left"
  }, leftSideFlipper), /*#__PURE__*/_react.default.createElement(_reactFilm.Flipper, {
    "aria-label": rightFlipperAriaLabel,
    blurFocusOnClick: true,
    mode: "right"
  }, rightSideFlipper))));
};

CarouselLayoutCore.defaultProps = {
  hideTimestamp: false,
  renderActivityStatus: false,
  renderAvatar: false,
  showCallout: true
};
CarouselLayoutCore.propTypes = {
  activity: _propTypes.default.shape({
    attachments: _propTypes.default.array
  }).isRequired,
  hideTimestamp: _propTypes.default.bool,
  renderActivityStatus: _propTypes.default.oneOfType([_propTypes.default.oneOf([false]), _propTypes.default.func]),
  renderAttachment: _propTypes.default.func.isRequired,
  renderAvatar: _propTypes.default.oneOfType([_propTypes.default.oneOf([false]), _propTypes.default.func]),
  showCallout: _propTypes.default.bool
};

var CarouselLayout = function CarouselLayout(props) {
  var _props$activity = props.activity;
  _props$activity = _props$activity === void 0 ? {} : _props$activity;
  var _props$activity$attac = _props$activity.attachments,
      attachments = _props$activity$attac === void 0 ? [] : _props$activity$attac;

  var _useDirection3 = useDirection(),
      _useDirection4 = _slicedToArray(_useDirection3, 1),
      direction = _useDirection4[0];

  var _useNonce = (0, _useNonce3.default)(),
      _useNonce2 = _slicedToArray(_useNonce, 1),
      nonce = _useNonce2[0];

  var filmStyleSet = (0, _react.useMemo)(function () {
    return (0, _reactFilm.createBasicStyleSet)({
      cursor: null
    });
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactFilm.Composer, {
    dir: direction,
    nonce: nonce,
    numItems: attachments.length,
    styleSet: filmStyleSet
  }, /*#__PURE__*/_react.default.createElement(CarouselLayoutCore, props));
};

CarouselLayout.defaultProps = _objectSpread({}, CarouselLayoutCore.defaultProps);
CarouselLayout.propTypes = _objectSpread({}, CarouselLayoutCore.propTypes);
var _default = CarouselLayout;
exports.default = _default;
