// botframework-webchat-component/lib/Activity/CarouselFilmStripAttachment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _botframeworkWebchatApi = require("botframework-webchat-api");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Bubble = _interopRequireDefault(require("./Bubble"));

var _ScreenReaderText = _interopRequireDefault(require("../ScreenReaderText"));

var _useStyleSet3 = _interopRequireDefault(require("../hooks/useStyleSet"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useDirection = _botframeworkWebchatApi.hooks.useDirection,
    useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer;

var CarouselFilmStripAttachment = function CarouselFilmStripAttachment(_ref) {
  var activity = _ref.activity,
      attachment = _ref.attachment,
      className = _ref.className,
      fromUser = _ref.fromUser,
      hasAvatar = _ref.hasAvatar,
      hideNub = _ref.hideNub,
      index = _ref.index,
      renderAttachment = _ref.renderAttachment,
      showAvatar = _ref.showAvatar,
      showNub = _ref.showNub;

  var _useDirection = useDirection(),
      _useDirection2 = _slicedToArray(_useDirection, 1),
      direction = _useDirection2[0];

  var localize = useLocalizer();

  var _useStyleSet = (0, _useStyleSet3.default)(),
      _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
      carouselFilmStripAttachmentStyleSet = _useStyleSet2[0].carouselFilmStripAttachment;

  var attachedAlt = localize(fromUser ? 'ACTIVITY_YOU_ATTACHED_ALT' : 'ACTIVITY_BOT_ATTACHED_ALT');
  return /*#__PURE__*/_react.default.createElement("li", {
    "aria-roledescription": "attachment",
    className: (0, _classnames.default)('webchat__carousel-filmstrip-attachment', {
      'webchat__carousel-filmstrip-attachment--hide-avatar': hasAvatar && !showAvatar,
      'webchat__carousel-filmstrip-attachment--hide-nub': hideNub,
      'webchat__carousel-filmstrip-attachment--rtl': direction === 'rtl',
      'webchat__carousel-filmstrip-attachment--show-avatar': showAvatar,
      'webchat__carousel-filmstrip-attachment--show-nub': showNub
    }, 'react-film__filmstrip__item', carouselFilmStripAttachmentStyleSet + '', (className || '') + ''),
    role: "listitem",
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_ScreenReaderText.default, {
    text: attachedAlt
  }), /*#__PURE__*/_react.default.createElement(_Bubble.default, {
    fromUser: fromUser,
    key: index,
    nub: false
  }, renderAttachment({
    activity: activity,
    attachment: attachment
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "webchat__carousel-filmstrip-attachment--focus"
  })));
};

CarouselFilmStripAttachment.defaultProps = {
  className: ''
};
CarouselFilmStripAttachment.propTypes = {
  activity: _propTypes.default.shape({
    attachments: _propTypes.default.array,
    from: _propTypes.default.shape({
      role: _propTypes.default.string.isRequired
    }).isRequired,
    text: _propTypes.default.string,
    textFormat: _propTypes.default.string,
    timestamp: _propTypes.default.string
  }).isRequired,
  attachment: _propTypes.default.shape({
    content: _propTypes.default.any
  }).isRequired,
  className: _propTypes.default.string,
  fromUser: _propTypes.default.any.isRequired,
  hasAvatar: _propTypes.default.any.isRequired,
  hideNub: _propTypes.default.bool.isRequired,
  index: _propTypes.default.number.isRequired,
  renderAttachment: _propTypes.default.func.isRequired,
  showAvatar: _propTypes.default.bool.isRequired,
  showNub: _propTypes.default.bool.isRequired
};
var _default = CarouselFilmStripAttachment;
exports.default = _default;
