// react-film/lib/esm/BasicFilm.js
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { useMemo } from 'react';
import Dots from './Dots';
import Filmstrip from './Filmstrip';
import Flipper from './Flipper';
import ScrollBar from './ScrollBar';
import useDir from './hooks/useDir';
import useHeight from './hooks/useHeight';
import useNumItems from './hooks/useNumItems';
import useScrollBarWidth from './hooks/useScrollBarWidth';
import useScrolling from './hooks/useScrolling';
import useStyleOptions from './hooks/useStyleOptions';
import useStyleSetClassNames from './hooks/useStyleSetClassNames';
var BasicFilm = function BasicFilm(_ref) {
  var children = _ref.children,
    className = _ref.className;
  var _useDir = useDir(),
    _useDir2 = _slicedToArray(_useDir, 1),
    dir = _useDir2[0];
  var _useHeight = useHeight(),
    _useHeight2 = _slicedToArray(_useHeight, 1),
    height = _useHeight2[0];
  var _useNumItems = useNumItems(),
    _useNumItems2 = _slicedToArray(_useNumItems, 1),
    numItems = _useNumItems2[0];
  var _useScrollBarWidth = useScrollBarWidth(),
    _useScrollBarWidth2 = _slicedToArray(_useScrollBarWidth, 1),
    scrollBarWidth = _useScrollBarWidth2[0];
  var _useScrolling = useScrolling(),
    _useScrolling2 = _slicedToArray(_useScrolling, 1),
    scrolling = _useScrolling2[0];
  var _useStyleSetClassName = useStyleSetClassNames(),
    _useStyleSetClassName2 = _slicedToArray(_useStyleSetClassName, 1),
    rootClassName = _useStyleSetClassName2[0].root;
  var _useStyleOptions = useStyleOptions(),
    _useStyleOptions2 = _slicedToArray(_useStyleOptions, 1),
    _useStyleOptions2$ = _useStyleOptions2[0],
    flipperBlurFocusOnClick = _useStyleOptions2$.flipperBlurFocusOnClick,
    leftFlipperAriaLabel = _useStyleOptions2$.leftFlipperAriaLabel,
    leftFlipperText = _useStyleOptions2$.leftFlipperText,
    rightFlipperAriaLabel = _useStyleOptions2$.rightFlipperAriaLabel,
    rightFlipperText = _useStyleOptions2$.rightFlipperText,
    showDots = _useStyleOptions2$.showDots,
    showFlipper = _useStyleOptions2$.showFlipper,
    showScrollBar = _useStyleOptions2$.showScrollBar;
  var contentStyle = useMemo(function () {
    return {
      height: height
    };
  }, [height]);
  return /*#__PURE__*/React.createElement("div", {
    className: classNames(rootClassName, (className || '') + ''),
    dir: dir
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('react-film__main', {
      'react-film__main--scrolling': scrolling
    }),
    style: contentStyle
  }, !!numItems && scrollBarWidth !== '100%' && !!showFlipper && /*#__PURE__*/React.createElement(Flipper, {
    "aria-label": leftFlipperAriaLabel,
    blurFocusOnClick: flipperBlurFocusOnClick,
    mode: "left"
  }, leftFlipperText), /*#__PURE__*/React.createElement(Filmstrip, null, children), !!numItems && scrollBarWidth !== '100%' && !!showFlipper && /*#__PURE__*/React.createElement(Flipper, {
    "aria-label": rightFlipperAriaLabel,
    blurFocusOnClick: flipperBlurFocusOnClick,
    mode: "right"
  }, rightFlipperText), !!numItems && scrollBarWidth !== '100%' && !!showScrollBar && /*#__PURE__*/React.createElement(ScrollBar, null)), !!numItems && scrollBarWidth !== '100%' && !!showDots && /*#__PURE__*/React.createElement(Dots, null));
};

// TODO: Move from styleSet to styleSheet.

BasicFilm.defaultProps = {
  children: undefined,
  className: undefined
};
BasicFilm.propTypes = {
  children: PropTypes.any,
  className: PropTypes.string
};
export default BasicFilm;
