// react-film/lib/esm/Dot.js
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import useIndex from './hooks/useIndex';
import useScrollTo from './hooks/useScrollTo';
var Dot = function Dot(_ref) {
  var ariaLabel = _ref['aria-label'],
    itemIndex = _ref.itemIndex;
  var _useIndex = useIndex(),
    _useIndex2 = _slicedToArray(_useIndex, 1),
    index = _useIndex2[0];
  var scrollTo = useScrollTo();
  var checked = index === itemIndex;
  var handleChange = useCallback(function () {
    return scrollTo(function () {
      return itemIndex;
    });
  }, [itemIndex, scrollTo]);
  var handleKeyPress = useCallback(function (event) {
    var keyCode = event.keyCode;

    // 13 is ENTER, 32 is SPACEBAR.
    // eslint-disable-next-line no-magic-numbers
    if (keyCode === 13 || keyCode === 32) {
      event.preventDefault();
      event.stopPropagation();
      handleChange();
    }
  }, [handleChange]);
  return /*#__PURE__*/React.createElement("li", {
    className: "react-film__dot"
  }, /*#__PURE__*/React.createElement("input", {
    "aria-label": ariaLabel,
    "aria-pressed": checked,
    checked: checked,
    className: "react-film__dot__input",
    onChange: handleChange,
    onKeyPress: handleKeyPress,
    role: "button",
    type: "checkbox"
  }), /*#__PURE__*/React.createElement("div", {
    className: "react-film__dot__handle"
  }));
};
Dot.defaultProps = {
  'aria-label': undefined
};
Dot.propTypes = {
  'aria-label': PropTypes.string,
  itemIndex: PropTypes.number.isRequired
};
export default Dot;
