// react-film/lib/esm/Flipper.js
import _slicedToArray from "@babel/runtime-corejs3/helpers/slicedToArray";
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { useCallback, useRef } from 'react';
import useDir from './hooks/useDir';
import useScrollBarPercentage from './hooks/useScrollBarPercentage';
import useScrollOneLeft from './hooks/useScrollOneLeft';
import useScrollOneRight from './hooks/useScrollOneRight';
var Flipper = function Flipper(_ref) {
  var ariaLabel = _ref['aria-label'],
    blurFocusOnClick = _ref.blurFocusOnClick,
    children = _ref.children,
    mode = _ref.mode;
  var _useDir = useDir(),
    _useDir2 = _slicedToArray(_useDir, 1),
    dir = _useDir2[0];
  var _useScrollBarPercenta = useScrollBarPercentage(),
    _useScrollBarPercenta2 = _slicedToArray(_useScrollBarPercenta, 1),
    scrollBarPercentage = _useScrollBarPercenta2[0];
  var buttonRef = useRef();
  var left = mode === 'left';
  var scrollOneLeft = useScrollOneLeft();
  var scrollOneRight = useScrollOneRight();
  var handleClick = useCallback(function () {
    left ? scrollOneLeft() : scrollOneRight();
    blurFocusOnClick && buttonRef.current && buttonRef.current.blur();
  }, [blurFocusOnClick, buttonRef, left, scrollOneLeft, scrollOneRight]);
  var handleKeyDown = useCallback(function (event) {
    var key = event.key;
    if (key === 'Enter' || key === ' ') {
      event.preventDefault();
      left ? scrollOneLeft() : scrollOneRight();
    }
  }, [left, scrollOneLeft, scrollOneRight]);
  var hide;
  if (dir === 'rtl') {
    if (left) {
      hide = scrollBarPercentage === '100%' || scrollBarPercentage === '-100%';
    } else {
      hide = scrollBarPercentage === '0%';
    }
  } else {
    if (left) {
      hide = scrollBarPercentage === '0%';
    } else {
      hide = scrollBarPercentage === '100%';
    }
  }
  return /*#__PURE__*/React.createElement("button", {
    "aria-label": ariaLabel || (left ? 'left' : 'right'),
    className: classNames('react-film__flipper', 'react-film__main__overlay', {
      'react-film__flipper--left': left,
      'react-film__flipper--right': !left
    }),
    onClick: handleClick,
    onKeyDown: handleKeyDown,
    ref: buttonRef,
    type: "button"
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames('react-film__flipper__slider', 'react-film__main__slider', {
      'react-film__main__slider--hide': hide,
      'react-film__main__slider--left': left,
      'react-film__main__slider--right': !left
    })
  }, /*#__PURE__*/React.createElement("div", {
    className: "react-film__flipper__body"
  }, children)));
};
Flipper.defaultProps = {
  'aria-label': undefined,
  blurFocusOnClick: false,
  children: undefined,
  mode: 'left'
};
Flipper.propTypes = {
  'aria-label': PropTypes.string,
  blurFocusOnClick: PropTypes.bool,
  children: PropTypes.any,
  mode: PropTypes.oneOf(['left', 'right'])
};
export default Flipper;
