// react-film/lib/esm/ReactFilm.js
import _objectSpread from "@babel/runtime-corejs3/helpers/objectSpread2";
import PropTypes from 'prop-types';
import React, { Children, useMemo } from 'react';
import BasicFilm from './BasicFilm';
import Composer from './Composer';
var ReactFilm = function ReactFilm(_ref) {
  var autoCenter = _ref.autoCenter,
    autoHide = _ref.autoHide,
    autoHideFlipperOnEdge = _ref.autoHideFlipperOnEdge,
    children = _ref.children,
    className = _ref.className,
    dir = _ref.dir,
    flipperBlurFocusOnClick = _ref.flipperBlurFocusOnClick,
    height = _ref.height,
    leftFlipperAriaLabel = _ref.leftFlipperAriaLabel,
    leftFlipperText = _ref.leftFlipperText,
    nonce = _ref.nonce,
    rightFlipperAriaLabel = _ref.rightFlipperAriaLabel,
    rightFlipperText = _ref.rightFlipperText,
    showDots = _ref.showDots,
    showFlipper = _ref.showFlipper,
    showScrollBar = _ref.showScrollBar,
    styleSet = _ref.styleSet;
  var styleOptions = useMemo(function () {
    return {
      autoCenter: autoCenter,
      autoHide: autoHide,
      autoHideFlipperOnEdge: autoHideFlipperOnEdge,
      dir: dir,
      leftFlipperAriaLabel: leftFlipperAriaLabel,
      leftFlipperText: leftFlipperText,
      flipperBlurFocusOnClick: flipperBlurFocusOnClick,
      rightFlipperAriaLabel: rightFlipperAriaLabel,
      rightFlipperText: rightFlipperText,
      showDots: showDots,
      showFlipper: showFlipper,
      showScrollBar: showScrollBar
    };
  }, [autoCenter, autoHide, autoHideFlipperOnEdge, dir, leftFlipperAriaLabel, leftFlipperText, flipperBlurFocusOnClick, rightFlipperAriaLabel, rightFlipperText, showDots, showFlipper, showScrollBar]);
  return /*#__PURE__*/React.createElement(Composer, {
    dir: dir,
    height: height,
    nonce: nonce,
    numItems: children ? Children.count(children) : 0,
    styleOptions: styleOptions,
    styleSet: styleSet
  }, /*#__PURE__*/React.createElement(BasicFilm, {
    className: className
  }, children));
};
ReactFilm.defaultProps = _objectSpread(_objectSpread({}, BasicFilm.defaultProps), {}, {
  autoCenter: undefined,
  autoHide: undefined,
  autoHideFlipperOnEdge: undefined,
  children: undefined,
  className: undefined,
  dir: undefined,
  flipperBlurFocusOnClick: undefined,
  leftFlipperAriaLabel: undefined,
  leftFlipperText: undefined,
  nonce: undefined,
  rightFlipperAriaLabel: undefined,
  rightFlipperText: undefined,
  showDots: undefined,
  showFlipper: undefined,
  showScrollBar: undefined,
  styleSet: undefined
});
ReactFilm.propTypes = _objectSpread(_objectSpread({}, BasicFilm.propTypes), {}, {
  autoCenter: PropTypes.bool,
  autoHide: PropTypes.bool,
  autoHideFlipperOnEdge: PropTypes.bool,
  children: PropTypes.any,
  className: PropTypes.string,
  dir: PropTypes.oneOf(['auto', 'ltr', 'rtl']),
  flipperBlurFocusOnClick: PropTypes.bool,
  height: PropTypes.number,
  leftFlipperAriaLabel: PropTypes.string,
  leftFlipperText: PropTypes.string,
  nonce: PropTypes.string,
  rightFlipperAriaLabel: PropTypes.string,
  rightFlipperText: PropTypes.string,
  showDots: PropTypes.bool,
  showFlipper: PropTypes.bool,
  showScrollBar: PropTypes.bool,
  styleSet: PropTypes.any
});
export default ReactFilm;
