// botframework-webchat-component/lib/BasicTranscript.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _botframeworkWebchatApi = require("botframework-webchat-api");

var _reactScrollToBottom = require("react-scroll-to-bottom");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _detectBrowser = require("./Utils/detectBrowser");

var _ActivityRow = _interopRequireDefault(require("./Transcript/ActivityRow"));

var _BasicTypingIndicator = _interopRequireDefault(require("./BasicTypingIndicator"));

var _FocusRedirector = _interopRequireDefault(require("./Utils/FocusRedirector"));

var _inputtableKey = _interopRequireDefault(require("./Utils/TypeFocusSink/inputtableKey"));

var _isZeroOrPositive = _interopRequireDefault(require("./Utils/isZeroOrPositive"));

var _KeyboardHelp = _interopRequireDefault(require("./Transcript/KeyboardHelp"));

var _LiveRegionTranscript = _interopRequireDefault(require("./Transcript/LiveRegionTranscript"));

var _tabbableElements = _interopRequireDefault(require("./Utils/tabbableElements"));

var _TranscriptFocusComposer = _interopRequireDefault(require("./providers/TranscriptFocus/TranscriptFocusComposer"));

var _useActiveDescendantId = _interopRequireDefault(require("./providers/TranscriptFocus/useActiveDescendantId"));

var _useActivityTreeWithRenderer = _interopRequireDefault(require("./providers/ActivityTree/useActivityTreeWithRenderer"));

var _useDispatchScrollPosition = _interopRequireDefault(require("./hooks/internal/useDispatchScrollPosition"));

var _useDispatchTranscriptFocusByActivityKey = _interopRequireDefault(require("./hooks/internal/useDispatchTranscriptFocusByActivityKey"));

var _useFocus = _interopRequireDefault(require("./hooks/useFocus"));

var _useFocusByActivityKey = _interopRequireDefault(require("./providers/TranscriptFocus/useFocusByActivityKey"));

var _useFocusedActivityKey = _interopRequireDefault(require("./providers/TranscriptFocus/useFocusedActivityKey"));

var _useFocusedExplicitly3 = _interopRequireDefault(require("./providers/TranscriptFocus/useFocusedExplicitly"));

var _useFocusRelativeActivity = _interopRequireDefault(require("./providers/TranscriptFocus/useFocusRelativeActivity"));

var _useObserveFocusVisible = _interopRequireDefault(require("./hooks/internal/useObserveFocusVisible"));

var _usePrevious = _interopRequireDefault(require("./hooks/internal/usePrevious"));

var _useRegisterFocusTranscript = _interopRequireDefault(require("./hooks/internal/useRegisterFocusTranscript"));

var _useRegisterScrollRelative = _interopRequireDefault(require("./hooks/internal/useRegisterScrollRelative"));

var _useRegisterScrollTo = _interopRequireDefault(require("./hooks/internal/useRegisterScrollTo"));

var _useRegisterScrollToEnd = _interopRequireDefault(require("./hooks/internal/useRegisterScrollToEnd"));

var _useStyleSet5 = _interopRequireDefault(require("./hooks/useStyleSet"));

var _useStyleToEmotionObject = _interopRequireDefault(require("./hooks/internal/useStyleToEmotionObject"));

var _useUniqueId = _interopRequireDefault(require("./hooks/internal/useUniqueId"));

var _useValueRef = _interopRequireDefault(require("./hooks/internal/useValueRef"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useActivityKeys = _botframeworkWebchatApi.hooks.useActivityKeys,
    useActivityKeysByRead = _botframeworkWebchatApi.hooks.useActivityKeysByRead,
    useCreateActivityStatusRenderer = _botframeworkWebchatApi.hooks.useCreateActivityStatusRenderer,
    useCreateAvatarRenderer = _botframeworkWebchatApi.hooks.useCreateAvatarRenderer,
    useCreateScrollToEndButtonRenderer = _botframeworkWebchatApi.hooks.useCreateScrollToEndButtonRenderer,
    useDirection = _botframeworkWebchatApi.hooks.useDirection,
    useGetActivityByKey = _botframeworkWebchatApi.hooks.useGetActivityByKey,
    useGetKeyByActivity = _botframeworkWebchatApi.hooks.useGetKeyByActivity,
    useGetKeyByActivityId = _botframeworkWebchatApi.hooks.useGetKeyByActivityId,
    useLastAcknowledgedActivityKey = _botframeworkWebchatApi.hooks.useLastAcknowledgedActivityKey,
    useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer,
    useMarkActivityKeyAsRead = _botframeworkWebchatApi.hooks.useMarkActivityKeyAsRead,
    useMarkAllAsAcknowledged = _botframeworkWebchatApi.hooks.useMarkAllAsAcknowledged,
    useStyleOptions = _botframeworkWebchatApi.hooks.useStyleOptions;
var ROOT_STYLE = {
  '&.webchat__basic-transcript': {
    display: 'flex',
    flexDirection: 'column',
    overflow: 'hidden',
    // Make sure to set "position: relative" here to form another stacking context for the scroll-to-end button.
    // Stacking context help isolating elements that use "z-index" from global pollution.
    // https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context
    position: 'relative',
    '& .webchat__basic-transcript__filler': {
      flex: 1
    },
    '& .webchat__basic-transcript__scrollable': {
      display: 'flex',
      flexDirection: 'column',
      overflowX: 'hidden',
      WebkitOverflowScrolling: 'touch'
    },
    '& .webchat__basic-transcript__transcript': {
      listStyleType: 'none'
    }
  }
};
// TODO: [P1] #4133 Add telemetry for computing how many re-render done so far.
var InternalTranscript = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var activityElementMapRef = _ref.activityElementMapRef,
      className = _ref.className;

  var _useStyleSet = (0, _useStyleSet5.default)(),
      _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
      basicTranscriptStyleSet = _useStyleSet2[0].basicTranscript;

  var _useStyleOptions = useStyleOptions(),
      _useStyleOptions2 = _slicedToArray(_useStyleOptions, 1),
      _useStyleOptions2$ = _useStyleOptions2[0],
      bubbleFromUserNubOffset = _useStyleOptions2$.bubbleFromUserNubOffset,
      bubbleNubOffset = _useStyleOptions2$.bubbleNubOffset,
      groupTimestamp = _useStyleOptions2$.groupTimestamp,
      showAvatarInGroup = _useStyleOptions2$.showAvatarInGroup;

  var _useActiveDescendantI = (0, _useActiveDescendantId.default)(),
      _useActiveDescendantI2 = _slicedToArray(_useActiveDescendantI, 1),
      activeDescendantId = _useActiveDescendantI2[0];

  var _useActivityTreeWithR = (0, _useActivityTreeWithRenderer.default)(),
      _useActivityTreeWithR2 = _slicedToArray(_useActivityTreeWithR, 1),
      activityWithRendererTree = _useActivityTreeWithR2[0];

  var _useDirection = useDirection(),
      _useDirection2 = _slicedToArray(_useDirection, 1),
      direction = _useDirection2[0];

  var _useFocusedActivityKe = (0, _useFocusedActivityKey.default)(),
      _useFocusedActivityKe2 = _slicedToArray(_useFocusedActivityKe, 1),
      focusedActivityKey = _useFocusedActivityKe2[0];

  var _useFocusedExplicitly = (0, _useFocusedExplicitly3.default)(),
      _useFocusedExplicitly2 = _slicedToArray(_useFocusedExplicitly, 1),
      focusedExplicitly = _useFocusedExplicitly2[0];

  var createActivityStatusRenderer = useCreateActivityStatusRenderer();
  var createAvatarRenderer = useCreateAvatarRenderer();
  var focus = (0, _useFocus.default)();
  var focusByActivityKey = (0, _useFocusByActivityKey.default)();
  var focusRelativeActivity = (0, _useFocusRelativeActivity.default)();
  var getActivityByKey = useGetActivityByKey();
  var getKeyByActivity = useGetKeyByActivity();
  var getKeyByActivityId = useGetKeyByActivityId();
  var localize = useLocalizer();
  var rootClassName = (0, _useStyleToEmotionObject.default)()(ROOT_STYLE) + '';
  var rootElementRef = (0, _react.useRef)();
  var terminatorLabelId = (0, _useUniqueId.default)('webchat__basic-transcript__terminator-label');
  var terminatorRef = (0, _react.useRef)();
  var focusedActivityKeyRef = (0, _useValueRef.default)(focusedActivityKey);
  var hideAllTimestamps = groupTimestamp === false;
  var terminatorText = localize('TRANSCRIPT_TERMINATOR_TEXT');
  var transcriptAriaLabel = localize('TRANSCRIPT_ARIA_LABEL_ALT');
  var callbackRef = (0, _react.useCallback)(function (element) {
    if (typeof ref === 'function') {
      ref(element);
    } else {
      ref.current = element;
    }

    rootElementRef.current = element;
  }, [ref, rootElementRef]); // Flatten the tree back into an array with information related to rendering.

  var renderingElements = (0, _react.useMemo)(function () {
    var renderingElements = [];
    var topSideBotNub = (0, _isZeroOrPositive.default)(bubbleNubOffset);
    var topSideUserNub = (0, _isZeroOrPositive.default)(bubbleFromUserNubOffset);
    activityWithRendererTree.forEach(function (entriesWithSameSender) {
      var _entriesWithSameSende = _slicedToArray(entriesWithSameSender, 1),
          _entriesWithSameSende2 = _slicedToArray(_entriesWithSameSende[0], 1),
          firstActivity = _entriesWithSameSende2[0].activity;

      var renderAvatar = createAvatarRenderer({
        activity: firstActivity
      });
      entriesWithSameSender.forEach(function (entriesWithSameSenderAndStatus, indexWithinSenderGroup) {
        var firstInSenderGroup = !indexWithinSenderGroup;
        var lastInSenderGroup = indexWithinSenderGroup === entriesWithSameSender.length - 1;
        entriesWithSameSenderAndStatus.forEach(function (_ref2, indexWithinSenderAndStatusGroup) {
          var _activity$from;

          var activity = _ref2.activity,
              renderActivity = _ref2.renderActivity;
          // We only show the timestamp at the end of the sender group. But we always show the "Send failed, retry" prompt.
          var firstInSenderAndStatusGroup = !indexWithinSenderAndStatusGroup;
          var key = getKeyByActivity(activity);
          var lastInSenderAndStatusGroup = indexWithinSenderAndStatusGroup === entriesWithSameSenderAndStatus.length - 1;
          var renderActivityStatus = createActivityStatusRenderer({
            activity: activity,
            nextVisibleActivity: undefined
          });
          var topSideNub = ((_activity$from = activity.from) === null || _activity$from === void 0 ? void 0 : _activity$from.role) === 'user' ? topSideUserNub : topSideBotNub;
          var showCallout; // Depending on the "showAvatarInGroup" setting, the avatar will render in different positions.

          if (showAvatarInGroup === 'sender') {
            if (topSideNub) {
              showCallout = firstInSenderGroup && firstInSenderAndStatusGroup;
            } else {
              showCallout = lastInSenderGroup && lastInSenderAndStatusGroup;
            }
          } else if (showAvatarInGroup === 'status') {
            if (topSideNub) {
              showCallout = firstInSenderAndStatusGroup;
            } else {
              showCallout = lastInSenderAndStatusGroup;
            }
          } else {
            showCallout = true;
          }

          renderingElements.push({
            activity: activity,
            // After the element is mounted, set it to activityElementsRef.
            callbackRef: function callbackRef(activityElement) {
              activityElement ? activityElementMapRef.current.set(key, activityElement) : activityElementMapRef.current.delete(key);
            },
            // "hideTimestamp" is a render-time parameter for renderActivityStatus().
            // If true, it will hide the timestamp, but it will continue to show the
            // retry prompt. And show the screen reader version of the timestamp.
            hideTimestamp: hideAllTimestamps || indexWithinSenderAndStatusGroup !== entriesWithSameSenderAndStatus.length - 1,
            key: key,
            renderActivity: renderActivity,
            renderActivityStatus: renderActivityStatus,
            renderAvatar: renderAvatar,
            showCallout: showCallout
          });
        });
      });
    });
    return renderingElements;
  }, [activityElementMapRef, activityWithRendererTree, bubbleFromUserNubOffset, bubbleNubOffset, createActivityStatusRenderer, createAvatarRenderer, getKeyByActivity, hideAllTimestamps, showAvatarInGroup]);
  var scrollToBottomScrollTo = (0, _reactScrollToBottom.useScrollTo)();
  var scrollToBottomScrollToEnd = (0, _reactScrollToBottom.useScrollToEnd)();
  var scrollTo = (0, _react.useCallback)(function (position) {
    var _ref3 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref3$behavior = _ref3.behavior,
        behavior = _ref3$behavior === void 0 ? 'auto' : _ref3$behavior;

    if (!position) {
      throw new Error('botframework-webchat: First argument passed to "useScrollTo" must be a ScrollPosition object.');
    }

    var activityId = position.activityID,
        scrollTop = position.scrollTop;

    if (typeof scrollTop !== 'undefined') {
      scrollToBottomScrollTo(scrollTop, {
        behavior: behavior
      });
    } else if (typeof activityId !== 'undefined') {
      var _activityElementMapRe;

      var activityBoundingBoxElement = (_activityElementMapRe = activityElementMapRef.current.get(getKeyByActivityId(activityId))) === null || _activityElementMapRe === void 0 ? void 0 : _activityElementMapRe.querySelector('.webchat__basic-transcript__activity-active-descendant');
      var scrollableElement = rootElementRef.current.querySelector('.webchat__basic-transcript__scrollable');

      if (scrollableElement && activityBoundingBoxElement) {
        // ESLint conflict with TypeScript. The result of getClientRects() is not an Array but DOMRectList, and cannot be destructured.
        // eslint-disable-next-line prefer-destructuring
        var activityBoundingBoxElementClientRect = activityBoundingBoxElement.getClientRects()[0]; // ESLint conflict with TypeScript. The result of getClientRects() is not an Array but DOMRectList, and cannot be destructured.
        // eslint-disable-next-line prefer-destructuring

        var scrollableElementClientRect = scrollableElement.getClientRects()[0]; // If either the activity or the transcript scrollable is not on DOM, we will not scroll the view.

        if (activityBoundingBoxElementClientRect && scrollableElementClientRect) {
          var activityHeight = activityBoundingBoxElementClientRect.height,
              activityY = activityBoundingBoxElementClientRect.y;
          var scrollableHeight = scrollableElementClientRect.height;
          var activityOffsetTop = activityY + scrollableElement.scrollTop;

          var _scrollTop = Math.min(activityOffsetTop, activityOffsetTop - scrollableHeight + activityHeight);

          scrollToBottomScrollTo(_scrollTop, {
            behavior: behavior
          });
        }
      }
    }
  }, [activityElementMapRef, getKeyByActivityId, rootElementRef, scrollToBottomScrollTo]);
  var scrollToEnd = (0, _react.useCallback)(function () {
    return scrollToBottomScrollToEnd({
      behavior: 'smooth'
    });
  }, [scrollToBottomScrollToEnd]);
  var scrollRelative = (0, _react.useCallback)(function (direction) {
    var _ref4 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        displacement = _ref4.displacement;

    var rootElement = rootElementRef.current;

    if (!rootElement) {
      return;
    }

    var scrollable = rootElement.querySelector('.webchat__basic-transcript__scrollable');
    var nextScrollTop;

    if (typeof displacement === 'number') {
      // eslint-disable-next-line no-magic-numbers
      nextScrollTop = scrollable.scrollTop + (direction === 'down' ? 1 : -1) * displacement;
    } else {
      // eslint-disable-next-line no-magic-numbers
      nextScrollTop = scrollable.scrollTop + (direction === 'down' ? 1 : -1) * scrollable.offsetHeight;
    }

    scrollTo({
      scrollTop: Math.max(0, Math.min(scrollable.scrollHeight - scrollable.offsetHeight, nextScrollTop))
    }, {
      behavior: 'smooth'
    });
  }, [rootElementRef, scrollTo]); // Since there could be multiple instances of <BasicTranscript> inside the <Composer>, when the developer calls `scrollXXX`, we need to call it on all instances.
  // We call `useRegisterScrollXXX` to register a callback function, the `useScrollXXX` will multiplex the call into each instance of <BasicTranscript>.

  (0, _useRegisterScrollTo.default)(scrollTo);
  (0, _useRegisterScrollToEnd.default)(scrollToEnd);
  (0, _useRegisterScrollRelative.default)(scrollRelative);
  var markActivityKeyAsRead = useMarkActivityKeyAsRead();
  var dispatchScrollPositionWithActivityId = (0, _useDispatchScrollPosition.default)(); // TODO: [P2] We should use IntersectionObserver to track what activity is in the scrollable.
  //            However, IntersectionObserver is not available on IE11, we need to make a limited polyfill in React style.

  var handleScrollPosition = (0, _react.useCallback)(function (_ref5) {
    var _ref6;

    var scrollTop = _ref5.scrollTop;
    var rootElement = rootElementRef.current;

    if (!rootElement) {
      return;
    }

    var scrollableElement = rootElement.querySelector('.webchat__basic-transcript__scrollable'); // "getClientRects()" is not returning an array, thus, it is not destructurable.
    // eslint-disable-next-line prefer-destructuring

    var scrollableElementClientRect = scrollableElement.getClientRects()[0]; // If the scrollable is not mounted, we cannot measure which activity is in view. Thus, we will not fire any events.

    if (!scrollableElementClientRect) {
      return;
    }

    var scrollableClientBottom = scrollableElementClientRect.bottom; // Find the activity just above scroll view bottom.
    // If the scroll view is already on top, get the first activity.

    var activityElements = Array.from(activityElementMapRef.current.entries());
    var activityKeyJustAboveScrollBottom = (_ref6 = scrollableElement.scrollTop ? activityElements.reverse() // Add subpixel tolerance
    .find(function (_ref7) {
      var _ref8 = _slicedToArray(_ref7, 2),
          element = _ref8[1];

      // "getClientRects()" is not returning an array, thus, it is not destructurable.
      // eslint-disable-next-line prefer-destructuring
      var elementClientRect = element.getClientRects()[0]; // If the activity is not attached to DOM tree, we should not count it as "bottommost visible activity", as it is not visible.

      return elementClientRect && elementClientRect.bottom < scrollableClientBottom + 1;
    }) : activityElements[0]) === null || _ref6 === void 0 ? void 0 : _ref6[0]; // When the end-user slowly scrolling the view down, we will mark activity as read when the message fully appear on the screen.

    activityKeyJustAboveScrollBottom && markActivityKeyAsRead(activityKeyJustAboveScrollBottom);

    if (dispatchScrollPositionWithActivityId) {
      var activity = getActivityByKey(activityKeyJustAboveScrollBottom);
      dispatchScrollPositionWithActivityId(_objectSpread(_objectSpread({}, activity ? {
        activityID: activity.id
      } : {}), {}, {
        scrollTop: scrollTop
      }));
    }
  }, [activityElementMapRef, dispatchScrollPositionWithActivityId, getActivityByKey, markActivityKeyAsRead, rootElementRef]);
  (0, _reactScrollToBottom.useObserveScrollPosition)(handleScrollPosition);
  var handleTranscriptKeyDown = (0, _react.useCallback)(function (event) {
    var target = event.target;
    var fromEndOfTranscriptIndicator = target === terminatorRef.current;
    var fromTranscript = target === event.currentTarget;

    if (!fromEndOfTranscriptIndicator && !fromTranscript) {
      return;
    }

    var handled = true;

    switch (event.key) {
      case 'ArrowDown':
        focusRelativeActivity(fromEndOfTranscriptIndicator ? 0 : 1);
        break;

      case 'ArrowUp':
        // eslint-disable-next-line no-magic-numbers
        focusRelativeActivity(fromEndOfTranscriptIndicator ? 0 : -1);
        break;

      case 'End':
        focusRelativeActivity(Infinity);
        break;

      case 'Enter':
        // This is capturing plain ENTER.
        // When screen reader is not running, or screen reader is running outside of scan mode, the ENTER key will be captured here.
        if (!fromEndOfTranscriptIndicator) {
          var _activityElementMapRe2, _tabbableElements$;

          var body = (_activityElementMapRe2 = activityElementMapRef.current.get(focusedActivityKeyRef.current)) === null || _activityElementMapRe2 === void 0 ? void 0 : _activityElementMapRe2.querySelector('.webchat__basic-transcript__activity-body');
          (_tabbableElements$ = (0, _tabbableElements.default)(body)[0]) === null || _tabbableElements$ === void 0 ? void 0 : _tabbableElements$.focus();
        }

        break;

      case 'Escape':
        focus('sendBoxWithoutKeyboard');
        break;

      case 'Home':
        focusRelativeActivity(-Infinity);
        break;

      default:
        handled = false;
        break;
    }

    if (handled) {
      event.preventDefault(); // If a custom HTML control wants to handle up/down arrow, we will prevent them from listening to this event to prevent bugs due to handling arrow keys twice.

      event.stopPropagation();
    }
  }, [activityElementMapRef, focus, focusedActivityKeyRef, focusRelativeActivity, terminatorRef]);
  var handleTranscriptKeyDownCapture = (0, _react.useCallback)(function (event) {
    var altKey = event.altKey,
        ctrlKey = event.ctrlKey,
        key = event.key,
        metaKey = event.metaKey,
        target = event.target;

    if (altKey || ctrlKey && key !== 'v' || metaKey || !(0, _inputtableKey.default)(key) && key !== 'Backspace') {
      // Ignore if one of the utility key (except SHIFT) is pressed
      // E.g. CTRL-C on a link in one of the message should not jump to chat box
      // E.g. "A" or "Backspace" should jump to chat box
      return;
    } // Send keystrokes to send box if we are focusing on the transcript or terminator.


    if (target === event.currentTarget || target === terminatorRef.current) {
      event.stopPropagation();
      focus('sendBox');
    }
  }, [focus]);
  (0, _useRegisterFocusTranscript.default)((0, _react.useCallback)(function () {
    return focusByActivityKey(undefined);
  }, [focusByActivityKey])); // When the focusing activity has changed, dispatch an event to observers of "useObserveTranscriptFocus".

  var dispatchTranscriptFocusByActivityKey = (0, _useDispatchTranscriptFocusByActivityKey.default)(); // Dispatch a "transcript focus" event based on user selection.
  // We should not dispatch "transcript focus" when a new activity come. Although the selection change, it is not initiated from the user.

  (0, _react.useMemo)(function () {
    return dispatchTranscriptFocusByActivityKey(focusedExplicitly ? focusedActivityKey : undefined);
  }, [dispatchTranscriptFocusByActivityKey, focusedActivityKey, focusedExplicitly]); // When the transcript is being focused on, we should dispatch a "transcriptfocus" event.

  var handleFocus = (0, _react.useCallback)( // We call "focusByActivityKey" with activity key of "true".
  // It means, tries to focus on anything.
  function (_ref9) {
    var currentTarget = _ref9.currentTarget,
        target = _ref9.target;
    return target === currentTarget && focusByActivityKey(true, false);
  }, [focusByActivityKey]); // This is required by IE11.
  // When the user clicks on and empty space (a.k.a. filler) in an empty transcript, IE11 says the focus is on the <div className="filler">,
  // despite the fact there are no "tabIndex" attributes set on the filler.
  // We need to artificially send the focus back to the transcript.

  var handleFocusFiller = (0, _react.useCallback)(function () {
    return focusByActivityKey(undefined);
  }, [focusByActivityKey]); // When focus into the transcript using TAB/SHIFT-TAB, scroll the focused activity into view.

  (0, _useObserveFocusVisible.default)(rootElementRef, (0, _react.useCallback)(function () {
    return focusByActivityKey(undefined);
  }, [focusByActivityKey]));
  return /*#__PURE__*/_react.default.createElement("div", {
    // Although Android TalkBack 12.1 does not support `aria-activedescendant`, when used, it become buggy and will narrate content twice.
    // We are disabling `aria-activedescendant` for Android. See <ActivityRow> for details.
    "aria-activedescendant": _detectBrowser.android ? undefined : activeDescendantId,
    "aria-label": transcriptAriaLabel,
    className: (0, _classnames.default)('webchat__basic-transcript', basicTranscriptStyleSet + '', rootClassName, (className || '') + ''),
    dir: direction,
    onFocus: handleFocus,
    onKeyDown: handleTranscriptKeyDown,
    onKeyDownCapture: handleTranscriptKeyDownCapture,
    ref: callbackRef // "aria-activedescendant" will only works with a number of roles and it must be explicitly set.
    // https://www.w3.org/TR/wai-aria/#aria-activedescendant
    ,
    role: "group" // For up/down arrow key navigation across activities, this component must be included in the tab sequence.
    // Otherwise, "aria-activedescendant" will not be narrated when the user press up/down arrow keys.
    // https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_focus_activedescendant
    ,
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_LiveRegionTranscript.default, {
    activityElementMapRef: activityElementMapRef
  }), /*#__PURE__*/_react.default.createElement(InternalTranscriptScrollable, {
    onFocusFiller: handleFocusFiller,
    terminatorRef: terminatorRef
  }, renderingElements.map(function (_ref10) {
    var activity = _ref10.activity,
        callbackRef = _ref10.callbackRef,
        hideTimestamp = _ref10.hideTimestamp,
        key = _ref10.key,
        renderActivity = _ref10.renderActivity,
        renderActivityStatus = _ref10.renderActivityStatus,
        renderAvatar = _ref10.renderAvatar,
        showCallout = _ref10.showCallout;
    return /*#__PURE__*/_react.default.createElement(_ActivityRow.default, {
      activity: activity,
      key: key,
      ref: callbackRef
    }, renderActivity({
      hideTimestamp: hideTimestamp,
      renderActivityStatus: renderActivityStatus,
      renderAvatar: renderAvatar,
      showCallout: showCallout
    }));
  })), !!renderingElements.length && /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_FocusRedirector.default, {
    redirectRef: rootElementRef
  }), /*#__PURE__*/_react.default.createElement("div", {
    "aria-labelledby": terminatorLabelId,
    className: "webchat__basic-transcript__terminator",
    ref: terminatorRef,
    role: "note",
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "webchat__basic-transcript__terminator-body"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "webchat__basic-transcript__terminator-text",
    id: terminatorLabelId
  }, terminatorText)))), /*#__PURE__*/_react.default.createElement("div", {
    className: "webchat__basic-transcript__focus-indicator"
  }));
});
InternalTranscript.defaultProps = {
  className: ''
};
InternalTranscript.displayName = 'InternalTranscript';
InternalTranscript.propTypes = {
  // PropTypes cannot validate precisely with its TypeScript counterpart.
  // @ts-ignore
  activityElementMapRef: _propTypes.default.shape({
    current: _propTypes.default.instanceOf(Map)
  }).isRequired,
  className: _propTypes.default.string
};

// Separating high-frequency hooks to improve performance.
var InternalTranscriptScrollable = function InternalTranscriptScrollable(_ref11) {
  var children = _ref11.children,
      onFocusFiller = _ref11.onFocusFiller,
      terminatorRef = _ref11.terminatorRef;

  var _useStyleSet3 = (0, _useStyleSet5.default)(),
      _useStyleSet4 = _slicedToArray(_useStyleSet3, 1),
      activitiesStyleSet = _useStyleSet4[0].activities;

  var _useAnimatingToEnd = (0, _reactScrollToBottom.useAnimatingToEnd)(),
      _useAnimatingToEnd2 = _slicedToArray(_useAnimatingToEnd, 1),
      animatingToEnd = _useAnimatingToEnd2[0];

  var _useAtEnd = (0, _reactScrollToBottom.useAtEnd)(),
      _useAtEnd2 = _slicedToArray(_useAtEnd, 1),
      atEnd = _useAtEnd2[0];

  var _useActivityKeysByRea = useActivityKeysByRead(),
      _useActivityKeysByRea2 = _slicedToArray(_useActivityKeysByRea, 2),
      unreadActivityKeys = _useActivityKeysByRea2[1];

  var _useSticky = (0, _reactScrollToBottom.useSticky)(),
      _useSticky2 = _slicedToArray(_useSticky, 1),
      sticky = _useSticky2[0];

  var _useStyleOptions3 = useStyleOptions(),
      _useStyleOptions4 = _slicedToArray(_useStyleOptions3, 1),
      styleOptions = _useStyleOptions4[0];

  var focusByActivityKey = (0, _useFocusByActivityKey.default)();
  var localize = useLocalizer();
  var markActivityKeyAsRead = useMarkActivityKeyAsRead();
  var markAllAsAcknowledged = useMarkAllAsAcknowledged();
  var scrollToEnd = (0, _reactScrollToBottom.useScrollToEnd)();
  var prevSticky = (0, _usePrevious.default)(sticky);
  var transcriptRoleDescription = localize('TRANSCRIPT_ARIA_ROLE_ALT');
  var stickyChangedToTrue = prevSticky !== sticky && sticky; // Acknowledged means either:
  // 1. The user sent a message
  //    - We don't need a condition here. When Web Chat sends the user's message, it will scroll to bottom, and it will trigger condition 2 below.
  // 2. The user scroll to the bottom of the transcript, from a non-bottom scroll position
  //    - If the transcript is already at the bottom, the user needs to scroll up and then go back down
  //    - What happens if we are relaxing "scrolled from a non-bottom scroll position":
  //      1. The condition will become solely "at the bottom of the transcript"
  //      2. Auto-scroll will always scroll the transcript to the bottom
  //      3. Web Chat will always acknowledge all activities as it is at the bottom
  //      4. Acknowledge flag become useless
  //      5. Therefore, even the developer set "pause after 3 activities", if activities are coming in at a slow pace (not batched in a single render)
  //         Web Chat will keep scrolling and not snapped/paused
  // Note: When Web Chat is loaded, there are no activities acknowledged. We need to assume all arriving activities are acknowledged until end-user sends their first activity.
  //       Activities loaded initially could be from conversation history. Without assuming acknowledgement, Web Chat will not scroll initially (as everything is not acknowledged).
  //       It would be better if the chat adapter should let Web Chat know if the activity is loaded from history or not.
  // TODO: [P2] #3670 Move the "conversation history acknowledgement" logic mentioned above to polyfill of chat adapters.
  //       1. Chat adapter should send "acknowledged" as part of "channelData"
  //       2. If "acknowledged" is "undefined", we set it to:
  //          a. true, if there are no egress activities yet
  //          b. Otherwise, false

  (0, _react.useMemo)(function () {
    return stickyChangedToTrue && // TODO: [P2] Both `markActivityKeyAsRead` and `markAllAsAcknowledged` hook are setters of useState.
    //       This means, in a render loop, we will be calling setter and will cause another re-render.
    //       This is not trivial but we should think if there is a way to avoid this.
    markAllAsAcknowledged();
  }, [markAllAsAcknowledged, stickyChangedToTrue]);

  var _useActivityTreeWithR3 = (0, _useActivityTreeWithRenderer.default)({
    flat: true
  }),
      _useActivityTreeWithR4 = _slicedToArray(_useActivityTreeWithR3, 1),
      flattenedActivityTreeWithRenderer = _useActivityTreeWithR4[0];

  var getKeyByActivity = useGetKeyByActivity();
  var renderingActivityKeys = (0, _react.useMemo)(function () {
    return flattenedActivityTreeWithRenderer.map(function (_ref12) {
      var activity = _ref12.activity;
      return getKeyByActivity(activity);
    });
  }, [flattenedActivityTreeWithRenderer, getKeyByActivity]);
  var renderingActivityKeysRef = (0, _useValueRef.default)(renderingActivityKeys); // To prevent flashy button, we are not waiting for another render loop to update the `[readActivityKeys, unreadActivityKeys]` state.
  // Instead, we are building the next one in this `useMemo` call.

  var nextUnreadActivityKeys = (0, _react.useMemo)(function () {
    // This code need to be careful reviewed as it will cause another render. The code should be converging.
    // After we call `markActivityKeyAsRead`, everything will be read and nothing will be unread.
    // That means, in next render, `unreadActivityKeys` will be emptied and the `markActivityKeyAsRead` will not get called again.
    if (sticky && unreadActivityKeys.length) {
      markActivityKeyAsRead(unreadActivityKeys[unreadActivityKeys.length - 1]);
      return [];
    }

    return unreadActivityKeys;
  }, [markActivityKeyAsRead, sticky, unreadActivityKeys]);
  var nextUnreadActivityKeysRef = (0, _useValueRef.default)(nextUnreadActivityKeys); // If we are rendering anything that is unread, we should show the "New messages" button.
  // Not everything in the `unreadActivityKeys` are rendered, say, bot typing indicator.
  // We should not show the "New messages" button for bot typing indicator as it will confuse the user.

  var unread = (0, _react.useMemo)(function () {
    return nextUnreadActivityKeys.some(function (key) {
      return renderingActivityKeys.includes(key);
    });
  }, [renderingActivityKeys, nextUnreadActivityKeys]);
  var handleScrollToEndButtonClick = (0, _react.useCallback)(function () {
    scrollToEnd({
      behavior: 'smooth'
    });
    var renderingActivityKeys = renderingActivityKeysRef.current; // After the "New message" button is clicked, focus on the first unread activity which will be rendered.

    var firstUnreadRenderingActivityKey = nextUnreadActivityKeysRef.current.find(function (key) {
      return renderingActivityKeys.includes(key);
    });

    if (firstUnreadRenderingActivityKey) {
      focusByActivityKey(firstUnreadRenderingActivityKey);
    } else {
      var _terminatorRef$curren;

      // If no unread activity, send the focus to the terminator block.
      (_terminatorRef$curren = terminatorRef.current) === null || _terminatorRef$curren === void 0 ? void 0 : _terminatorRef$curren.focus();
    }
  }, [focusByActivityKey, nextUnreadActivityKeysRef, renderingActivityKeysRef, scrollToEnd, terminatorRef]);
  var renderScrollToEndButton = useCreateScrollToEndButtonRenderer()({
    atEnd: animatingToEnd || atEnd || sticky,
    styleOptions: styleOptions,
    unread: unread
  });
  var hasAnyChild = !!_react.default.Children.count(children);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, renderScrollToEndButton && renderScrollToEndButton({
    onClick: handleScrollToEndButtonClick
  }), hasAnyChild && /*#__PURE__*/_react.default.createElement(_FocusRedirector.default, {
    redirectRef: terminatorRef
  }), /*#__PURE__*/_react.default.createElement(_reactScrollToBottom.Panel, {
    className: "webchat__basic-transcript__scrollable"
  }, /*#__PURE__*/_react.default.createElement("div", {
    "aria-hidden": true,
    className: "webchat__basic-transcript__filler",
    onFocus: onFocusFiller
  }), hasAnyChild && /*#__PURE__*/_react.default.createElement("section", {
    "aria-roledescription": transcriptRoleDescription,
    className: (0, _classnames.default)(activitiesStyleSet + '', 'webchat__basic-transcript__transcript'),
    role: "feed"
  }, children), /*#__PURE__*/_react.default.createElement(_BasicTypingIndicator.default, null)));
};

InternalTranscriptScrollable.propTypes = {
  children: _propTypes.default.any.isRequired,
  onFocusFiller: _propTypes.default.func.isRequired,
  terminatorRef: _propTypes.default.any.isRequired
};

// "scroller" is the auto-scroll limiter, a.k.a. auto scroll snap.
var useScroller = function useScroller(activityElementMapRef) {
  var _useActivityKeys = useActivityKeys(),
      _useActivityKeys2 = _slicedToArray(_useActivityKeys, 1),
      activityKeys = _useActivityKeys2[0];

  var _useLastAcknowledgedA = useLastAcknowledgedActivityKey(),
      _useLastAcknowledgedA2 = _slicedToArray(_useLastAcknowledgedA, 1),
      lastAcknowledgedActivityKey = _useLastAcknowledgedA2[0];

  var _useStyleOptions5 = useStyleOptions(),
      _useStyleOptions6 = _slicedToArray(_useStyleOptions5, 1),
      styleOptions = _useStyleOptions6[0];

  var activityKeysRef = (0, _useValueRef.default)(activityKeys);
  var lastAcknowledgedActivityKeyRef = (0, _useValueRef.default)(lastAcknowledgedActivityKey);
  var styleOptionsRef = (0, _useValueRef.default)(styleOptions);
  return (0, _react.useCallback)(function (_ref13) {
    var offsetHeight = _ref13.offsetHeight,
        scrollTop = _ref13.scrollTop;
    var _styleOptionsRef$curr = styleOptionsRef.current,
        autoScrollSnapOnActivity = _styleOptionsRef$curr.autoScrollSnapOnActivity,
        autoScrollSnapOnActivityOffset = _styleOptionsRef$curr.autoScrollSnapOnActivityOffset,
        autoScrollSnapOnPage = _styleOptionsRef$curr.autoScrollSnapOnPage,
        autoScrollSnapOnPageOffset = _styleOptionsRef$curr.autoScrollSnapOnPageOffset;
    var patchedAutoScrollSnapOnActivity = typeof autoScrollSnapOnActivity === 'number' ? Math.max(0, autoScrollSnapOnActivity) : autoScrollSnapOnActivity ? 1 : 0;
    var patchedAutoScrollSnapOnPage = typeof autoScrollSnapOnPage === 'number' ? Math.max(0, Math.min(1, autoScrollSnapOnPage)) : autoScrollSnapOnPage ? 1 : 0;
    var patchedAutoScrollSnapOnActivityOffset = typeof autoScrollSnapOnActivityOffset === 'number' ? autoScrollSnapOnActivityOffset : 0;
    var patchedAutoScrollSnapOnPageOffset = typeof autoScrollSnapOnPageOffset === 'number' ? autoScrollSnapOnPageOffset : 0;

    if (patchedAutoScrollSnapOnActivity || patchedAutoScrollSnapOnPage) {
      var activityElementMap = activityElementMapRef.current;
      var _activityKeys = activityKeysRef.current;
      var _lastAcknowledgedActivityKey = lastAcknowledgedActivityKeyRef.current;
      var values = [];

      var lastAcknowledgedActivityKeyIndex = _activityKeys.indexOf(_lastAcknowledgedActivityKey);

      if (~lastAcknowledgedActivityKeyIndex) {
        // The activity that we acknowledged could be not rendered, such as post back activity.
        // When calculating scroll snap, we can only base on the first unacknowledged-and-rendering activity.
        var renderingActivityKeys = Array.from(activityElementMap.keys());
        var firstUnacknowledgedActivityElementIndex = -1;

        var _iterator = _createForOfIteratorHelper(_activityKeys.slice(0, lastAcknowledgedActivityKeyIndex + 1).reverse()),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var acknowledgedActivityKey = _step.value;
            var index = renderingActivityKeys.indexOf(acknowledgedActivityKey);

            if (~index) {
              if (index !== renderingActivityKeys.length - 1) {
                firstUnacknowledgedActivityElementIndex = index + 1;
              }

              break;
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        if (~firstUnacknowledgedActivityElementIndex) {
          var activityElements = Array.from(activityElementMap.values());

          if (patchedAutoScrollSnapOnActivity) {
            // Gets the activity element which we should snap to.
            var nthUnacknowledgedActivityElement = activityElements[firstUnacknowledgedActivityElementIndex + patchedAutoScrollSnapOnActivity - 1];

            if (nthUnacknowledgedActivityElement) {
              var nthUnacknowledgedActivityBoundingBoxElement = nthUnacknowledgedActivityElement === null || nthUnacknowledgedActivityElement === void 0 ? void 0 : nthUnacknowledgedActivityElement.querySelector('.webchat__basic-transcript__activity-active-descendant');
              var nthUnacknowledgedActivityOffsetTop = nthUnacknowledgedActivityElement.offsetTop + nthUnacknowledgedActivityBoundingBoxElement.offsetTop;
              values.push(nthUnacknowledgedActivityOffsetTop + nthUnacknowledgedActivityBoundingBoxElement.offsetHeight - offsetHeight - scrollTop + patchedAutoScrollSnapOnActivityOffset);
            }
          }

          if (patchedAutoScrollSnapOnPage) {
            var firstUnacknowledgedActivityElement = activityElements[+firstUnacknowledgedActivityElementIndex];
            var firstUnacknowledgedActivityBoundingBoxElement = firstUnacknowledgedActivityElement.querySelector('.webchat__basic-transcript__activity-active-descendant');
            var firstUnacknowledgedActivityOffsetTop = firstUnacknowledgedActivityElement.offsetTop + firstUnacknowledgedActivityBoundingBoxElement.offsetTop;
            values.push(firstUnacknowledgedActivityOffsetTop - scrollTop - offsetHeight * (1 - patchedAutoScrollSnapOnPage) + patchedAutoScrollSnapOnPageOffset);
          }
        }
      }

      return Math.min.apply(Math, values);
    }

    return Infinity;
  }, [activityElementMapRef, activityKeysRef, lastAcknowledgedActivityKeyRef, styleOptionsRef]);
};

var BasicTranscript = function BasicTranscript(_ref14) {
  var className = _ref14.className;
  var activityElementMapRef = (0, _react.useRef)(new Map());
  var containerRef = (0, _react.useRef)();
  var scroller = useScroller(activityElementMapRef);
  return /*#__PURE__*/_react.default.createElement(_TranscriptFocusComposer.default, {
    containerRef: containerRef
  }, /*#__PURE__*/_react.default.createElement(_reactScrollToBottom.Composer, {
    scroller: scroller
  }, /*#__PURE__*/_react.default.createElement(_KeyboardHelp.default, null), /*#__PURE__*/_react.default.createElement(InternalTranscript, {
    activityElementMapRef: activityElementMapRef,
    className: className,
    ref: containerRef
  })));
};

BasicTranscript.defaultProps = {
  className: ''
};
BasicTranscript.propTypes = {
  className: _propTypes.default.string
};
var _default = BasicTranscript;
exports.default = _default;
