// botframework-webchat-component/lib/BasicWebChat.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _botframeworkWebchatApi = require("botframework-webchat-api");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Surface = _interopRequireDefault(require("./Utils/AccessKeySink/Surface"));

var _BasicConnectivityStatus = _interopRequireDefault(require("./BasicConnectivityStatus"));

var _BasicSendBox = _interopRequireDefault(require("./BasicSendBox"));

var _BasicToaster = _interopRequireDefault(require("./BasicToaster"));

var _BasicTranscript = _interopRequireDefault(require("./BasicTranscript"));

var _useStyleSet3 = _interopRequireDefault(require("./hooks/useStyleSet"));

var _useStyleToEmotionObject = _interopRequireDefault(require("./hooks/internal/useStyleToEmotionObject"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var useStyleOptions = _botframeworkWebchatApi.hooks.useStyleOptions;
var ROOT_STYLE = {
  display: 'flex',
  flexDirection: 'column'
};
var CONNECTIVITY_STATUS_STYLE = {
  flexShrink: 0
};
var SEND_BOX_CSS = {
  flexShrink: 0
};
var TOASTER_STYLE = {
  flexShrink: 0
};
var TRANSCRIPT_STYLE = {
  flex: 1
}; // Subset of landmark roles: https://w3.org/TR/wai-aria/#landmark_roles

var ARIA_LANDMARK_ROLES = ['complementary', 'contentinfo', 'form', 'main', 'region'];

var BasicWebChat = function BasicWebChat(_ref) {
  var className = _ref.className,
      role = _ref.role;

  var _useStyleSet = (0, _useStyleSet3.default)(),
      _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
      rootStyleSet = _useStyleSet2[0].root;

  var _useStyleOptions = useStyleOptions(),
      _useStyleOptions2 = _slicedToArray(_useStyleOptions, 1),
      options = _useStyleOptions2[0];

  var styleToEmotionObject = (0, _useStyleToEmotionObject.default)();
  var connectivityStatusClassName = styleToEmotionObject(CONNECTIVITY_STATUS_STYLE) + '';
  var rootClassName = styleToEmotionObject(ROOT_STYLE) + '';
  var sendBoxClassName = styleToEmotionObject(SEND_BOX_CSS) + '';
  var toasterClassName = styleToEmotionObject(TOASTER_STYLE) + '';
  var transcriptClassName = styleToEmotionObject(TRANSCRIPT_STYLE) + ''; // Fallback to "complementary" if specified is not a valid landmark role.

  if (!ARIA_LANDMARK_ROLES.includes(role)) {
    role = 'complementary';
  }

  return /*#__PURE__*/_react.default.createElement(_Surface.default, {
    className: (0, _classnames.default)(rootClassName, rootStyleSet + '', (className || '') + ''),
    role: role
  }, !options.hideToaster && /*#__PURE__*/_react.default.createElement(_BasicToaster.default, {
    className: toasterClassName
  }), /*#__PURE__*/_react.default.createElement(_BasicTranscript.default, {
    className: transcriptClassName
  }), /*#__PURE__*/_react.default.createElement(_BasicConnectivityStatus.default, {
    className: connectivityStatusClassName
  }), !options.hideSendBox && /*#__PURE__*/_react.default.createElement(_BasicSendBox.default, {
    className: sendBoxClassName
  }));
};

BasicWebChat.defaultProps = {
  className: '',
  role: 'complementary'
};
BasicWebChat.propTypes = {
  className: _propTypes.default.string,
  // Ignoring deficiencies with TypeScript/PropTypes inference.
  // @ts-ignore
  role: _propTypes.default.oneOf(ARIA_LANDMARK_ROLES)
};
var _default = BasicWebChat;
exports.default = _default;
