// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/default/defaultProps/defaultProactiveChatPaneControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultProactiveChatPaneControlProps = void 0;

var _Constants = require("../../../../../common/Constants");

const defaultProactiveChatPaneControlProps = {
  id: _Constants.Ids.DefaultProactiveChatPaneId,
  dir: "ltr",
  hideProactiveChatPane: false,
  proactiveChatPaneAriaLabel: _Constants.AriaLabels.ProactiveChatPane,
  hideTitle: false,
  titleText: _Constants.Texts.ProactiveChatPaneTitleText,
  hideSubtitle: false,
  subtitleText: _Constants.Texts.ProactiveChatPaneSubtitleText,
  hideCloseButton: false,
  closeButtonProps: {
    type: _Constants.ButtonTypes.Icon,
    iconName: _Constants.IconNames.ChromeClose,
    hideButtonTitle: true
  },
  isBodyContainerHorizantal: false,
  hideBodyTitle: false,
  bodyTitleText: _Constants.Texts.ProactiveChatPaneBodyTitleText,
  hideStartButton: false,
  startButtonText: _Constants.Texts.ProactiveChatPaneStartButtonText,
  startButtonAriaLabel: _Constants.AriaLabels.ChatNow,
  onClose: function () {
    console.log("on close");
  },
  onStart: function () {
    console.log("on start");
  }
};
exports.defaultProactiveChatPaneControlProps = defaultProactiveChatPaneControlProps;
