// botframework-webchat-component/lib/Middleware/Avatar/createCoreMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultAvatar = void 0;
exports.default = createCoreAvatarMiddleware;

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ImageAvatar = _interopRequireDefault(require("../../Avatar/ImageAvatar"));

var _InitialsAvatar = _interopRequireDefault(require("../../Avatar/InitialsAvatar"));

var _useStyleSet3 = _interopRequireDefault(require("../../hooks/useStyleSet"));

var _useStyleToEmotionObject = _interopRequireDefault(require("../../hooks/internal/useStyleToEmotionObject"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var ROOT_STYLE = {
  overflow: 'hidden',
  position: 'relative',
  '> *': {
    left: 0,
    position: 'absolute',
    top: 0
  }
};

var DefaultAvatar = function DefaultAvatar(_ref) {
  var ariaHidden = _ref['aria-hidden'],
      className = _ref.className,
      fromUser = _ref.fromUser;

  var _useStyleSet = (0, _useStyleSet3.default)(),
      _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
      avatarStyleSet = _useStyleSet2[0].avatar;

  var rootClassName = (0, _useStyleToEmotionObject.default)()(ROOT_STYLE) + '';
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-hidden": ariaHidden,
    className: (0, _classnames.default)('webchat__defaultAvatar', {
      'webchat__defaultAvatar--fromUser': fromUser
    }, rootClassName, avatarStyleSet + '', (className || '') + '')
  }, /*#__PURE__*/_react.default.createElement(_InitialsAvatar.default, {
    fromUser: fromUser
  }), /*#__PURE__*/_react.default.createElement(_ImageAvatar.default, {
    fromUser: fromUser
  }));
};

exports.DefaultAvatar = DefaultAvatar;
DefaultAvatar.defaultProps = {
  'aria-hidden': true,
  className: ''
};
DefaultAvatar.propTypes = {
  'aria-hidden': _propTypes.default.bool,
  className: _propTypes.default.string,
  fromUser: _propTypes.default.bool.isRequired
};

function createCoreAvatarMiddleware() {
  return [function () {
    return function () {
      return function (_ref2) {
        var fromUser = _ref2.fromUser,
            styleOptions = _ref2.styleOptions;
        var botAvatarImage = styleOptions.botAvatarImage,
            botAvatarInitials = styleOptions.botAvatarInitials,
            userAvatarImage = styleOptions.userAvatarImage,
            userAvatarInitials = styleOptions.userAvatarInitials;

        if (fromUser ? userAvatarImage || userAvatarInitials : botAvatarImage || botAvatarInitials) {
          return function () {
            return /*#__PURE__*/_react.default.createElement(DefaultAvatar, {
              fromUser: fromUser
            });
          };
        }

        return false;
      };
    };
  }];
}
