// @microsoft/omnichannel-chat-sdk/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultLocaleString = exports.defaultLocaleId = exports.getLocaleIdFromString = exports.getLocaleStringFromId = exports.isCustomerMessage = exports.isSystemMessage = exports.PersonType = exports.MessageType = exports.DeliveryMode = exports.MessageContentType = exports.uuidv4 = exports.OCSDKProvider = exports.IC3SDKProvider = exports.OmnichannelChatSDK = void 0;
var ocsdk_1 = require("@microsoft/ocsdk");
Object.defineProperty(exports, "OCSDKProvider", { enumerable: true, get: function () { return ocsdk_1.SDKProvider; } });
Object.defineProperty(exports, "uuidv4", { enumerable: true, get: function () { return ocsdk_1.uuidv4; } });
var locale_1 = require("./utils/locale");
Object.defineProperty(exports, "defaultLocaleId", { enumerable: true, get: function () { return locale_1.defaultLocaleId; } });
Object.defineProperty(exports, "defaultLocaleString", { enumerable: true, get: function () { return locale_1.defaultLocaleString; } });
Object.defineProperty(exports, "getLocaleIdFromString", { enumerable: true, get: function () { return locale_1.getLocaleIdFromString; } });
Object.defineProperty(exports, "getLocaleStringFromId", { enumerable: true, get: function () { return locale_1.getLocaleStringFromId; } });
var utilities_1 = require("./utils/utilities");
Object.defineProperty(exports, "isCustomerMessage", { enumerable: true, get: function () { return utilities_1.isCustomerMessage; } });
Object.defineProperty(exports, "isSystemMessage", { enumerable: true, get: function () { return utilities_1.isSystemMessage; } });
var DeliveryMode_1 = require("@microsoft/omnichannel-ic3core/lib/model/DeliveryMode");
exports.DeliveryMode = DeliveryMode_1.default;
var omnichannel_ic3core_1 = require("@microsoft/omnichannel-ic3core");
Object.defineProperty(exports, "IC3SDKProvider", { enumerable: true, get: function () { return omnichannel_ic3core_1.SDKProvider; } });
var MessageContentType_1 = require("@microsoft/omnichannel-ic3core/lib/model/MessageContentType");
exports.MessageContentType = MessageContentType_1.default;
var MessageType_1 = require("@microsoft/omnichannel-ic3core/lib/model/MessageType");
exports.MessageType = MessageType_1.default;
var OmnichannelChatSDK_1 = require("./OmnichannelChatSDK");
exports.OmnichannelChatSDK = OmnichannelChatSDK_1.default;
var PersonType_1 = require("@microsoft/omnichannel-ic3core/lib/model/PersonType");
exports.PersonType = PersonType_1.default;
exports.default = {
    OmnichannelChatSDK: OmnichannelChatSDK_1.default,
    ic3sdk: { IC3SDKProvider: omnichannel_ic3core_1.SDKProvider },
    ocsdk: { OCSDKProvider: ocsdk_1.SDKProvider, uuidv4: ocsdk_1.uuidv4 }
};
