// botframework-webchat/lib/adaptiveCards/Attachment/private/renderAdaptiveCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderAdaptiveCard;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

/**
 * Returns `true`, if the object is a plain object and not a class, otherwise, `false`.
 */
function isPlainObject(obj) {
  return Object.getPrototypeOf(obj) === Object.prototype;
}

function renderAdaptiveCard(adaptiveCard, _ref) {
  var adaptiveCardsHostConfig = _ref.adaptiveCardsHostConfig,
      _ref$adaptiveCardsPac = _ref.adaptiveCardsPackage,
      GlobalSettings = _ref$adaptiveCardsPac.GlobalSettings,
      HostConfig = _ref$adaptiveCardsPac.HostConfig,
      renderMarkdownAsHTML = _ref.renderMarkdownAsHTML,
      setTabIndexAtCardRoot = _ref.setTabIndexAtCardRoot;

  // Currently, the only way to set the Markdown engine is to set it thru static member of AdaptiveCard class
  // TODO: [P3] Checks if we could make the "renderMarkdownAsHTML" per card
  //       This could be limitations from Adaptive Cards package (not supported as of 1.2.5)
  //       Because there could be timing difference between .parse and .render, we could be using wrong Markdown engine
  // "onProcessMarkdown" is a static function but we are trying to scope it to the current object instead.
  // eslint-disable-next-line dot-notation
  adaptiveCard.constructor['onProcessMarkdown'] = function (text, result) {
    if (renderMarkdownAsHTML) {
      result.outputHtml = renderMarkdownAsHTML(text);
      result.didProcess = true;
    }
  };

  if (adaptiveCardsHostConfig) {
    adaptiveCard.hostConfig = isPlainObject(adaptiveCardsHostConfig) ? new HostConfig(adaptiveCardsHostConfig) : adaptiveCardsHostConfig;
  } // For accessibility issue #1340, `tabindex="0"` must not be set for the root container if it is not interactive.


  GlobalSettings.setTabIndexAtCardRoot = setTabIndexAtCardRoot;

  var _adaptiveCard$validat = adaptiveCard.validateProperties(),
      validationEvents = _adaptiveCard$validat.validationEvents;

  if (validationEvents.length) {
    return {
      errors: validationEvents.reduce(function (items, _ref2) {
        var message = _ref2.message;
        return [].concat((0, _toConsumableArray2.default)(items), [new Error(message)]);
      }, [])
    };
  }

  var element;

  try {
    element = adaptiveCard.render();
  } catch (error) {
    return {
      errors: [error]
    };
  }

  if (!element) {
    return {
      errors: [new Error('Adaptive Card rendered as empty element')]
    };
  }

  return {
    element: element
  };
}
