// @microsoft/omnichannel-chat-components/lib/cjs/components/header/common/defaultProps/azureHeaderControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.azureHeaderControlProps = void 0;

var _Icons = require("../../../../assets/Icons");

var _Constants = require("../../../../common/Constants");

const azureHeaderControlProps = {
  id: _Constants.Ids.DefaultHeaderId,
  hideIcon: false,
  hideTitle: true,
  hideCloseButton: false,
  hideMinimizeButton: true,
  onMinimizeClick: function () {
    console.log("minimize clicked");
  },
  onCloseClick: function () {
    console.log("close clicked");
  },
  middleGroup: {
    children: []
  },
  leftGroup: {
    children: []
  },
  rightGroup: {
    children: []
  },
  minimizeButtonProps: {
    id: "oc-lcw-header-minimizebutton",
    type: _Constants.ButtonTypes.Icon
  },
  closeButtonProps: {
    id: "oc-lcw-header-closebutton",
    type: _Constants.ButtonTypes.Text,
    text: "End Chat"
  },
  headerIconProps: {
    id: _Constants.Ids.HeaderIconId,
    src: _Icons.LegacyChatIconBase64,
    alt: _Constants.Texts.HeaderIcon
  },
  headerTitleProps: {
    id: _Constants.Ids.HeaderTitleId,
    text: "Azure Chat Support"
  }
};
exports.azureHeaderControlProps = azureHeaderControlProps;
