// @microsoft/omnichannel-chat-components/lib/cjs/components/header/common/defaultProps/customHeaderControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.customHeaderControlProps = void 0;

var _Icons = require("../../../../assets/Icons");

var _Constants = require("../../../../common/Constants");

const customHeaderControlProps = {
  id: _Constants.Ids.DefaultHeaderId,
  hideIcon: false,
  hideTitle: false,
  hideCloseButton: false,
  hideMinimizeButton: false,
  onMinimizeClick: function () {
    console.log("minimize clicked");
  },
  onCloseClick: function () {
    console.log("close clicked");
  },
  middleGroup: {
    children: []
  },
  leftGroup: {
    children: []
  },
  rightGroup: {
    children: []
  },
  minimizeButtonProps: {
    id: "oc-lcw-header-minimizebutton",
    type: _Constants.ButtonTypes.Icon,
    imageIconProps: {
      src: _Icons.MinimizeChatButtonIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    hideButtonTitle: true
  },
  closeButtonProps: {
    id: "oc-lcw-header-closebutton",
    type: _Constants.ButtonTypes.Icon,
    imageIconProps: {
      src: _Icons.CloseChatButtonIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    hideButtonTitle: true
  },
  headerIconProps: {
    id: _Constants.Ids.HeaderIconId,
    src: _Icons.LegacyChatIconBase64,
    alt: _Constants.Texts.HeaderIcon
  },
  headerTitleProps: {
    id: _Constants.Ids.HeaderTitleId,
    text: _Constants.Texts.HeaderTitle
  }
};
exports.customHeaderControlProps = customHeaderControlProps;
