// @microsoft/omnichannel-chat-components/lib/cjs/components/header/common/defaultProps/defaultHeaderControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultHeaderControlProps = void 0;

var _Icons = require("../../../../assets/Icons");

var _Constants = require("../../../../common/Constants");

const defaultHeaderControlProps = {
  id: _Constants.Ids.DefaultHeaderId,
  hideIcon: false,
  hideTitle: false,
  hideCloseButton: false,
  hideMinimizeButton: false,
  onMinimizeClick: function () {
    console.log("minimize clicked");
  },
  onCloseClick: function () {
    console.log("close clicked");
  },
  middleGroup: {
    children: []
  },
  leftGroup: {
    children: []
  },
  rightGroup: {
    children: []
  },
  minimizeButtonProps: {
    id: _Constants.Ids.MinimizeButtonId,
    type: _Constants.ButtonTypes.Icon,
    iconName: _Constants.IconNames.ChromeMinimize,
    hideButtonTitle: true
  },
  closeButtonProps: {
    id: _Constants.Ids.CloseButtonId,
    type: _Constants.ButtonTypes.Icon,
    iconName: _Constants.IconNames.ChromeClose,
    hideButtonTitle: true
  },
  headerIconProps: {
    id: _Constants.Ids.HeaderIconId,
    src: _Icons.ModernChatIconBase64,
    alt: _Constants.Texts.HeaderIcon
  },
  headerTitleProps: {
    id: _Constants.Ids.HeaderTitleId,
    text: _Constants.Texts.HeaderTitle
  }
};
exports.defaultHeaderControlProps = defaultHeaderControlProps;
