// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/presetThree/presetThreeStyles/presetThreeProactiveChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetThreeProactiveChatPaneStyles = void 0;

var _presetThreeProactiveChatPaneBodyContainerStyles = require("./presetThreeProactiveChatPaneBodyContainerStyles");

var _presetThreeProactiveChatPaneGeneralStyles = require("./presetThreeProactiveChatPaneGeneralStyles");

var _presetThreeProactiveChatPaneCloseButtonStyles = require("./presetThreeProactiveChatPaneCloseButtonStyles");

var _presetThreeProactiveChatPaneHeaderContainerStyles = require("./presetThreeProactiveChatPaneHeaderContainerStyles");

var _presetThreeProactiveChatPaneStartButtonStyles = require("./presetThreeProactiveChatPaneStartButtonStyles");

var _presetThreeProactiveChatPaneStartButtonHoveredStyles = require("./presetThreeProactiveChatPaneStartButtonHoveredStyles");

var _presetThreeProactiveChatPaneCloseButtonHoveredStyles = require("./presetThreeProactiveChatPaneCloseButtonHoveredStyles");

const presetThreeProactiveChatPaneStyles = {
  generalStyleProps: _presetThreeProactiveChatPaneGeneralStyles.presetThreeProactiveChatPaneGeneralStyles,
  headerContainerStyleProps: _presetThreeProactiveChatPaneHeaderContainerStyles.presetThreeProactiveChatPaneHeaderContainerStyles,
  closeButtonStyleProps: _presetThreeProactiveChatPaneCloseButtonStyles.presetThreeProactiveChatPaneCloseButtonStyles,
  closeButtonHoveredStyleProps: _presetThreeProactiveChatPaneCloseButtonHoveredStyles.presetThreeProactiveChatPaneCloseButtonHoveredStyles,
  bodyContainerStyleProps: _presetThreeProactiveChatPaneBodyContainerStyles.presetThreeProactiveChatPaneBodyContainerStyles,
  startButtonStyleProps: _presetThreeProactiveChatPaneStartButtonStyles.presetThreeProactiveChatPaneStartButtonStyles,
  startButtonHoveredStyleProps: _presetThreeProactiveChatPaneStartButtonHoveredStyles.presetThreeProactiveChatPaneStartButtonHoveredStyles
};
exports.presetThreeProactiveChatPaneStyles = presetThreeProactiveChatPaneStyles;
