// botframework-webchat-component/lib/Attachment/ImageAttachment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ImageContent = _interopRequireDefault(require("./ImageContent"));

var _readDataURIToBlob = _interopRequireDefault(require("../Utils/readDataURIToBlob"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ImageAttachment = function ImageAttachment(_ref) {
  var attachment = _ref.attachment;
  var imageURL = attachment.thumbnailUrl || attachment.contentUrl; // To support Content Security Policy, data URI cannot be used.
  // We need to parse the data URI into a blob: URL.

  var blob = (0, _readDataURIToBlob.default)(imageURL);

  if (blob) {
    imageURL = URL.createObjectURL(blob);
  }

  return /*#__PURE__*/_react.default.createElement(_ImageContent.default, {
    alt: attachment.name,
    src: imageURL
  });
};

ImageAttachment.propTypes = {
  // Either attachment.contentUrl or attachment.thumbnailUrl must be specified.
  attachment: _propTypes.default.oneOfType([_propTypes.default.shape({
    contentUrl: _propTypes.default.string.isRequired,
    name: _propTypes.default.string,
    thumbnailUrl: _propTypes.default.string
  }), _propTypes.default.shape({
    contentUrl: _propTypes.default.string,
    name: _propTypes.default.string,
    thumbnailUrl: _propTypes.default.string.isRequired
  })]).isRequired
};
var _default = ImageAttachment;
exports.default = _default;
