// @microsoft/omnichannel-chat-components/lib/cjs/components/loadingpane/common/defaultProps/defaultStyles/defaultLoadingPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultLoadingPaneStyles = void 0;

var _defaultLoadingPaneGeneralStyles = require("./defaultLoadingPaneGeneralStyles");

var _defaultLoadingPaneIconStyles = require("./defaultLoadingPaneIconStyles");

var _defaultLoadingPaneIconImageProps = require("./defaultLoadingPaneIconImageProps");

var _defaultLoadingPaneSubtitleStyles = require("./defaultLoadingPaneSubtitleStyles");

var _defaultLoadingPaneTitleStyles = require("./defaultLoadingPaneTitleStyles");

var _defaultLoadingPaneSpinnerStyles = require("./defaultLoadingPaneSpinnerStyles");

var _defaultLoadingPaneSpinnerTextStyles = require("./defaultLoadingPaneSpinnerTextStyles");

const defaultLoadingPaneStyles = {
  generalStyleProps: _defaultLoadingPaneGeneralStyles.defaultLoadingPaneGeneralStyles,
  titleStyleProps: _defaultLoadingPaneTitleStyles.defaultLoadingPaneTitleStyles,
  subtitleStyleProps: _defaultLoadingPaneSubtitleStyles.defaultLoadingPaneSubtitleStyles,
  iconStyleProps: _defaultLoadingPaneIconStyles.defaultLoadingPaneIconStyles,
  iconImageProps: _defaultLoadingPaneIconImageProps.defaultLoadingPaneIconImageProps,
  spinnerStyleProps: _defaultLoadingPaneSpinnerStyles.defaultLoadingPaneSpinnerStyles,
  spinnerTextStyleProps: _defaultLoadingPaneSpinnerTextStyles.defaultLoadingPaneSpinnerTextStyles
};
exports.defaultLoadingPaneStyles = defaultLoadingPaneStyles;
