// botframework-webchat-component/lib/hooks/internal/BypassSpeechSynthesisPonyfill.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.speechSynthesis = exports.SpeechSynthesisVoice = exports.SpeechSynthesisUtterance = exports.SpeechSynthesisEvent = void 0;

var _es = _interopRequireWildcard(require("event-target-shim/es5"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var SpeechSynthesisEvent = /*#__PURE__*/function (_Event) {
  _inherits(SpeechSynthesisEvent, _Event);

  var _super = _createSuper(SpeechSynthesisEvent);

  function SpeechSynthesisEvent(type, utterance) {
    var _this;

    _classCallCheck(this, SpeechSynthesisEvent);

    _this = _super.call(this, type);
    _this._utterance = utterance;
    return _this;
  }

  _createClass(SpeechSynthesisEvent, [{
    key: "charIndex",
    get: function get() {
      return 0;
    }
  }, {
    key: "elapsedTime",
    get: function get() {
      return 0;
    }
  }, {
    key: "name",
    get: function get() {// It is expected to return `undefined`, keeping the function empty to reduce footprint.
    }
  }, {
    key: "utterance",
    get: function get() {
      return this._utterance;
    }
  }]);

  return SpeechSynthesisEvent;
}(_es.Event);

exports.SpeechSynthesisEvent = SpeechSynthesisEvent;

var SpeechSynthesisUtterance = /*#__PURE__*/function (_EventTarget) {
  _inherits(SpeechSynthesisUtterance, _EventTarget);

  var _super2 = _createSuper(SpeechSynthesisUtterance);

  function SpeechSynthesisUtterance(text) {
    var _this2;

    _classCallCheck(this, SpeechSynthesisUtterance);

    _this2 = _super2.call(this);
    _this2._lang = 'en-US';
    _this2._pitch = 1;
    _this2._rate = 1;
    _this2._text = text;
    _this2._voice = null;
    _this2._volume = 1;
    return _this2;
  }

  _createClass(SpeechSynthesisUtterance, [{
    key: "lang",
    get: function get() {
      return this._lang;
    },
    set: function set(value) {
      this._lang = value;
    }
  }, {
    key: "pitch",
    get: function get() {
      return this._pitch;
    },
    set: function set(value) {
      this._pitch = value;
    }
  }, {
    key: "rate",
    get: function get() {
      return this._rate;
    },
    set: function set(value) {
      this._rate = value;
    }
  }, {
    key: "text",
    get: function get() {
      return this._text;
    },
    set: function set(value) {
      this._text = value;
    }
  }, {
    key: "voice",
    get: function get() {
      return this._voice;
    },
    set: function set(value) {
      this._voice = value;
    }
  }, {
    key: "volume",
    get: function get() {
      return this._volume;
    },
    set: function set(value) {
      this._volume = value;
    }
  }, {
    key: "onboundary",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'boundary');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'boundary', value);
    }
  }, {
    key: "onend",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'end');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'end', value);
    }
  }, {
    key: "onerror",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'error');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'error', value);
    }
  }, {
    key: "onmark",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'mark');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'mark', value);
    }
  }, {
    key: "onpause",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'pause');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'pause', value);
    }
  }, {
    key: "onresume",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'resume');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'resume', value);
    }
  }, {
    key: "onstart",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'start');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'start', value);
    }
  }]);

  return SpeechSynthesisUtterance;
}(_es.default);

exports.SpeechSynthesisUtterance = SpeechSynthesisUtterance;

var SpeechSynthesisVoice = /*#__PURE__*/function () {
  function SpeechSynthesisVoice() {
    _classCallCheck(this, SpeechSynthesisVoice);
  }

  _createClass(SpeechSynthesisVoice, [{
    key: "default",
    get: function get() {
      return true;
    }
  }, {
    key: "lang",
    get: function get() {
      return 'en-US';
    }
  }, {
    key: "localService",
    get: function get() {
      return true;
    }
  }, {
    key: "name",
    get: function get() {
      return 'English (US)';
    }
  }, {
    key: "voiceURI",
    get: function get() {
      return 'English (US)';
    }
  }]);

  return SpeechSynthesisVoice;
}();

exports.SpeechSynthesisVoice = SpeechSynthesisVoice;

var SpeechSynthesis = /*#__PURE__*/function (_EventTarget2) {
  _inherits(SpeechSynthesis, _EventTarget2);

  var _super3 = _createSuper(SpeechSynthesis);

  function SpeechSynthesis() {
    _classCallCheck(this, SpeechSynthesis);

    return _super3.apply(this, arguments);
  }

  _createClass(SpeechSynthesis, [{
    key: "paused",
    get: function get() {
      return false;
    }
  }, {
    key: "pending",
    get: function get() {
      return false;
    }
  }, {
    key: "speaking",
    get: function get() {
      return false;
    }
  }, {
    key: "cancel",
    value: function cancel() {}
  }, {
    key: "getVoices",
    value: function getVoices() {
      return [new SpeechSynthesisVoice()];
    }
  }, {
    key: "pause",
    value: function pause() {
      throw new Error('pause is not implemented.');
    }
  }, {
    key: "resume",
    value: function resume() {
      throw new Error('resume is not implemented.');
    }
  }, {
    key: "speak",
    value: function speak(utterance) {
      utterance.dispatchEvent(new SpeechSynthesisEvent('start', utterance));
      utterance.dispatchEvent(new SpeechSynthesisEvent('end', utterance));
    }
  }, {
    key: "onvoiceschanged",
    get: function get() {
      return (0, _es.getEventAttributeValue)(this, 'voiceschanged');
    },
    set: function set(value) {
      (0, _es.setEventAttributeValue)(this, 'voiceschanged', value);
    }
  }]);

  return SpeechSynthesis;
}(_es.default);

var speechSynthesis = new SpeechSynthesis();
exports.speechSynthesis = speechSynthesis;
