// @microsoft/omnichannel-chat-components/lib/cjs/components/reconnectchatpane/common/presetThree/presetThreeProps/presetThreeReconnectChatPaneControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetThreeReconnectChatPaneControlProps = void 0;

var _Constants = require("../../../../../common/Constants");

const presetThreeReconnectChatPaneControlProps = {
  id: "lcw-components-reconnect-chat-pane",
  dir: "ltr",
  hideReconnectChatPane: false,
  reconnectChatPaneAriaLabel: _Constants.AriaLabels.ReconnectChatPane,
  hideTitle: false,
  titleText: _Constants.Texts.ReconnectChatPaneTitleText,
  hideSubtitle: true,
  subtitleText: _Constants.Texts.ReconnectChatPaneSubtitleText,
  hideIcon: false,
  iconAriaLabel: _Constants.AriaLabels.ReconnectChatPaneIcon,
  isButtonGroupHorizontal: false,
  hideContinueChatButton: false,
  continueChatButtonText: _Constants.Texts.ReconnectChatPaneContinueChatButtonText,
  continueChatButtonAriaLabel: _Constants.AriaLabels.ContinueConversation,
  hideStartNewChatButton: true,
  startNewChatButtonText: _Constants.Texts.ReconnectChatPaneStartNewChatButtonText,
  startNewChatButtonAriaLabel: _Constants.AriaLabels.StartNewConversation,
  onContinueChat: function () {
    console.log("on continue conversation");
  },
  onStartNewChat: function () {
    console.log("on start new conversation");
  }
};
exports.presetThreeReconnectChatPaneControlProps = presetThreeReconnectChatPaneControlProps;
