// react-film/lib/browser.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.safari = exports.internetExplorer = exports.firefox = exports.edgeUWP = exports.edgeAnaheim = exports.chromium = exports.chrome = void 0;
var _navigator = navigator,
  userAgent = _navigator.userAgent;
var chromium = !/Edge\//.test(userAgent) && /Chrome\//.test(userAgent);
exports.chromium = chromium;
var edgeAnaheim = /Edg\//.test(userAgent);
exports.edgeAnaheim = edgeAnaheim;
var edgeUWP = /Edge\//.test(userAgent);
exports.edgeUWP = edgeUWP;
var firefox = /Firefox\//.test(userAgent);
exports.firefox = firefox;
var internetExplorer = /Trident\/7(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])0/.test(userAgent);
exports.internetExplorer = internetExplorer;
var chrome = chromium && !edgeAnaheim;
exports.chrome = chrome;
var safari = !(chrome || edgeUWP || internetExplorer || firefox);
exports.safari = safari;
