// react-film/lib/Dot.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _useIndex3 = _interopRequireDefault(require("./hooks/useIndex"));
var _useScrollTo = _interopRequireDefault(require("./hooks/useScrollTo"));
var Dot = function Dot(_ref) {
  var ariaLabel = _ref['aria-label'],
    itemIndex = _ref.itemIndex;
  var _useIndex = (0, _useIndex3.default)(),
    _useIndex2 = (0, _slicedToArray2.default)(_useIndex, 1),
    index = _useIndex2[0];
  var scrollTo = (0, _useScrollTo.default)();
  var checked = index === itemIndex;
  var handleChange = (0, _react.useCallback)(function () {
    return scrollTo(function () {
      return itemIndex;
    });
  }, [itemIndex, scrollTo]);
  var handleKeyPress = (0, _react.useCallback)(function (event) {
    var keyCode = event.keyCode;

    // 13 is ENTER, 32 is SPACEBAR.
    // eslint-disable-next-line no-magic-numbers
    if (keyCode === 13 || keyCode === 32) {
      event.preventDefault();
      event.stopPropagation();
      handleChange();
    }
  }, [handleChange]);
  return /*#__PURE__*/_react.default.createElement("li", {
    className: "react-film__dot"
  }, /*#__PURE__*/_react.default.createElement("input", {
    "aria-label": ariaLabel,
    "aria-pressed": checked,
    checked: checked,
    className: "react-film__dot__input",
    onChange: handleChange,
    onKeyPress: handleKeyPress,
    role: "button",
    type: "checkbox"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__dot__handle"
  }));
};
Dot.defaultProps = {
  'aria-label': undefined
};
Dot.propTypes = {
  'aria-label': _propTypes.default.string,
  itemIndex: _propTypes.default.number.isRequired
};
var _default = Dot;
exports.default = _default;
