// react-film/lib/Flipper.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));
var _classnames = _interopRequireDefault(require("classnames"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _useDir3 = _interopRequireDefault(require("./hooks/useDir"));
var _useScrollBarPercentage = _interopRequireDefault(require("./hooks/useScrollBarPercentage"));
var _useScrollOneLeft = _interopRequireDefault(require("./hooks/useScrollOneLeft"));
var _useScrollOneRight = _interopRequireDefault(require("./hooks/useScrollOneRight"));
var Flipper = function Flipper(_ref) {
  var ariaLabel = _ref['aria-label'],
    blurFocusOnClick = _ref.blurFocusOnClick,
    children = _ref.children,
    mode = _ref.mode;
  var _useDir = (0, _useDir3.default)(),
    _useDir2 = (0, _slicedToArray2.default)(_useDir, 1),
    dir = _useDir2[0];
  var _useScrollBarPercenta = (0, _useScrollBarPercentage.default)(),
    _useScrollBarPercenta2 = (0, _slicedToArray2.default)(_useScrollBarPercenta, 1),
    scrollBarPercentage = _useScrollBarPercenta2[0];
  var buttonRef = (0, _react.useRef)();
  var left = mode === 'left';
  var scrollOneLeft = (0, _useScrollOneLeft.default)();
  var scrollOneRight = (0, _useScrollOneRight.default)();
  var handleClick = (0, _react.useCallback)(function () {
    left ? scrollOneLeft() : scrollOneRight();
    blurFocusOnClick && buttonRef.current && buttonRef.current.blur();
  }, [blurFocusOnClick, buttonRef, left, scrollOneLeft, scrollOneRight]);
  var handleKeyDown = (0, _react.useCallback)(function (event) {
    var key = event.key;
    if (key === 'Enter' || key === ' ') {
      event.preventDefault();
      left ? scrollOneLeft() : scrollOneRight();
    }
  }, [left, scrollOneLeft, scrollOneRight]);
  var hide;
  if (dir === 'rtl') {
    if (left) {
      hide = scrollBarPercentage === '100%' || scrollBarPercentage === '-100%';
    } else {
      hide = scrollBarPercentage === '0%';
    }
  } else {
    if (left) {
      hide = scrollBarPercentage === '0%';
    } else {
      hide = scrollBarPercentage === '100%';
    }
  }
  return /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": ariaLabel || (left ? 'left' : 'right'),
    className: (0, _classnames.default)('react-film__flipper', 'react-film__main__overlay', {
      'react-film__flipper--left': left,
      'react-film__flipper--right': !left
    }),
    onClick: handleClick,
    onKeyDown: handleKeyDown,
    ref: buttonRef,
    type: "button"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('react-film__flipper__slider', 'react-film__main__slider', {
      'react-film__main__slider--hide': hide,
      'react-film__main__slider--left': left,
      'react-film__main__slider--right': !left
    })
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__flipper__body"
  }, children)));
};
Flipper.defaultProps = {
  'aria-label': undefined,
  blurFocusOnClick: false,
  children: undefined,
  mode: 'left'
};
Flipper.propTypes = {
  'aria-label': _propTypes.default.string,
  blurFocusOnClick: _propTypes.default.bool,
  children: _propTypes.default.any,
  mode: _propTypes.default.oneOf(['left', 'right'])
};
var _default = Flipper;
exports.default = _default;
