// react-film/lib/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
_Object$defineProperty(exports, "BasicFilm", {
  enumerable: true,
  get: function get() {
    return _BasicFilm.default;
  }
});
_Object$defineProperty(exports, "Composer", {
  enumerable: true,
  get: function get() {
    return _Composer.default;
  }
});
_Object$defineProperty(exports, "Context", {
  enumerable: true,
  get: function get() {
    return _LegacyContext.default;
  }
});
_Object$defineProperty(exports, "Dots", {
  enumerable: true,
  get: function get() {
    return _Dots.default;
  }
});
_Object$defineProperty(exports, "FilmStrip", {
  enumerable: true,
  get: function get() {
    return _DeprecatedFilmStrip.default;
  }
});
_Object$defineProperty(exports, "Filmstrip", {
  enumerable: true,
  get: function get() {
    return _Filmstrip.default;
  }
});
_Object$defineProperty(exports, "Flipper", {
  enumerable: true,
  get: function get() {
    return _Flipper.default;
  }
});
_Object$defineProperty(exports, "ScrollBar", {
  enumerable: true,
  get: function get() {
    return _ScrollBar.default;
  }
});
_Object$defineProperty(exports, "createBasicStyleSet", {
  enumerable: true,
  get: function get() {
    return _createBasicStyleSet.default;
  }
});
exports.default = void 0;
_Object$defineProperty(exports, "useDir", {
  enumerable: true,
  get: function get() {
    return _useDir.default;
  }
});
_Object$defineProperty(exports, "useHeight", {
  enumerable: true,
  get: function get() {
    return _useHeight.default;
  }
});
_Object$defineProperty(exports, "useIndex", {
  enumerable: true,
  get: function get() {
    return _useIndex.default;
  }
});
_Object$defineProperty(exports, "useIndexFraction", {
  enumerable: true,
  get: function get() {
    return _useIndexFraction.default;
  }
});
_Object$defineProperty(exports, "useItemContainerCallbackRef", {
  enumerable: true,
  get: function get() {
    return _useItemContainerCallbackRef.default;
  }
});
_Object$defineProperty(exports, "useNumItems", {
  enumerable: true,
  get: function get() {
    return _useNumItems.default;
  }
});
_Object$defineProperty(exports, "useScrollBarPercentage", {
  enumerable: true,
  get: function get() {
    return _useScrollBarPercentage.default;
  }
});
_Object$defineProperty(exports, "useScrollBarWidth", {
  enumerable: true,
  get: function get() {
    return _useScrollBarWidth.default;
  }
});
_Object$defineProperty(exports, "useScrollOneLeft", {
  enumerable: true,
  get: function get() {
    return _useScrollOneLeft.default;
  }
});
_Object$defineProperty(exports, "useScrollOneRight", {
  enumerable: true,
  get: function get() {
    return _useScrollOneRight.default;
  }
});
_Object$defineProperty(exports, "useScrollTo", {
  enumerable: true,
  get: function get() {
    return _useScrollTo.default;
  }
});
_Object$defineProperty(exports, "useScrollableCallbackRef", {
  enumerable: true,
  get: function get() {
    return _useScrollableCallbackRef.default;
  }
});
_Object$defineProperty(exports, "useScrolling", {
  enumerable: true,
  get: function get() {
    return _useScrolling.default;
  }
});
_Object$defineProperty(exports, "useStyleOptions", {
  enumerable: true,
  get: function get() {
    return _useStyleOptions.default;
  }
});
_Object$defineProperty(exports, "useStyleSetClassNames", {
  enumerable: true,
  get: function get() {
    return _useStyleSetClassNames.default;
  }
});
var _BasicFilm = _interopRequireDefault(require("./BasicFilm"));
var _Composer = _interopRequireDefault(require("./Composer"));
var _LegacyContext = _interopRequireDefault(require("./LegacyContext"));
var _createBasicStyleSet = _interopRequireDefault(require("./createBasicStyleSet"));
var _DeprecatedFilmStrip = _interopRequireDefault(require("./DeprecatedFilmStrip"));
var _Dots = _interopRequireDefault(require("./Dots"));
var _Filmstrip = _interopRequireDefault(require("./Filmstrip"));
var _Flipper = _interopRequireDefault(require("./Flipper"));
var _ReactFilm = _interopRequireDefault(require("./ReactFilm"));
var _ScrollBar = _interopRequireDefault(require("./ScrollBar"));
var _useDir = _interopRequireDefault(require("./hooks/useDir"));
var _useHeight = _interopRequireDefault(require("./hooks/useHeight"));
var _useIndex = _interopRequireDefault(require("./hooks/useIndex"));
var _useIndexFraction = _interopRequireDefault(require("./hooks/useIndexFraction"));
var _useItemContainerCallbackRef = _interopRequireDefault(require("./hooks/useItemContainerCallbackRef"));
var _useNumItems = _interopRequireDefault(require("./hooks/useNumItems"));
var _useScrollableCallbackRef = _interopRequireDefault(require("./hooks/useScrollableCallbackRef"));
var _useScrollBarPercentage = _interopRequireDefault(require("./hooks/useScrollBarPercentage"));
var _useScrollBarWidth = _interopRequireDefault(require("./hooks/useScrollBarWidth"));
var _useScrolling = _interopRequireDefault(require("./hooks/useScrolling"));
var _useScrollOneLeft = _interopRequireDefault(require("./hooks/useScrollOneLeft"));
var _useScrollOneRight = _interopRequireDefault(require("./hooks/useScrollOneRight"));
var _useScrollTo = _interopRequireDefault(require("./hooks/useScrollTo"));
var _useStyleOptions = _interopRequireDefault(require("./hooks/useStyleOptions"));
var _useStyleSetClassNames = _interopRequireDefault(require("./hooks/useStyleSetClassNames"));
var _default = _ReactFilm.default;
exports.default = _default;
if (typeof document !== 'undefined' && document.head) {
  var meta = document.createElement('meta');
  meta.setAttribute('name', 'react-film');

  // eslint-disable-next-line no-undef
  meta.setAttribute('content', "version=".concat("3.1.1-main.df870ea"));
  document.head.appendChild(meta);
}
