// react-film/lib/ScrollBar.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parseFloat2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/parse-float"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/objectSpread2"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var browser = _interopRequireWildcard(require("./browser"));
var _useDir3 = _interopRequireDefault(require("./hooks/useDir"));
var _useScrollBarPercentage = _interopRequireDefault(require("./hooks/useScrollBarPercentage"));
var _useScrollBarWidth3 = _interopRequireDefault(require("./hooks/useScrollBarWidth"));
/* eslint no-magic-numbers: ["error", { "ignore": [1, 100] }] */

var ScrollBar = function ScrollBar() {
  var _useDir = (0, _useDir3.default)(),
    _useDir2 = (0, _slicedToArray2.default)(_useDir, 1),
    dir = _useDir2[0];
  var _useScrollBarPercenta = (0, _useScrollBarPercentage.default)(),
    _useScrollBarPercenta2 = (0, _slicedToArray2.default)(_useScrollBarPercenta, 1),
    scrollBarPercentage = _useScrollBarPercenta2[0];
  var _useScrollBarWidth = (0, _useScrollBarWidth3.default)(),
    _useScrollBarWidth2 = (0, _slicedToArray2.default)(_useScrollBarWidth, 1),
    scrollBarWidth = _useScrollBarWidth2[0];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__scroll-bar react-film__main__slider react-film__main__slider--bottom"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "react-film__scroll-bar__handle",
    style: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, dir === 'rtl' ? browser.edgeUWP || browser.internetExplorer ? {
      marginRight: "".concat((1 - (0, _parseFloat2.default)(scrollBarWidth) / 100) * (0, _parseFloat2.default)(scrollBarPercentage), "%")
    } : {
      marginRight: "".concat(((0, _parseFloat2.default)(scrollBarWidth) / 100 - 1) * (0, _parseFloat2.default)(scrollBarPercentage), "%")
    } : {
      marginLeft: "".concat((1 - (0, _parseFloat2.default)(scrollBarWidth) / 100) * (0, _parseFloat2.default)(scrollBarPercentage), "%")
    }), {}, {
      width: scrollBarWidth
    })
  }));
};
var _default = ScrollBar;
exports.default = _default;
