// @microsoft/omnichannel-chat-components/lib/esm/components/reconnectchatpane/common/presetThree/presetThreeProps/presetThreeReconnectChatPaneControlProps.js
import { AriaLabels, Texts } from "../../../../../common/Constants";
export const presetThreeReconnectChatPaneControlProps = {
  id: "lcw-components-reconnect-chat-pane",
  dir: "ltr",
  hideReconnectChatPane: false,
  reconnectChatPaneAriaLabel: AriaLabels.ReconnectChatPane,
  hideTitle: false,
  titleText: Texts.ReconnectChatPaneTitleText,
  hideSubtitle: true,
  subtitleText: Texts.ReconnectChatPaneSubtitleText,
  hideIcon: false,
  iconAriaLabel: AriaLabels.ReconnectChatPaneIcon,
  isButtonGroupHorizontal: false,
  hideContinueChatButton: false,
  continueChatButtonText: Texts.ReconnectChatPaneContinueChatButtonText,
  continueChatButtonAriaLabel: AriaLabels.ContinueConversation,
  hideStartNewChatButton: true,
  startNewChatButtonText: Texts.ReconnectChatPaneStartNewChatButtonText,
  startNewChatButtonAriaLabel: AriaLabels.StartNewConversation,
  onContinueChat: function () {
    console.log("on continue conversation");
  },
  onStartNewChat: function () {
    console.log("on start new conversation");
  }
};
