// @microsoft/omnichannel-chat-widget/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BroadcastService", {
  enumerable: true,
  get: function () {
    return _omnichannelChatComponents.BroadcastService;
  }
});
Object.defineProperty(exports, "ConversationState", {
  enumerable: true,
  get: function () {
    return _ConversationState.ConversationState;
  }
});
Object.defineProperty(exports, "LiveChatWidget", {
  enumerable: true,
  get: function () {
    return _LiveChatWidget.default;
  }
});
Object.defineProperty(exports, "decodeComponentString", {
  enumerable: true,
  get: function () {
    return _omnichannelChatComponents.decodeComponentString;
  }
});
Object.defineProperty(exports, "encodeComponentString", {
  enumerable: true,
  get: function () {
    return _omnichannelChatComponents.encodeComponentString;
  }
});
Object.defineProperty(exports, "getWidgetCacheId", {
  enumerable: true,
  get: function () {
    return _utils.getWidgetCacheId;
  }
});
Object.defineProperty(exports, "getWidgetEndChatEventName", {
  enumerable: true,
  get: function () {
    return _utils.getWidgetEndChatEventName;
  }
});
Object.defineProperty(exports, "useChatContextStore", {
  enumerable: true,
  get: function () {
    return _useChatContextStore.default;
  }
});
Object.defineProperty(exports, "useChatSDKStore", {
  enumerable: true,
  get: function () {
    return _useChatSDKStore.default;
  }
});
var _omnichannelChatComponents = require("@microsoft/omnichannel-chat-components");
var _useChatContextStore = _interopRequireDefault(require("./hooks/useChatContextStore"));
var _useChatSDKStore = _interopRequireDefault(require("./hooks/useChatSDKStore"));
var _utils = require("./common/utils");
var _ConversationState = require("./contexts/common/ConversationState");
var _LiveChatWidget = _interopRequireDefault(require("./components/livechatwidget/LiveChatWidget"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
