// @microsoft/omnichannel-chat-components/lib/cjs/components/outofofficehourspane/OOOHPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("@fluentui/react");

var _react2 = _interopRequireDefault(require("react"));

var _decodeComponentString = require("../../common/decodeComponentString");

var _defaultOOOHPaneControlProps = require("./common/defaultProps/defaultOOOHPaneControlProps");

var _defaultOOOHPaneGeneralStyles = require("./common/defaultProps/defaultStyles/defaultOOOHPaneGeneralStyles");

var _defaultOOOHPaneTitleStyles = require("./common/defaultProps/defaultStyles/defaultOOOHPaneTitleStyles");

var _utils = require("../../common/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function OOOHPane(props) {
  var _props$controlProps, _props$styleProps, _props$styleProps2, _props$controlProps2, _props$controlProps3, _props$controlProps4, _props$controlProps5, _props$styleProps3, _props$styleProps3$cl, _props$controlProps6, _props$controlProps7, _props$componentOverr, _props$styleProps4, _props$styleProps4$cl;

  const elementId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? _defaultOOOHPaneControlProps.defaultOOOHPaneControlProps.id;
  const containerStyles = {
    root: Object.assign({}, _defaultOOOHPaneGeneralStyles.defaultOOOHPaneGeneralStyles, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const titleStyles = {
    root: Object.assign({}, _defaultOOOHPaneTitleStyles.defaultOOOHPaneTitleStyles, (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.titleStyleProps)
  };
  const displayText = (0, _utils.replaceURLWithAnchor)(((_props$controlProps2 = props.controlProps) === null || _props$controlProps2 === void 0 ? void 0 : _props$controlProps2.titleText) ?? _defaultOOOHPaneControlProps.defaultOOOHPaneControlProps.titleText, ((_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.openLinkInNewTab) ?? _defaultOOOHPaneControlProps.defaultOOOHPaneControlProps.openLinkInNewTab) ?? "";
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, !((_props$controlProps4 = props.controlProps) !== null && _props$controlProps4 !== void 0 && _props$controlProps4.hideOOOHPane) && /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    id: elementId,
    tabIndex: -1,
    styles: containerStyles,
    role: ((_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.role) ?? _defaultOOOHPaneControlProps.defaultOOOHPaneControlProps.role,
    className: (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : (_props$styleProps3$cl = _props$styleProps3.classNames) === null || _props$styleProps3$cl === void 0 ? void 0 : _props$styleProps3$cl.containerClassName,
    dir: ((_props$controlProps6 = props.controlProps) === null || _props$controlProps6 === void 0 ? void 0 : _props$controlProps6.dir) ?? _defaultOOOHPaneControlProps.defaultOOOHPaneControlProps.dir
  }, !((_props$controlProps7 = props.controlProps) !== null && _props$controlProps7 !== void 0 && _props$controlProps7.hideTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr = props.componentOverrides) === null || _props$componentOverr === void 0 ? void 0 : _props$componentOverr.title) || /*#__PURE__*/_react2.default.createElement(_react.Text, {
    className: (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : (_props$styleProps4$cl = _props$styleProps4.classNames) === null || _props$styleProps4$cl === void 0 ? void 0 : _props$styleProps4$cl.titleClassName,
    styles: titleStyles,
    tabIndex: -1,
    id: elementId + "-title"
  }, /*#__PURE__*/_react2.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: displayText
    }
  })))));
}

var _default = OOOHPane;
exports.default = _default;
