// @microsoft/omnichannel-chat-components/lib/esm/components/callingcontainer/CallingContainer.js
import * as React from "react";
import { Stack } from "@fluentui/react";
import CurrentCall from "./subcomponents/CurrentCall/CurrentCall";
import IncomingCall from "./subcomponents/IncomingCall/IncomingCall";
import { defaultCallingContainerProps } from "./common/defaultProps/defaultCallingContainerProps";

function CallingContainer(props) {
  var _props$controlProps, _defaultCallingContai, _defaultCallingContai2, _props$styleProps, _props$controlProps2, _defaultCallingContai3, _props$styleProps2, _props$controlProps3, _defaultCallingContai4, _props$controlProps4, _props$styleProps3, _props$controlProps5, _props$styleProps4;

  const callingContainerId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? ((_defaultCallingContai = defaultCallingContainerProps.controlProps) === null || _defaultCallingContai === void 0 ? void 0 : _defaultCallingContai.id);
  let stackStyles = {
    root: Object.assign({}, (_defaultCallingContai2 = defaultCallingContainerProps.styleProps) === null || _defaultCallingContai2 === void 0 ? void 0 : _defaultCallingContai2.generalStyleProps, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const hideCallingContainer = ((_props$controlProps2 = props.controlProps) === null || _props$controlProps2 === void 0 ? void 0 : _props$controlProps2.hideCallingContainer) ?? ((_defaultCallingContai3 = defaultCallingContainerProps.controlProps) === null || _defaultCallingContai3 === void 0 ? void 0 : _defaultCallingContai3.hideCallingContainer);

  if (hideCallingContainer) {
    stackStyles = {
      root: Object.assign({}, stackStyles.root, {
        display: "none"
      })
    };
  }

  return /*#__PURE__*/React.createElement(Stack, {
    id: callingContainerId,
    className: (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.className,
    horizontalAlign: "space-between",
    dir: ((_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.dir) ?? ((_defaultCallingContai4 = defaultCallingContainerProps.controlProps) === null || _defaultCallingContai4 === void 0 ? void 0 : _defaultCallingContai4.dir),
    styles: stackStyles
  }, ((_props$controlProps4 = props.controlProps) === null || _props$controlProps4 === void 0 ? void 0 : _props$controlProps4.isIncomingCall) && /*#__PURE__*/React.createElement(IncomingCall, {
    controlProps: props.controlProps.incomingCallControlProps,
    styleProps: (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.incomingCallStyleProps
  }), ((_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.isIncomingCall) === false && /*#__PURE__*/React.createElement(CurrentCall, {
    controlProps: props.controlProps.currentCallControlProps,
    styleProps: (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.currentCallStyleProps
  }));
}

export default CallingContainer;
