// @microsoft/omnichannel-chat-components/lib/cjs/components/confirmationpane/common/defaultStyles/defaultConfirmationPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultConfirmationPaneStyles = void 0;

var _defaultConfirmationPaneButtonGroupStyles = require("./defaultConfirmationPaneButtonGroupStyles");

var _defaultConfirmationPaneCancelButtonStyles = require("./defaultConfirmationPaneCancelButtonStyles");

var _defaultConfirmationPaneConfirmButtonStyles = require("./defaultConfirmationPaneConfirmButtonStyles");

var _defaultConfirmationPaneGeneralStyles = require("./defaultConfirmationPaneGeneralStyles");

var _defaultConfirmationPaneSubtitleStyles = require("./defaultConfirmationPaneSubtitleStyles");

var _defaultConfirmationPaneTitleStyles = require("./defaultConfirmationPaneTitleStyles");

const defaultConfirmationPaneStyles = {
  generalStyleProps: _defaultConfirmationPaneGeneralStyles.defaultConfirmationPaneGeneralStyles,
  titleStyleProps: _defaultConfirmationPaneTitleStyles.defaultConfirmationPaneTitleStyles,
  subtitleStyleProps: _defaultConfirmationPaneSubtitleStyles.defaultConfirmationPaneSubtitleStyles,
  buttonGroupStyleProps: _defaultConfirmationPaneButtonGroupStyles.defaultConfirmationPaneButtonGroupStyles,
  confirmButtonStyleProps: _defaultConfirmationPaneConfirmButtonStyles.defaultConfirmationPaneConfirmButtonStyles,
  cancelButtonStyleProps: _defaultConfirmationPaneCancelButtonStyles.defaultConfirmationPaneCancelButtonStyles
};
exports.defaultConfirmationPaneStyles = defaultConfirmationPaneStyles;
