// botframework-webchat/lib/adaptiveCards/DOMManipulationWithUndo/durableDisableInputElementAccessiblyWithUndo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = durableDisableInputElementAccessiblyWithUndo;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _addEventListenerWithUndo = _interopRequireDefault(require("./addEventListenerWithUndo"));

var _bunchUndos = _interopRequireDefault(require("./bunchUndos"));

var _noOp = _interopRequireDefault(require("./private/noOp"));

var _setOrRemoveAttributeIfFalseWithUndo = _interopRequireDefault(require("./setOrRemoveAttributeIfFalseWithUndo"));

/**
 * An event handler for disabling event bubbling and propagation.
 */
function disabledHandler(event) {
  event.preventDefault();
  event.stopImmediatePropagation();
  event.stopPropagation();
}

function disable(element, undoStack) {
  var tag = element.nodeName.toLowerCase();
  /* eslint-disable-next-line default-case */

  switch (tag) {
    case 'button':
    case 'input':
    case 'select':
    case 'textarea':
      undoStack.push( // "click" handler in capture phase to make sure we can block as much "click" event listeners as possible.
      (0, _addEventListenerWithUndo.default)(element, 'click', disabledHandler, {
        capture: true
      }), (0, _setOrRemoveAttributeIfFalseWithUndo.default)(element, 'aria-disabled', 'true'), (0, _setOrRemoveAttributeIfFalseWithUndo.default)(element, 'tabindex', '-1'));

      if (tag === 'input' || tag === 'textarea') {
        undoStack.push((0, _setOrRemoveAttributeIfFalseWithUndo.default)(element, 'readonly', 'readonly'));
      } else if (tag === 'select') {
        undoStack.push.apply(undoStack, (0, _toConsumableArray2.default)(Array.from(element.querySelectorAll('option')).map(function (option) {
          return (0, _setOrRemoveAttributeIfFalseWithUndo.default)(option, 'disabled', 'disabled');
        })));
      }

      break;
  }
}
/**
 * Disables an input element in accessible fashion with undo function.
 *
 * This is designed for accessibility and mimick the behavior of `disabled` attribute in accessible form:
 *
 * - Take away from focus ring;
 *   - If currently focused, do not move focus;
 * - Mark content as readonly.
 *
 * Thus, it should not impact hyperlinks or other contents which are not affected by `disabled` attribute.
 *
 * For simplicity, currently, we did not disable element with `contenteditable` attribute.
 *
 * We only disable these elements: `<button>`, `<input>`, `<select>`, `<textarea>`.
 *
 * We need durability as Adaptive Cards occasionally reset `tabindex="0"`.
 *
 * @returns {function} A function, when called, will restore to previous state.
 */


function durableDisableInputElementAccessiblyWithUndo(element) {
  if (!element) {
    return _noOp.default;
  }

  var undoStack = [];

  var apply = function apply() {
    return disable(element, undoStack);
  };

  apply();
  var observer = new MutationObserver(apply);
  observer.observe(element, {
    attributeFilter: ['tabindex']
  });
  undoStack.push(function () {
    return observer.disconnect();
  });
  return (0, _bunchUndos.default)(undoStack);
}
