// botframework-webchat/lib/adaptiveCards/Attachment/AdaptiveCardHacks/private/useAdaptiveCardModEffect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useAdaptiveCardModEffect;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _classPrivateFieldGet5 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldGet"));

var _classPrivateFieldSet2 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldSet"));

var _react = require("react");

var _useLazyRef = _interopRequireDefault(require("./useLazyRef"));

var _useValueRef = _interopRequireDefault(require("./useValueRef"));

function _classPrivateFieldInitSpec(obj, privateMap, value) { _checkPrivateRedeclaration(obj, privateMap); privateMap.set(obj, value); }

function _checkPrivateRedeclaration(obj, privateCollection) { if (privateCollection.has(obj)) { throw new TypeError("Cannot initialize the same private elements twice on an object"); } }

var _mod = /*#__PURE__*/new WeakMap();

var _undo = /*#__PURE__*/new WeakMap();

var Mod = /*#__PURE__*/function () {
  function Mod(mod) {
    (0, _classCallCheck2.default)(this, Mod);

    _classPrivateFieldInitSpec(this, _mod, {
      writable: true,
      value: void 0
    });

    _classPrivateFieldInitSpec(this, _undo, {
      writable: true,
      value: void 0
    });

    (0, _classPrivateFieldSet2.default)(this, _mod, mod);
  } // @ts-ignore We are using Babel to transpile and it will transpile private modifier.


  (0, _createClass2.default)(Mod, [{
    key: "apply",
    value: function apply(adaptiveCard, cardElement) {
      var _classPrivateFieldGet2, _classPrivateFieldGet3;

      (_classPrivateFieldGet2 = (0, _classPrivateFieldGet5.default)(this, _undo)) === null || _classPrivateFieldGet2 === void 0 ? void 0 : _classPrivateFieldGet2.call(this);

      for (var _len = arguments.length, args = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
        args[_key - 2] = arguments[_key];
      }

      (0, _classPrivateFieldSet2.default)(this, _undo, adaptiveCard && cardElement && (_classPrivateFieldGet3 = (0, _classPrivateFieldGet5.default)(this, _mod)).call.apply(_classPrivateFieldGet3, [this, adaptiveCard, cardElement].concat(args)));
    }
  }, {
    key: "undo",
    value: function undo() {
      var _classPrivateFieldGet4;

      (_classPrivateFieldGet4 = (0, _classPrivateFieldGet5.default)(this, _undo)) === null || _classPrivateFieldGet4 === void 0 ? void 0 : _classPrivateFieldGet4.call(this);
      (0, _classPrivateFieldSet2.default)(this, _undo, undefined);
    }
  }]);
  return Mod;
}();
/**
 * Creates a mod effect for Adaptive Card.
 *
 * When this hook is executed, it will return two functions for applying and undo the mod.
 * It will also monitor the DOM tree and undo-then-reapply if mutation occurred.
 *
 * The first function must be called right after DOM is mounted. The second function must be called right before re-render.
 *
 * @return {[function, function]} Two functions, the first one to apply the mod, the second one to undo the mod.
 */


function useAdaptiveCardModEffect(modder, adaptiveCard) {
  var adaptiveCardRef = (0, _useValueRef.default)(adaptiveCard);
  var mod = (0, _react.useMemo)(function () {
    return new Mod(modder);
  }, [modder]);
  var reapplyRef = (0, _react.useRef)();
  var observerRef = (0, _useLazyRef.default)(function () {
    return new MutationObserver(function () {
      var _reapplyRef$current;

      (_reapplyRef$current = reapplyRef.current) === null || _reapplyRef$current === void 0 ? void 0 : _reapplyRef$current.call(reapplyRef);
    });
  });
  (0, _react.useEffect)(function () {
    return function () {
      observerRef.current.disconnect();
    };
  }, [observerRef]);
  var handleApply = (0, _react.useCallback)(function (cardElement) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    if (adaptiveCardRef.current && cardElement) {
      // Apply the mod immediately, then assign the function to reapply() so we can call later when mutation happens.
      (reapplyRef.current = function () {
        return mod.apply.apply(mod, [adaptiveCardRef.current, cardElement].concat(args));
      })();
    }

    var observer = observerRef.current;
    observer.disconnect();
    observer.observe(cardElement, {
      childList: true,
      subtree: true
    });
  }, [adaptiveCardRef, observerRef, mod]);
  var handleUndo = (0, _react.useCallback)(function () {
    mod.undo(); // If we have undo-ed the mod, calling reapply() through MutationObserver should be no-op.

    reapplyRef.current = undefined;
  }, [mod, reapplyRef]);
  return (0, _react.useMemo)(function () {
    return Object.freeze([handleApply, handleUndo]);
  }, [handleApply, handleUndo]);
}
