// @microsoft/omnichannel-chat-components/lib/cjs/components/common/commandbutton/CommandButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("@fluentui/react");

var _react2 = _interopRequireWildcard(require("react"));

var _BroadcastService = require("../../../services/BroadcastService");

var _Constants = require("../../../common/Constants");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function CommandButton(props) {
  var _props$hoverStyles, _props$hoverStyles2, _props$focusStyles;

  //imageIconProps > iconName
  const iconProp = props.imageIconProps ? {
    imageProps: props.imageIconProps
  } : {
    iconName: props.iconName
  };
  let iconStyles = {};

  if (props.type === _Constants.ButtonTypes.Icon) {
    var _props$styles;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    iconStyles = { ...(props === null || props === void 0 ? void 0 : (_props$styles = props.styles) === null || _props$styles === void 0 ? void 0 : _props$styles.icon)
    };
  }

  const buttonStyles = {
    icon: { ...iconStyles
    },
    root: {
      selectors: {
        ":hover .ms-Button-icon": { // eslint-disable-next-line @typescript-eslint/no-explicit-any
          ...(props === null || props === void 0 ? void 0 : (_props$hoverStyles = props.hoverStyles) === null || _props$hoverStyles === void 0 ? void 0 : _props$hoverStyles.icon)
        },
        ":active .ms-Button-icon": { // eslint-disable-next-line @typescript-eslint/no-explicit-any
          ...(props === null || props === void 0 ? void 0 : (_props$hoverStyles2 = props.hoverStyles) === null || _props$hoverStyles2 === void 0 ? void 0 : _props$hoverStyles2.icon)
        },
        ":focus .ms-Button-icon": { // eslint-disable-next-line @typescript-eslint/no-explicit-any
          ...(props === null || props === void 0 ? void 0 : (_props$focusStyles = props.focusStyles) === null || _props$focusStyles === void 0 ? void 0 : _props$focusStyles.icon)
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(props === null || props === void 0 ? void 0 : props.styles)
    },
    rootHovered: props.hoverStyles,
    rootFocused: props.focusStyles,
    rootPressed: props.hoverStyles
  };
  const handleOnClick = (0, _react2.useCallback)(() => {
    if (props !== null && props !== void 0 && props.onClick) {
      const customEvent = props.customEvent ?? {
        eventName: _Constants.EventNames.OnClick,
        elementType: _Constants.ElementType.Custom,
        elementId: props === null || props === void 0 ? void 0 : props.id
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      props === null || props === void 0 ? void 0 : props.onClick();
    }
  }, []);
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, props.type === _Constants.ButtonTypes.Text && /*#__PURE__*/_react2.default.createElement(_react.DefaultButton, {
    id: props.id,
    text: props.text,
    ariaLabel: props.ariaLabel,
    onClick: handleOnClick,
    disabled: props.disabled,
    className: props.className,
    styles: buttonStyles,
    title: props.hideButtonTitle ? undefined : props.ariaLabel
  }), props.type === _Constants.ButtonTypes.Icon && /*#__PURE__*/_react2.default.createElement(_react.IconButton, {
    id: props.id,
    iconProps: iconProp,
    title: props.hideButtonTitle ? undefined : props.ariaLabel,
    ariaLabel: props.ariaLabel,
    disabled: props.disabled,
    styles: buttonStyles,
    onClick: handleOnClick,
    className: props.className
  }));
}

var _default = CommandButton;
exports.default = _default;
