// @microsoft/omnichannel-chat-widget/lib/cjs/components/webchatcontainerstateful/webchatcontroller/notification/NotificationHandler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotificationHandler = void 0;
var _Constants = require("../../../../common/Constants");
var _NotificationLevel = require("../enums/NotificationLevel");
var _WebChatActionType = require("../enums/WebChatActionType");
var _WebChatStoreLoader = require("../WebChatStoreLoader");
var _utils = require("../../../../common/utils");
class NotificationHandler {
  static notify(id, level, message) {
    if (_WebChatStoreLoader.WebChatStoreLoader.store) {
      _WebChatStoreLoader.WebChatStoreLoader.store.dispatch({
        type: _WebChatActionType.WebChatActionType.WEB_CHAT_SET_NOTIFICATION,
        payload: {
          id,
          level,
          message
        }
      });
      NotificationHandler.setFocusOnNotificationCloseButton();
    }
  }
  static dismissNotification(id) {
    if (_WebChatStoreLoader.WebChatStoreLoader.store) {
      _WebChatStoreLoader.WebChatStoreLoader.store.dispatch({
        type: _WebChatActionType.WebChatActionType.WEB_CHAT_DISMISS_NOTIFICATION,
        payload: {
          id
        }
      });
    }
  }
  static notifyError(id, message) {
    this.notify(id, _NotificationLevel.NotificationLevel.Error, message);
  }
  static notifyWarning(id, message) {
    this.notify(id, _NotificationLevel.NotificationLevel.Warning, message);
  }
  static notifyInfo(id, message) {
    this.notify(id, _NotificationLevel.NotificationLevel.Info, message);
  }
  static notifySuccess(id, message) {
    this.notify(id, _NotificationLevel.NotificationLevel.Success, message);
  }
  static notifyWithLevel(id, message, level) {
    if (!level) {
      this.notifyInfo(id, message);
    } else {
      this.notify(id, level, message);
    }
  }

  /* Bypass WebChat behavior and focus on banner close button */
  static setFocusOnNotificationCloseButton() {
    const expandButton = document.querySelector(`#${_Constants.HtmlIdNames.MSLiveChatWidget} .${_Constants.HtmlClassNames.webChatBannerExpandButton}`);
    if (expandButton) {
      expandButton.focus();
      return;
    }
    const closeButton = document.querySelector(`#${_Constants.HtmlIdNames.MSLiveChatWidget} .${_Constants.HtmlClassNames.webChatBannerCloseButton}`);
    if (closeButton) {
      closeButton.focus();
      NotificationHandler.registerOnCloseEvent(closeButton);
      return;
    }
  }
  static registerOnCloseEvent(button) {
    button.onclick = () => {
      (0, _utils.setFocusOnSendBox)();
    };
  }
}
exports.NotificationHandler = NotificationHandler;
