// @microsoft/omnichannel-chat-widget/lib/esm/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/sanitizationMiddleware.js
/******
 * SanitizationMiddleware
 * 
 * Sanitizes the text.
 ******/

import DOMPurify from "dompurify";
import { LogLevel, TelemetryEvent } from "../../../../../common/telemetry/TelemetryConstants";
import { TelemetryHelper } from "../../../../../common/telemetry/TelemetryHelper";
import { WebChatActionType } from "../../enums/WebChatActionType";

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars
const sanitizationMiddleware = _ref => {
  let {
    dispatch
  } = _ref;
  return next => action => {
    if (action.type === WebChatActionType.WEB_CHAT_SEND_MESSAGE) {
      try {
        var _action$payload;
        let text = (_action$payload = action.payload) === null || _action$payload === void 0 ? void 0 : _action$payload.text;
        if (text) {
          text = DOMPurify.sanitize(text) ?? " ";
        }
      } catch (e) {
        const copyDataForTelemetry = {
          ...action,
          payload: {
            ...action.payload,
            text: undefined
          }
        };
        let errorMessage = "Failed to apply action: ";
        try {
          errorMessage += JSON.stringify(copyDataForTelemetry);
        } catch (e) {
          errorMessage += " (unable to stringify action)";
        }
        TelemetryHelper.logActionEvent(LogLevel.ERROR, {
          Event: TelemetryEvent.ProcessingSanitizationMiddlewareFailed,
          ExceptionDetails: {
            ErrorData: errorMessage,
            Exception: e
          }
        });
      }
    }
    return next(action);
  };
};
export default sanitizationMiddleware;
