// @microsoft/omnichannel-chat-widget/lib/cjs/contexts/common/LiveChatWidgetContextInitialState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLiveChatWidgetContextInitialState = void 0;
var _ConversationState = require("./ConversationState");
var _defaultMiddlewareLocalizedTexts = require("../../components/webchatcontainerstateful/common/defaultProps/defaultMiddlewareLocalizedTexts");
var _utils = require("../../common/utils");
var _defaultClientDataStoreProvider = require("../../common/storage/default/defaultClientDataStoreProvider");
var _Constants = require("../../common/Constants");
const getLiveChatWidgetContextInitialState = props => {
  var _props$controlProps, _props$webChatContain;
  const widgetCacheId = (0, _utils.getWidgetCacheIdfromProps)(props);
  const cacheTtlInMins = (props === null || props === void 0 ? void 0 : (_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.cacheTtlInMins) ?? _Constants.Constants.CacheTtlInMinutes;
  const storageType = (props === null || props === void 0 ? void 0 : props.useSessionStorage) === true ? _Constants.StorageType.sessionStorage : _Constants.StorageType.localStorage;
  const initialState = (0, _defaultClientDataStoreProvider.defaultClientDataStoreProvider)(cacheTtlInMins, storageType).getData(widgetCacheId);
  if (!(0, _utils.isNullOrUndefined)(initialState)) {
    const initialStateFromCache = JSON.parse(initialState);
    return initialStateFromCache;
  }
  const LiveChatWidgetContextInitialState = {
    domainStates: {
      liveChatConfig: props.chatConfig,
      widgetElementId: "",
      renderingMiddlewareProps: (_props$webChatContain = props.webChatContainerProps) === null || _props$webChatContain === void 0 ? void 0 : _props$webChatContain.renderingMiddlewareProps,
      middlewareLocalizedTexts: _defaultMiddlewareLocalizedTexts.defaultMiddlewareLocalizedTexts,
      preChatSurveyResponse: "{}",
      chatToken: undefined,
      postChatContext: undefined,
      telemetryInternalData: {},
      globalDir: "ltr",
      liveChatContext: undefined,
      customContext: undefined,
      widgetSize: undefined,
      widgetInstanceId: "",
      initialChatSdkRequestId: "",
      transcriptRequestId: "",
      confirmationPaneConfirmedOptionClicked: false,
      confirmationState: _Constants.ConfirmationState.NotSet
    },
    appStates: {
      conversationState: _ConversationState.ConversationState.Closed,
      isMinimized: undefined,
      previousElementIdOnFocusBeforeModalOpen: null,
      startChatFailed: false,
      outsideOperatingHours: false,
      preChatResponseEmail: "",
      isAudioMuted: null,
      newMessage: false,
      hideStartChatButton: false,
      reconnectId: undefined,
      proactiveChatStates: {
        proactiveChatBodyTitle: "",
        proactiveChatEnablePrechat: false,
        proactiveChatInNewWindow: false
      },
      e2vvEnabled: false,
      unreadMessageCount: 0,
      conversationEndedBy: _Constants.ConversationEndEntity.NotSet,
      chatDisconnectEventReceived: false,
      selectedSurveyMode: null
    },
    uiStates: {
      showConfirmationPane: false,
      showEmailTranscriptPane: false,
      showCallingPopup: false,
      isIncomingCall: true,
      disableVideoCall: true,
      disableRemoteVideo: true,
      disableSelfVideo: true,
      focusChatButton: false
    }
  };
  return LiveChatWidgetContextInitialState;
};
exports.getLiveChatWidgetContextInitialState = getLiveChatWidgetContextInitialState;
