// @microsoft/omnichannel-chat-components/lib/cjs/components/reconnectchatpane/ReconnectChatPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("@fluentui/react");

var _react2 = _interopRequireWildcard(require("react"));

var _BroadcastService = require("../../services/BroadcastService");

var _Constants = require("../../common/Constants");

var _decodeComponentString = require("../../common/decodeComponentString");

var _defaultReconnectChatPaneButtonGroupStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneButtonGroupStyles");

var _defaultReconnectChatPaneContinueChatButtonHoveredStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneContinueChatButtonHoveredStyles");

var _defaultReconnectChatPaneContinueChatButtonStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneContinueChatButtonStyles");

var _defaultReconnectChatPaneControlProps = require("./common/default/defaultProps/defaultReconnectChatPaneControlProps");

var _defaultReconnectChatPaneGeneralStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneGeneralStyles");

var _defaultReconnectChatPaneIconStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneIconStyles");

var _defaultReconnectChatPaneStartNewChatButtonHoveredStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneStartNewChatButtonHoveredStyles");

var _defaultReconnectChatPaneStartNewChatButtonStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneStartNewChatButtonStyles");

var _defaultReconnectChatPaneSubtitleStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneSubtitleStyles");

var _defaultReconnectChatPaneTitleStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneTitleStyles");

var _defaultReconnectChatPaneWrapperStyles = require("./common/default/defaultStyles/defaultReconnectChatPaneWrapperStyles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ReconnectChatPane(props) {
  var _props$controlProps, _props$styleProps, _props$styleProps2, _props$styleProps3, _props$styleProps4, _props$styleProps5, _props$styleProps6, _props$styleProps7, _props$styleProps8, _props$styleProps9, _props$styleProps10, _props$styleProps11, _props$styleProps12, _props$controlProps8, _props$styleProps13, _props$styleProps13$c, _props$controlProps9, _props$controlProps10, _props$styleProps14, _props$styleProps14$c, _props$controlProps11, _props$componentOverr, _props$styleProps15, _props$styleProps15$c, _props$controlProps12, _props$controlProps13, _props$componentOverr2, _props$styleProps16, _props$styleProps16$c, _props$controlProps14, _props$controlProps15, _props$componentOverr3, _props$styleProps17, _props$styleProps17$c, _props$controlProps16, _props$controlProps17, _props$styleProps18, _props$styleProps18$c, _props$controlProps18, _props$componentOverr4, _props$styleProps19, _props$styleProps19$c, _props$controlProps19, _props$controlProps20, _props$controlProps21, _props$componentOverr5, _props$styleProps20, _props$styleProps20$c, _props$controlProps22, _props$controlProps23;

  const elementId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.id;
  const handleContinueChatClick = (0, _react2.useCallback)(() => {
    var _props$controlProps2;

    if ((_props$controlProps2 = props.controlProps) !== null && _props$controlProps2 !== void 0 && _props$controlProps2.onContinueChat) {
      var _props$controlProps3;

      const customEvent = {
        elementType: _Constants.ElementType.ReconnectChatContinueChatButton,
        elementId: elementId + "-continueconversationbutton",
        eventName: _Constants.EventNames.OnClick
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.onContinueChat();
    }
  }, []);
  const handleStartNewChatClick = (0, _react2.useCallback)(() => {
    var _props$controlProps4;

    if ((_props$controlProps4 = props.controlProps) !== null && _props$controlProps4 !== void 0 && _props$controlProps4.onStartNewChat) {
      var _props$controlProps5;

      const customEvent = {
        elementType: _Constants.ElementType.ReconnectChatStartNewChatButton,
        elementId: elementId + "-startnewconversationbutton",
        eventName: _Constants.EventNames.OnClick
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.onStartNewChat();
    }
  }, []); // eslint-disable-next-line @typescript-eslint/no-explicit-any

  const handleEscKeyDown = (0, _react2.useCallback)(e => {
    var _props$controlProps6;

    if (e.code === _Constants.KeyCodes.ESCAPE && (_props$controlProps6 = props.controlProps) !== null && _props$controlProps6 !== void 0 && _props$controlProps6.onMinimize) {
      var _props$controlProps7;

      const customEvent = {
        elementType: _Constants.ElementType.ReconnectChatPane,
        elementId: elementId,
        eventName: _Constants.EventNames.OnEscapeKeyDown
      };

      _BroadcastService.BroadcastService.postMessage(customEvent);

      (_props$controlProps7 = props.controlProps) === null || _props$controlProps7 === void 0 ? void 0 : _props$controlProps7.onMinimize();
    }
  }, []);
  const containerStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneGeneralStyles.defaultReconnectChatPaneGeneralStyles, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const wrapperStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneWrapperStyles.defaultReconnectChatPaneWrapperStyles, (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.wrapperStyleProps)
  };
  const titleStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneTitleStyles.defaultReconnectChatPaneTitleStyles, (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.titleStyleProps)
  };
  const subtitleStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneSubtitleStyles.defaultReconnectChatPaneSubtitleStyles, (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.subtitleStyleProps)
  };
  const iconStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneIconStyles.defaultReconnectChatPaneIconStyles, (_props$styleProps5 = props.styleProps) === null || _props$styleProps5 === void 0 ? void 0 : _props$styleProps5.iconStyleProps)
  };
  const buttonGroupStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneButtonGroupStyles.defaultReconnectChatPaneButtonGroupStyles, (_props$styleProps6 = props.styleProps) === null || _props$styleProps6 === void 0 ? void 0 : _props$styleProps6.buttonGroupStyleProps)
  };
  const continueChatButtonStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneContinueChatButtonStyles.defaultReconnectChatPaneContinueChatButtonStyles, (_props$styleProps7 = props.styleProps) === null || _props$styleProps7 === void 0 ? void 0 : _props$styleProps7.continueChatButtonStyleProps),
    rootHovered: Object.assign({}, _defaultReconnectChatPaneContinueChatButtonHoveredStyles.defaultReconnectChatPaneContinueChatButtonHoveredStyles, (_props$styleProps8 = props.styleProps) === null || _props$styleProps8 === void 0 ? void 0 : _props$styleProps8.continueChatButtonHoveredStyleProps),
    rootPressed: Object.assign({}, _defaultReconnectChatPaneContinueChatButtonHoveredStyles.defaultReconnectChatPaneContinueChatButtonHoveredStyles, (_props$styleProps9 = props.styleProps) === null || _props$styleProps9 === void 0 ? void 0 : _props$styleProps9.continueChatButtonHoveredStyleProps)
  };
  const startNewChatButtonStyles = {
    root: Object.assign({}, _defaultReconnectChatPaneStartNewChatButtonStyles.defaultReconnectChatPaneStartNewChatButtonStyles, (_props$styleProps10 = props.styleProps) === null || _props$styleProps10 === void 0 ? void 0 : _props$styleProps10.startNewChatButtonStyleProps),
    rootHovered: Object.assign({}, _defaultReconnectChatPaneStartNewChatButtonHoveredStyles.defaultReconnectChatPaneStartNewChatButtonHoveredStyles, (_props$styleProps11 = props.styleProps) === null || _props$styleProps11 === void 0 ? void 0 : _props$styleProps11.startNewChatButtonHoveredStyleProps),
    rootPressed: Object.assign({}, _defaultReconnectChatPaneStartNewChatButtonHoveredStyles.defaultReconnectChatPaneStartNewChatButtonHoveredStyles, (_props$styleProps12 = props.styleProps) === null || _props$styleProps12 === void 0 ? void 0 : _props$styleProps12.startNewChatButtonHoveredStyleProps)
  };
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, !((_props$controlProps8 = props.controlProps) !== null && _props$controlProps8 !== void 0 && _props$controlProps8.hideReconnectChatPane) && /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps13 = props.styleProps) === null || _props$styleProps13 === void 0 ? void 0 : (_props$styleProps13$c = _props$styleProps13.classNames) === null || _props$styleProps13$c === void 0 ? void 0 : _props$styleProps13$c.containerClassName,
    id: elementId,
    tabIndex: -1,
    onKeyDown: handleEscKeyDown,
    dir: ((_props$controlProps9 = props.controlProps) === null || _props$controlProps9 === void 0 ? void 0 : _props$controlProps9.dir) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.dir,
    "aria-label": ((_props$controlProps10 = props.controlProps) === null || _props$controlProps10 === void 0 ? void 0 : _props$controlProps10.reconnectChatPaneAriaLabel) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.reconnectChatPaneAriaLabel,
    styles: containerStyles
  }, /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    className: (_props$styleProps14 = props.styleProps) === null || _props$styleProps14 === void 0 ? void 0 : (_props$styleProps14$c = _props$styleProps14.classNames) === null || _props$styleProps14$c === void 0 ? void 0 : _props$styleProps14$c.wrapperClassName,
    styles: wrapperStyles,
    tabIndex: -1,
    id: elementId + "-wrapper"
  }, !((_props$controlProps11 = props.controlProps) !== null && _props$controlProps11 !== void 0 && _props$controlProps11.hideTitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr = props.componentOverrides) === null || _props$componentOverr === void 0 ? void 0 : _props$componentOverr.title) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps15 = props.styleProps) === null || _props$styleProps15 === void 0 ? void 0 : (_props$styleProps15$c = _props$styleProps15.classNames) === null || _props$styleProps15$c === void 0 ? void 0 : _props$styleProps15$c.titleClassName,
    styles: titleStyles,
    tabIndex: -1,
    id: elementId + "-title"
  }, ((_props$controlProps12 = props.controlProps) === null || _props$controlProps12 === void 0 ? void 0 : _props$controlProps12.titleText) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.titleText)), !((_props$controlProps13 = props.controlProps) !== null && _props$controlProps13 !== void 0 && _props$controlProps13.hideSubtitle) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr2 = props.componentOverrides) === null || _props$componentOverr2 === void 0 ? void 0 : _props$componentOverr2.subtitle) || /*#__PURE__*/_react2.default.createElement(_react.Label, {
    className: (_props$styleProps16 = props.styleProps) === null || _props$styleProps16 === void 0 ? void 0 : (_props$styleProps16$c = _props$styleProps16.classNames) === null || _props$styleProps16$c === void 0 ? void 0 : _props$styleProps16$c.subtitleClassName,
    styles: subtitleStyles,
    tabIndex: -1,
    id: elementId + "-subtitle"
  }, ((_props$controlProps14 = props.controlProps) === null || _props$controlProps14 === void 0 ? void 0 : _props$controlProps14.subtitleText) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.subtitleText)), !((_props$controlProps15 = props.controlProps) !== null && _props$controlProps15 !== void 0 && _props$controlProps15.hideIcon) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr3 = props.componentOverrides) === null || _props$componentOverr3 === void 0 ? void 0 : _props$componentOverr3.icon) || /*#__PURE__*/_react2.default.createElement(_react.Icon, {
    className: (_props$styleProps17 = props.styleProps) === null || _props$styleProps17 === void 0 ? void 0 : (_props$styleProps17$c = _props$styleProps17.classNames) === null || _props$styleProps17$c === void 0 ? void 0 : _props$styleProps17$c.iconClassName,
    styles: iconStyles,
    tabIndex: -1,
    role: "img",
    id: elementId + "-icon",
    "aria-label": ((_props$controlProps16 = props.controlProps) === null || _props$controlProps16 === void 0 ? void 0 : _props$controlProps16.iconAriaLabel) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.iconAriaLabel
  })), /*#__PURE__*/_react2.default.createElement(_react.Stack, {
    horizontal: ((_props$controlProps17 = props.controlProps) === null || _props$controlProps17 === void 0 ? void 0 : _props$controlProps17.isButtonGroupHorizontal) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.isButtonGroupHorizontal,
    className: (_props$styleProps18 = props.styleProps) === null || _props$styleProps18 === void 0 ? void 0 : (_props$styleProps18$c = _props$styleProps18.classNames) === null || _props$styleProps18$c === void 0 ? void 0 : _props$styleProps18$c.buttonGroupClassName,
    styles: buttonGroupStyles,
    tabIndex: -1,
    id: elementId + "-buttongroup"
  }, !((_props$controlProps18 = props.controlProps) !== null && _props$controlProps18 !== void 0 && _props$controlProps18.hideContinueChatButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr4 = props.componentOverrides) === null || _props$componentOverr4 === void 0 ? void 0 : _props$componentOverr4.continueChatButton) || /*#__PURE__*/_react2.default.createElement(_react.DefaultButton, {
    className: (_props$styleProps19 = props.styleProps) === null || _props$styleProps19 === void 0 ? void 0 : (_props$styleProps19$c = _props$styleProps19.classNames) === null || _props$styleProps19$c === void 0 ? void 0 : _props$styleProps19$c.continueChatButtonClassName,
    styles: continueChatButtonStyles,
    tabIndex: 0,
    text: ((_props$controlProps19 = props.controlProps) === null || _props$controlProps19 === void 0 ? void 0 : _props$controlProps19.continueChatButtonText) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.continueChatButtonText,
    onClick: handleContinueChatClick,
    id: elementId + "-continueconversationbutton",
    "aria-label": ((_props$controlProps20 = props.controlProps) === null || _props$controlProps20 === void 0 ? void 0 : _props$controlProps20.continueChatButtonAriaLabel) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.continueChatButtonAriaLabel
  })), !((_props$controlProps21 = props.controlProps) !== null && _props$controlProps21 !== void 0 && _props$controlProps21.hideStartNewChatButton) && ((0, _decodeComponentString.decodeComponentString)((_props$componentOverr5 = props.componentOverrides) === null || _props$componentOverr5 === void 0 ? void 0 : _props$componentOverr5.startNewChatButton) || /*#__PURE__*/_react2.default.createElement(_react.DefaultButton, {
    className: (_props$styleProps20 = props.styleProps) === null || _props$styleProps20 === void 0 ? void 0 : (_props$styleProps20$c = _props$styleProps20.classNames) === null || _props$styleProps20$c === void 0 ? void 0 : _props$styleProps20$c.startNewChatButtonClassName,
    styles: startNewChatButtonStyles,
    tabIndex: 0,
    text: ((_props$controlProps22 = props.controlProps) === null || _props$controlProps22 === void 0 ? void 0 : _props$controlProps22.startNewChatButtonText) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.startNewChatButtonText,
    onClick: handleStartNewChatClick,
    id: elementId + "-startnewconversationbutton",
    "aria-label": ((_props$controlProps23 = props.controlProps) === null || _props$controlProps23 === void 0 ? void 0 : _props$controlProps23.startNewChatButtonAriaLabel) || _defaultReconnectChatPaneControlProps.defaultReconnectChatPaneControlProps.startNewChatButtonAriaLabel
  }))))));
}

var _default = ReconnectChatPane;
exports.default = _default;
