// @microsoft/omnichannel-chat-components/lib/cjs/components/chatbutton/common/defaultStyles/defaultChatButtonStyleProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultChatButtonStyleProps = void 0;

var _defaultChatButtonGeneralStyles = require("./defaultChatButtonGeneralStyles");

var _defaultChatButtonIconContainerStyles = require("./defaultChatButtonIconContainerStyles");

var _defaultChatButtonNotificationBubbleStyles = require("./defaultChatButtonNotificationBubbleStyles");

var _defaultChatButtonSubTitleStyles = require("./defaultChatButtonSubTitleStyles");

var _defaultChatButtonTitleStyles = require("./defaultChatButtonTitleStyles");

var _defaultChatButtonTextContainerStyles = require("./defaultChatButtonTextContainerStyles");

const defaultChatButtonStyleProps = {
  generalStyleProps: _defaultChatButtonGeneralStyles.defaultChatButtonGeneralStyles,
  // Icon
  iconStyleProps: _defaultChatButtonIconContainerStyles.defaultChatButtonIconContainerStyles,
  //Notification Bubble
  notificationBubbleStyleProps: _defaultChatButtonNotificationBubbleStyles.defaultChatButtonNotificationBubbleStyles,
  //Text Container having title and subtitle
  textContainerStyleProps: _defaultChatButtonTextContainerStyles.defaultChatButtonTextContainerStyles,
  titleStyleProps: _defaultChatButtonTitleStyles.defaultChatButtonTitleStyles,
  subtitleStyleProps: _defaultChatButtonSubTitleStyles.defaultChatButtonSubTitleStyles
};
exports.defaultChatButtonStyleProps = defaultChatButtonStyleProps;
