// botframework-webchat-component/lib/LiveRegion/private/LiveRegionSuggestedActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _computeSuggestedActionText = _interopRequireDefault(require("../../Utils/computeSuggestedActionText"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LiveRegionSuggestedActions = function LiveRegionSuggestedActions(_ref) {
  var _suggestedActions$act;

  var suggestedActions = _ref.suggestedActions;
  return ((_suggestedActions$act = suggestedActions.actions) === null || _suggestedActions$act === void 0 ? void 0 : _suggestedActions$act.length) && /*#__PURE__*/_react.default.createElement("p", {
    className: "webchat__live-region-activity__suggested-actions"
  }, suggestedActions.actions.map(function (action, index) {
    return (
      /*#__PURE__*/
      // Direct Line schema does not have key other than index.
      // eslint-disable-next-line react/no-array-index-key
      _react.default.createElement("button", {
        className: "webchat__live-region-activity__suggested-action",
        key: index,
        tabIndex: -1,
        type: "button"
      }, (0, _computeSuggestedActionText.default)(action))
    );
  }));
};

LiveRegionSuggestedActions.propTypes = {
  suggestedActions: _propTypes.default.shape({
    actions: _propTypes.default.array
  }).isRequired
};
var _default = LiveRegionSuggestedActions;
exports.default = _default;
