// botframework-webchat-api/lib/hooks/middleware/createDefaultGroupActivitiesMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createDefaultGroupActivitiesMiddleware;

var _SendStatus = require("../../types/internal/SendStatus");

function bin(items, grouping) {
  var lastBin;
  var bins = [];
  var lastItem;
  items.forEach(function (item) {
    if (lastItem && grouping(lastItem, item)) {
      lastBin.push(item);
    } else {
      lastBin = [item];
      bins.push(lastBin);
    }

    lastItem = item;
  });
  return bins;
}

function sending(activity) {
  if (activity.from.role === 'user') {
    var _activity$channelData = activity.channelData,
        state = _activity$channelData.state,
        sendStatus = _activity$channelData['webchat:send-status']; // `channelData.state` is being deprecated in favor of `channelData['webchat:send-status']`.
    // Please refer to #4362 for details. Remove on or after 2024-07-31.

    return sendStatus || (state === _SendStatus.SENT ? _SendStatus.SENT : _SendStatus.SENDING);
  }
}

function shouldGroupTimestamp(activityX, activityY, groupTimestamp, _ref) {
  var Date = _ref.Date;

  if (groupTimestamp === false) {
    // Hide timestamp for all activities.
    return true;
  } else if (activityX && activityY) {
    if (sending(activityX) !== sending(activityY)) {
      return false;
    }

    groupTimestamp = typeof groupTimestamp === 'number' ? groupTimestamp : Infinity;
    var timeX = new Date(activityX.timestamp).getTime();
    var timeY = new Date(activityY.timestamp).getTime();
    return Math.abs(timeX - timeY) <= groupTimestamp;
  }

  return false;
}

function createDefaultGroupActivitiesMiddleware(_ref2) {
  var groupTimestamp = _ref2.groupTimestamp,
      ponyfill = _ref2.ponyfill;
  return function () {
    return function () {
      return function (_ref3) {
        var activities = _ref3.activities;
        return {
          sender: bin(activities, function (x, y) {
            return x.from.role === y.from.role;
          }),
          status: bin(activities, function (x, y) {
            return shouldGroupTimestamp(x, y, groupTimestamp, ponyfill);
          })
        };
      };
    };
  };
}
