// @microsoft/omnichannel-chat-components/lib/cjs/components/loadingpane/common/presetTwoProps/presetTwoStyles/presetTwoLoadingPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetTwoLoadingPaneStyles = void 0;

var _presetTwoLoadingPaneGeneralStyles = require("./presetTwoLoadingPaneGeneralStyles");

var _presetTwoLoadingPaneSpinnerStyles = require("./presetTwoLoadingPaneSpinnerStyles");

var _defaultLoadingPaneTitleStyles = require("../../defaultProps/defaultStyles/defaultLoadingPaneTitleStyles");

var _defaultLoadingPaneSubtitleStyles = require("../../defaultProps/defaultStyles/defaultLoadingPaneSubtitleStyles");

var _defaultLoadingPaneIconStyles = require("../../defaultProps/defaultStyles/defaultLoadingPaneIconStyles");

var _defaultLoadingPaneIconImageProps = require("../../defaultProps/defaultStyles/defaultLoadingPaneIconImageProps");

var _defaultLoadingPaneSpinnerTextStyles = require("../../defaultProps/defaultStyles/defaultLoadingPaneSpinnerTextStyles");

const presetTwoLoadingPaneStyles = {
  generalStyleProps: _presetTwoLoadingPaneGeneralStyles.presetTwoLoadingPaneGeneralStyles,
  titleStyleProps: _defaultLoadingPaneTitleStyles.defaultLoadingPaneTitleStyles,
  subtitleStyleProps: _defaultLoadingPaneSubtitleStyles.defaultLoadingPaneSubtitleStyles,
  iconStyleProps: _defaultLoadingPaneIconStyles.defaultLoadingPaneIconStyles,
  iconImageProps: _defaultLoadingPaneIconImageProps.defaultLoadingPaneIconImageProps,
  spinnerStyleProps: _presetTwoLoadingPaneSpinnerStyles.presetTwoLoadingPaneSpinnerStyles,
  spinnerTextStyleProps: _defaultLoadingPaneSpinnerTextStyles.defaultLoadingPaneSpinnerTextStyles
};
exports.presetTwoLoadingPaneStyles = presetTwoLoadingPaneStyles;
