// @microsoft/omnichannel-chat-components/lib/cjs/components/callingcontainer/subcomponents/CurrentCall/common/defaultProps/defaultCurrentCallControlPropsRtl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultCurrentCallControlPropsRtl = void 0;

var _Icons = require("../../../../../../assets/Icons");

var _Constants = require("../../../../../../common/Constants");

const defaultCurrentCallControlPropsRtl = {
  id: _Constants.Ids.DefaultCurrentCallId,
  nonActionIds: {
    currentCallActionGroupId: _Constants.Ids.CurrentCallActionGroupId,
    currentCallFooterId: _Constants.Ids.CurrentCallFooterId,
    remoteVideoTileId: _Constants.Ids.RemoteVideoTileId,
    selfVideoTileId: _Constants.Ids.SelfVideoTileId,
    videoTileGroupId: _Constants.Ids.VideoTileGroupId
  },
  hideMicButton: false,
  hideVideoButton: false,
  hideEndCallButton: false,
  videoCallDisabled: false,
  hideCallTimer: false,
  onEndCallClick: function () {
    console.log("end call clicked");
  },
  onMicCallClick: function () {
    console.log("mute clicked");
  },
  onVideoOffClick: function () {
    console.log("video off clicked");
  },
  middleGroup: {
    gap: 1,
    children: []
  },
  leftGroup: {
    gap: 1,
    children: []
  },
  rightGroup: {
    gap: 1,
    children: []
  },
  dir: "rtl",
  endCallButtonProps: {
    id: _Constants.Ids.EndCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.EndCall,
    imageIconProps: {
      src: _Icons.CallRejectButtonBase64,
      styles: {
        image: {
          height: "18px",
          width: "18px"
        }
      }
    }
  },
  micButtonProps: {
    id: _Constants.Ids.MicButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.MicMute,
    toggleAriaLabel: _Constants.AriaLabels.MicUnmute,
    imageIconProps: {
      src: _Icons.VoiceOnIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    imageToggleIconProps: {
      src: _Icons.VoiceOffIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    iconSize: 18
  },
  videoButtonProps: {
    id: _Constants.Ids.VideoButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.VideoTurnCameraOn,
    toggleAriaLabel: _Constants.AriaLabels.VideoTurnCameraOff,
    imageIconProps: {
      src: _Icons.VideoOnIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    imageToggleIconProps: {
      src: _Icons.VideoOffIconBase64,
      styles: {
        image: {
          height: "16px",
          width: "16px"
        }
      }
    },
    iconSize: 18
  },
  callTimerProps: {
    id: _Constants.Ids.CallTimerId
  }
};
exports.defaultCurrentCallControlPropsRtl = defaultCurrentCallControlPropsRtl;
