// @microsoft/omnichannel-chat-components/lib/cjs/components/callingcontainer/subcomponents/CurrentCall/common/defaultProps/defaultCurrentCallControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultCurrentCallControlProps = void 0;

var _Constants = require("../../../../../../common/Constants");

const defaultCurrentCallControlProps = {
  id: _Constants.Ids.DefaultCurrentCallId,
  nonActionIds: {
    currentCallActionGroupId: _Constants.Ids.CurrentCallActionGroupId,
    currentCallFooterId: _Constants.Ids.CurrentCallFooterId,
    remoteVideoTileId: _Constants.Ids.RemoteVideoTileId,
    selfVideoTileId: _Constants.Ids.SelfVideoTileId,
    videoTileGroupId: _Constants.Ids.VideoTileGroupId
  },
  hideMicButton: false,
  hideVideoButton: false,
  hideEndCallButton: false,
  videoCallDisabled: false,
  hideCallTimer: false,
  onEndCallClick: function () {
    console.log("end call clicked");
  },
  onMicCallClick: function () {
    console.log("mute clicked");
  },
  onVideoOffClick: function () {
    console.log("video off clicked");
  },
  middleGroup: {
    gap: 1,
    children: []
  },
  leftGroup: {
    gap: 1,
    children: []
  },
  rightGroup: {
    gap: 1,
    children: []
  },
  endCallButtonProps: {
    id: _Constants.Ids.EndCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.EndCall,
    iconName: _Constants.IconNames.DeclineCall,
    iconSize: 18
  },
  micButtonProps: {
    id: _Constants.Ids.MicButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.MicMute,
    toggleAriaLabel: _Constants.AriaLabels.MicUnmute,
    iconName: _Constants.IconNames.Microphone,
    toggleIconName: _Constants.IconNames.MicOff2,
    iconSize: 18
  },
  videoButtonProps: {
    id: _Constants.Ids.VideoButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.VideoTurnCameraOn,
    toggleAriaLabel: _Constants.AriaLabels.VideoTurnCameraOff,
    iconName: _Constants.IconNames.Video,
    toggleIconName: _Constants.IconNames.VideoOff,
    iconSize: 18
  },
  callTimerProps: {
    id: _Constants.Ids.CallTimerId,
    showHours: false,
    timerStyles: {
      color: "white",
      textAlign: "center",
      backgroundColor: "#3d3c3c",
      height: "45px",
      width: "50px",
      lineHeight: "40px",
      borderRadius: "2px",
      margin: "1px"
    }
  }
};
exports.defaultCurrentCallControlProps = defaultCurrentCallControlProps;
