// @microsoft/ocsdk/lib/Utils/LoggingSanitizer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoggingSanitizer = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _Constants = _interopRequireDefault(require("../Common/Constants"));

var LoggingSanitizer = /*#__PURE__*/function () {
  function LoggingSanitizer() {
    (0, _classCallCheck2["default"])(this, LoggingSanitizer);
  }

  (0, _createClass2["default"])(LoggingSanitizer, null, [{
    key: "stripCustomContextDataValues",
    value: function stripCustomContextDataValues(customContextData) {
      // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
      if (customContextData) {
        var _Object$keys;

        (_Object$keys = Object.keys(customContextData)) === null || _Object$keys === void 0 ? void 0 : _Object$keys.forEach(function (contextKey) {
          var _customContextData$;

          if ((_customContextData$ = customContextData["".concat(contextKey)]) !== null && _customContextData$ !== void 0 && _customContextData$.value) {
            customContextData["".concat(contextKey)].value = _Constants["default"].hiddenContentPlaceholder;
          }
        });
      }
    }
  }, {
    key: "stripPreChatResponse",
    value: function stripPreChatResponse(preChatResponse) {
      // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
      if (preChatResponse) {
        Object.keys(preChatResponse).forEach(function (responseKey) {
          if (preChatResponse["".concat(responseKey)] && responseKey !== 'Type') {
            preChatResponse["".concat(responseKey)] = _Constants["default"].hiddenContentPlaceholder;
          }
        });
      }
    }
  }, {
    key: "stripGeolocation",
    value: function stripGeolocation(data) {
      // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
      if (data) {
        if (Object.keys(data).includes('longitude')) {
          data['longitude'] = _Constants["default"].hiddenContentPlaceholder;
        }

        if (Object.keys(data).includes('latitude')) {
          data['latitude'] = _Constants["default"].hiddenContentPlaceholder;
        }
      }
    }
  }, {
    key: "stripErrorSensitiveProperties",
    value: function stripErrorSensitiveProperties(errorObject) {
      var _Object$keys2,
          _this = this;

      // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
      if (errorObject && (0, _typeof2["default"])(errorObject) === 'object' && ((_Object$keys2 = Object.keys(errorObject)) === null || _Object$keys2 === void 0 ? void 0 : _Object$keys2.length) > 0) {
        var _Object$keys3;

        (_Object$keys3 = Object.keys(errorObject)) === null || _Object$keys3 === void 0 ? void 0 : _Object$keys3.forEach(function (key) {
          if (_Constants["default"].sensitiveProperties.indexOf(key) !== -1) {
            // remove sensitive properties from error object
            delete errorObject["".concat(key)];
          }

          if (key === 'data') {
            var data;

            if (typeof errorObject[key] === 'string') {
              // eslint-disable-line security/detect-object-injection
              try {
                data = JSON.parse(errorObject[key]); // eslint-disable-line security/detect-object-injection
              } catch (_unused) {
                data = undefined;
              }
            }

            if (data) {
              if (Object.keys(data).includes('preChatResponse')) {
                LoggingSanitizer.stripPreChatResponse(data.preChatResponse);
              }

              if (Object.keys(data).includes('customContextData')) {
                LoggingSanitizer.stripCustomContextDataValues(data.customContextData);
              }

              LoggingSanitizer.stripGeolocation(data);
              errorObject[key] = JSON.stringify(data); // eslint-disable-line security/detect-object-injection
            }
          }

          if (errorObject["".concat(key)] !== null && (0, _typeof2["default"])(errorObject["".concat(key)]) === 'object') {
            // check sensitive properties in nested error object
            _this.stripErrorSensitiveProperties(errorObject["".concat(key)]);

            return;
          }
        });
      }
    }
  }]);
  return LoggingSanitizer;
}();

exports.LoggingSanitizer = LoggingSanitizer;
