// @microsoft/omnichannel-chat-components/lib/cjs/components/reconnectchatpane/common/presetThree/presetThreeStyles/presetThreeReconnectChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetThreeReconnectChatPaneStyles = void 0;

var _presetThreeReconnectChatPaneButtonGroupStyles = require("./presetThreeReconnectChatPaneButtonGroupStyles");

var _presetThreeReconnectChatPaneContinueChatButtonStyles = require("./presetThreeReconnectChatPaneContinueChatButtonStyles");

var _presetThreeReconnectChatPaneGeneralStyles = require("./presetThreeReconnectChatPaneGeneralStyles");

var _presetThreeReconnectChatPaneTitleStyles = require("./presetThreeReconnectChatPaneTitleStyles");

var _presetThreeReconnectChatPaneWrapperStyles = require("./presetThreeReconnectChatPaneWrapperStyles");

const presetThreeReconnectChatPaneStyles = {
  generalStyleProps: _presetThreeReconnectChatPaneGeneralStyles.presetThreeReconnectChatPaneGeneralStyles,
  wrapperStyleProps: _presetThreeReconnectChatPaneWrapperStyles.presetThreeReconnectChatPaneWrapperStyles,
  titleStyleProps: _presetThreeReconnectChatPaneTitleStyles.presetThreeReconnectChatPaneTitleStyles,
  buttonGroupStyleProps: _presetThreeReconnectChatPaneButtonGroupStyles.presetThreeReconnectChatPaneButtonGroupStyles,
  continueChatButtonStyleProps: _presetThreeReconnectChatPaneContinueChatButtonStyles.presetThreeReconnectChatPaneContinueChatButtonStyles
};
exports.presetThreeReconnectChatPaneStyles = presetThreeReconnectChatPaneStyles;
