// @microsoft/omnichannel-chat-components/lib/cjs/components/confirmationpane/common/defaultProps/defaultConfirmationPaneControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultConfirmationPaneControlProps = void 0;

var _Constants = require("../../../../common/Constants");

const defaultConfirmationPaneControlProps = {
  id: _Constants.Ids.DefaultConfirmationPaneId,
  dir: "ltr",
  hideConfirmationPane: false,
  hideTitle: false,
  titleText: _Constants.Texts.ConfirmationPaneTitle,
  hideSubtitle: false,
  subtitleText: _Constants.Texts.ConfirmationPaneSubtitle,
  hideConfirmButton: false,
  confirmButtonText: _Constants.Texts.ConfirmButtonText,
  confirmButtonAriaLabel: _Constants.AriaLabels.ConfirmationPaneConfirm,
  hideCancelButton: false,
  cancelButtonText: _Constants.Texts.CancelButtonText,
  cancelButtonAriaLabel: _Constants.AriaLabels.ConfirmationPaneCancel,
  onConfirm: function () {
    console.log("on confirm");
  },
  onCancel: function () {
    console.log("on cancel");
  }
};
exports.defaultConfirmationPaneControlProps = defaultConfirmationPaneControlProps;
