// @microsoft/omnichannel-chat-components/lib/cjs/components/footer/common/defaultProps/customFooterControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.customFooterControlProps = void 0;

var _Icons = require("../../../../assets/Icons");

var _Constants = require("../../../../common/Constants");

const customFooterControlProps = {
  id: _Constants.Ids.CustomFooterId,
  hideDownloadTranscriptButton: false,
  hideEmailTranscriptButton: false,
  hideAudioNotificationButton: false,
  onDownloadTranscriptClick: function () {
    console.log("download transcript clicked");
  },
  onEmailTranscriptClick: function () {
    console.log("email transcript clicked");
  },
  onAudioNotificationClick: function () {
    console.log("audio notification clicked");
  },
  middleGroup: {
    children: []
  },
  leftGroup: {
    children: []
  },
  rightGroup: {
    children: []
  },
  downloadTranscriptButtonProps: {
    id: _Constants.Ids.DownloadTranscriptButtonId,
    type: _Constants.ButtonTypes.Icon,
    imageIconProps: {
      src: _Icons.TranscriptDownloadIconBase64
    },
    ariaLabel: _Constants.AriaLabels.DownloadChatTranscript,
    hideButtonTitle: true
  },
  emailTranscriptButtonProps: {
    id: _Constants.Ids.EmailTranscriptButtonId,
    type: _Constants.ButtonTypes.Icon,
    imageIconProps: {
      src: _Icons.TranscriptEmailIconBase64
    },
    ariaLabel: _Constants.AriaLabels.EmailTranscript,
    hideButtonTitle: true
  },
  audioNotificationButtonProps: {
    id: _Constants.Ids.AudioNotificationButtonId,
    ariaLabel: _Constants.AriaLabels.TurnSoundOff,
    toggleAriaLabel: _Constants.AriaLabels.TurnSoundOn,
    imageIconProps: {
      src: _Icons.AudioNotificationOnIconBase64
    },
    imageToggleIconProps: {
      src: _Icons.AudioNotificationOffIconBase64
    },
    hideButtonTitle: true
  }
};
exports.customFooterControlProps = customFooterControlProps;
