// @microsoft/omnichannel-chat-components/lib/cjs/components/footer/common/defaultProps/defaultFooterControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultFooterControlProps = void 0;

var _Constants = require("../../../../common/Constants");

const defaultFooterControlProps = {
  id: _Constants.Ids.DefaultFooterId,
  hideDownloadTranscriptButton: false,
  hideEmailTranscriptButton: false,
  hideAudioNotificationButton: false,
  onDownloadTranscriptClick: function () {
    console.log("download transcript clicked");
  },
  onEmailTranscriptClick: function () {
    console.log("email transcript clicked");
  },
  onAudioNotificationClick: function () {
    console.log("audio notification clicked");
  },
  middleGroup: {
    children: []
  },
  leftGroup: {
    children: []
  },
  rightGroup: {
    children: []
  },
  downloadTranscriptButtonProps: {
    id: _Constants.Ids.DownloadTranscriptButtonId,
    type: _Constants.ButtonTypes.Icon,
    iconName: _Constants.IconNames.Download,
    ariaLabel: _Constants.AriaLabels.DownloadChatTranscript,
    hideButtonTitle: true
  },
  emailTranscriptButtonProps: {
    id: _Constants.Ids.EmailTranscriptButtonId,
    type: _Constants.ButtonTypes.Icon,
    iconName: _Constants.IconNames.Mail,
    ariaLabel: _Constants.AriaLabels.EmailTranscript,
    hideButtonTitle: true
  },
  audioNotificationButtonProps: {
    id: _Constants.Ids.AudioNotificationButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.TurnSoundOff,
    toggleAriaLabel: _Constants.AriaLabels.TurnSoundOn,
    iconName: _Constants.IconNames.Volume3,
    toggleIconName: _Constants.IconNames.Volume0,
    hideButtonTitle: true
  }
};
exports.defaultFooterControlProps = defaultFooterControlProps;
