// @microsoft/omnichannel-chat-widget/lib/cjs/components/webchatcontainerstateful/webchatcontroller/middlewares/renderingmiddlewares/activityStatusMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.activityStatusMiddleware = void 0;
var _Constants = require("../../../../../common/Constants");
var _DeliveredTimestamp = require("./timestamps/DeliveredTimestamp");
var _NotDeliveredTimestamp = require("./timestamps/NotDeliveredTimestamp");
var _react = _interopRequireDefault(require("react"));
var _SendStatus = require("../../enums/SendStatus");
var _SendingTimestamp = require("./timestamps/SendingTimestamp");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// eslint-disable-next-line @typescript-eslint/no-explicit-any, react/display-name, @typescript-eslint/no-unused-vars
const activityStatusMiddleware = () => next => args => {
  const {
    activity: {
      channelData: {
        tags
      },
      from: {
        name,
        role
      },
      timestamp
    },
    sendState,
    sameTimestampGroup
  } = args;
  const current_tags = tags,
    current_name = name,
    current_role = role,
    current_timestamp = timestamp;
  let sameTimestampGroupTemp = sameTimestampGroup;
  if (args[_Constants.WebChatMiddlewareConstants.nextVisibleActivity]) {
    const {
      nextVisibleActivity: {
        from: {
          name
        },
        timestamp
      }
    } = args;
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const next_name = name,
      next_timestamp = timestamp;
    const timestampCurrentInMillisecond = Date.parse(current_timestamp);
    const timestampNextInMillisecond = Date.parse(next_timestamp);
    if (current_name !== next_name || timestampNextInMillisecond - timestampCurrentInMillisecond > _Constants.WebChatMiddlewareConstants.timeBetweenTimestampGroups) {
      sameTimestampGroupTemp = false;
    }
  }
  if (current_tags && current_tags.includes(_Constants.Constants.systemMessageTag) || sameTimestampGroupTemp && sendState === _SendStatus.SendStatus.Sent) {
    return;
  } else {
    return /*#__PURE__*/_react.default.createElement("span", {
      style: {
        padding: "2px"
      }
    }, sendState === _SendStatus.SendStatus.Sending && /*#__PURE__*/_react.default.createElement(_SendingTimestamp.SendingTimestamp, null), sendState === _SendStatus.SendStatus.SendFailed && /*#__PURE__*/_react.default.createElement(_NotDeliveredTimestamp.NotDeliveredTimestamp, {
      args: args
    }), sendState === _SendStatus.SendStatus.Sent && /*#__PURE__*/_react.default.createElement(_DeliveredTimestamp.DeliveredTimestamp, {
      args: args,
      role: current_role,
      name: current_name
    }))
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;
  }
};
exports.activityStatusMiddleware = activityStatusMiddleware;
