// @microsoft/omnichannel-chat-components/lib/cjs/components/inputvalidationpane/common/presetTwo/presetTwoStyles/presetTwoInputValidationPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetTwoInputValidationPaneStyles = void 0;

var _presetTwoInputValidationPaneTitleStyles = require("./presetTwoInputValidationPaneTitleStyles");

var _presetTwoInputValidationPaneInputStyles = require("./presetTwoInputValidationPaneInputStyles");

var _presetTwoInputValidationPaneSendButtonStyles = require("./presetTwoInputValidationPaneSendButtonStyles");

var _presetTwoInputValidationPaneCancelButtonStyles = require("./presetTwoInputValidationPaneCancelButtonStyles");

var _presetTwoInputValidationPaneHeaderGroupStyles = require("./presetTwoInputValidationPaneHeaderGroupStyles");

const presetTwoInputValidationPaneStyles = {
  headerGroupStyleProps: _presetTwoInputValidationPaneHeaderGroupStyles.presetTwoInputValidationPaneHeaderGroupStyles,
  titleStyleProps: _presetTwoInputValidationPaneTitleStyles.presetTwoInputValidationPaneTitleStyles,
  inputStyleProps: _presetTwoInputValidationPaneInputStyles.presetTwoInputValidationPaneInputStyles,
  sendButtonStyleProps: _presetTwoInputValidationPaneSendButtonStyles.presetTwoInputValidationPaneSendButtonStyles,
  cancelButtonStyleProps: _presetTwoInputValidationPaneCancelButtonStyles.presetTwoInputValidationPaneCancelButtonStyles
};
exports.presetTwoInputValidationPaneStyles = presetTwoInputValidationPaneStyles;
