// react-film/lib/esm/hooks/internal/useObserveScrollLeft.js
import _regeneratorRuntime from "@babel/runtime-corejs3/helpers/regeneratorRuntime";
import _asyncToGenerator from "@babel/runtime-corejs3/helpers/asyncToGenerator";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";
/* eslint no-magic-numbers: ["error", { "ignore": [0, 100] }] */

import memoize from 'memoize-one';
import { useEffect } from 'react';
import debounce from '../../util/debounce';
function sleepZero() {
  return new _Promise(function (resolve) {
    return _setTimeout(function () {
      return resolve();
    });
  });
}
export default function useObserveScrollLeft(callbackRefWithSubscribe, observer) {
  useEffect(function () {
    return callbackRefWithSubscribe.subscribe(function (current) {
      if (!current) {
        return;
      }
      var memoizedEmitValue = memoize(function (initial, fraction, width) {
        return observer && observer({
          initial: initial,
          fraction: fraction,
          width: width
        });
      });
      var emitValue = function emitValue(initial) {
        var offsetWidth = current.offsetWidth,
          scrollLeft = current.scrollLeft,
          scrollWidth = current.scrollWidth;
        memoizedEmitValue(initial, "".concat(scrollWidth === offsetWidth ? 0 : scrollLeft / (scrollWidth - offsetWidth) * 100, "%"), "".concat(offsetWidth / scrollWidth * 100, "%"));
      };
      var handleScroll = function handleScroll() {
        return emitValue(false);
      };
      var handlePointerOver = debounce(function () {
        // We need to send "onScroll" because "scrollWidth" might have changed
        // For example, the container resized, the scroll width will be incorrect
        // We will debounce to prevent "pointerOver" calculating too often
        // We will memoize to prevent firing unnecessary "onScroll"
        emitValue(false);
      });
      current.addEventListener('pointerover', handlePointerOver, {
        passive: true
      });
      current.addEventListener('scroll', handleScroll, {
        passive: true
      });

      // eslint-disable-next-line wrap-iife
      _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (!(current.scrollWidth === current.offsetWidth)) {
                _context.next = 3;
                break;
              }
              _context.next = 3;
              return sleepZero();
            case 3:
              emitValue(true);
            case 4:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }))();
      return function () {
        current.removeEventListener('pointerover', handlePointerOver);
        current.removeEventListener('scroll', handleScroll);
      };
    });
  }, [callbackRefWithSubscribe, observer]);
}
