// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/presetTwo/presetTwoProps/presetTwoProactiveChatPaneComponentOverrides.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetTwoProactiveChatPaneComponentOverrides = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = require("../../../../..");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const customReactNode1 = (0, _.encodeComponentString)( /*#__PURE__*/_react.default.createElement("div", {
  style: {
    color: "rgb(255, 255, 255)",
    float: "right",
    backgroundColor: "green",
    borderRadius: "50px",
    fontSize: "10px",
    height: "20px",
    margin: "10px 10px 0 0",
    padding: "10px",
    width: "20px"
  }
}, "End Chat"));

const customReactNode2 = /*#__PURE__*/_react.default.createElement("button", {
  style: {
    color: "rgb(255, 255, 255)",
    backgroundColor: "green",
    borderRadius: "30px",
    borderColor: "green",
    borderStyle: "solid",
    height: "80px",
    margin: "30px 15px 0 0",
    padding: "10px",
    width: "160px"
  }
}, "This is a custom button");

const presetTwoProactiveChatPaneComponentOverrides = {
  closeButton: customReactNode1,
  bodyTitle: customReactNode2
};
exports.presetTwoProactiveChatPaneComponentOverrides = presetTwoProactiveChatPaneComponentOverrides;
