// @microsoft/omnichannel-chat-widget/lib/cjs/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/dataMaskingMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _TelemetryConstants = require("../../../../../common/telemetry/TelemetryConstants");
var _TelemetryHelper = require("../../../../../common/telemetry/TelemetryHelper");
var _WebChatActionType = require("../../enums/WebChatActionType");
var _isMaskingFromCustomer = require("../../../common/utils/isMaskingFromCustomer");
/******
 * DataMaskingMiddleware
 *
 * Applies data masking rules that Omnichannel Admin has configured.
 ******/

const applyDataMasking = (action, regexCollection) => {
  const maskedChar = "#"; //to do: load from env once OC SDK supporting.
  let {
    text
  } = action.payload;
  if (!regexCollection) {
    _TelemetryHelper.TelemetryHelper.logActionEvent(_TelemetryConstants.LogLevel.ERROR, {
      Event: _TelemetryConstants.TelemetryEvent.InvalidConfiguration,
      ExceptionDetails: {
        ErrorDetails: "Invalid data masking rules"
      }
    });
    return action;
  }
  for (const ruleId of Object.keys(regexCollection)) {
    const item = regexCollection[ruleId];
    if (item) {
      let ruleInfiniteException = false;
      let ruleApplied = false;
      try {
        const regex = new RegExp(item, "gi");
        let match;
        // eslint-disable-next-line no-cond-assign
        while (match = regex.exec(text)) {
          const replaceStr = match[0].replace(/./gi, maskedChar);
          const modifiedText = text.replace(match[0], replaceStr);
          if (modifiedText == text) {
            ruleInfiniteException = true;
            console.warn(`The data masking rule ${item} is ignored because it matches empty strings. Please modify this rule.`);
            break;
          }
          ruleApplied = true;
          text = modifiedText;
        }
      } catch (err) {
        _TelemetryHelper.TelemetryHelper.logActionEvent(_TelemetryConstants.LogLevel.ERROR, {
          Event: _TelemetryConstants.TelemetryEvent.DataMaskingRuleApplyFailed,
          ExceptionDetails: {
            RuleId: ruleId,
            Exception: err
          }
        });
      }
      if (ruleApplied) {
        _TelemetryHelper.TelemetryHelper.logActionEvent(_TelemetryConstants.LogLevel.INFO, {
          Event: _TelemetryConstants.TelemetryEvent.DataMaskingRuleApplied,
          Description: `Data Masking Rule Id: ${ruleId} applied.`
        });
      }
      if (ruleInfiniteException) {
        _TelemetryHelper.TelemetryHelper.logActionEvent(_TelemetryConstants.LogLevel.ERROR, {
          Event: _TelemetryConstants.TelemetryEvent.DataMaskingRuleApplyFailed,
          ExceptionDetails: {
            RuleId: ruleId,
            Exception: "The data masking rule matches empty strings."
          }
        });
      }
    }
  }
  action.payload.text = text;
  return action;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars
const createDataMaskingMiddleware = dataMaskingInfo => _ref => {
  let {
    dispatch
  } = _ref;
  return next => action => {
    var _action$payload;
    if ((0, _isMaskingFromCustomer.isMaskingforCustomer)(dataMaskingInfo) && (_action$payload = action.payload) !== null && _action$payload !== void 0 && _action$payload.text && action.type === _WebChatActionType.WebChatActionType.WEB_CHAT_SEND_MESSAGE) {
      const regexCollection = dataMaskingInfo === null || dataMaskingInfo === void 0 ? void 0 : dataMaskingInfo.dataMaskingRules;
      return next(applyDataMasking(action, regexCollection));
    }
    return next(action);
  };
};
var _default = createDataMaskingMiddleware;
exports.default = _default;
