// @microsoft/omnichannel-chat-components/lib/esm/components/header/subcomponents/MinimizeButton.js
import React from "react";
import { AriaLabels, ElementType, EventNames, IconNames, Texts } from "../../../common/Constants";
import CommandButton from "../../common/commandbutton/CommandButton";

function MinimizeButton(props) {
  const {
    type
  } = props;
  const customEvent = {
    elementType: ElementType.HeaderMinimizeButton,
    elementId: props === null || props === void 0 ? void 0 : props.id,
    eventName: EventNames.OnClick
  };
  return /*#__PURE__*/React.createElement(CommandButton, {
    id: props === null || props === void 0 ? void 0 : props.id,
    type: type,
    text: props.text ?? Texts.MinimizeText,
    styles: props.styles,
    hoverStyles: props.hoverStyles,
    focusStyles: props.focusStyles,
    iconName: props.iconName ?? IconNames.ChromeMinimize,
    imageIconProps: props.imageIconProps,
    onClick: props.onClick,
    ariaLabel: props.ariaLabel ?? AriaLabels.Minimize,
    className: props.className,
    customEvent: customEvent,
    hideButtonTitle: props.hideButtonTitle
  });
}

export default MinimizeButton;
