// @microsoft/omnichannel-chat-components/lib/cjs/components/callingcontainer/subcomponents/IncomingCall/common/defaultProps/defaultIncomingCallControlPropsRtl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultIncomingCallControlPropsRtl = void 0;

var _Icons = require("../../../../../../assets/Icons");

var _Constants = require("../../../../../../common/Constants");

const defaultIncomingCallControlPropsRtl = {
  id: _Constants.Ids.DefaultIncomingCallPopupId,
  ariaLabel: _Constants.AriaLabels.IncomingCallArea,
  hideAudioCall: false,
  hideVideoCall: false,
  hideDeclineCall: false,
  hideIncomingCallTitle: false,
  onDeclineCallClick: function () {
    console.log("decline call clicked");
  },
  onAudioCallClick: function () {
    console.log("audio call clicked");
  },
  onVideoCallClick: function () {
    console.log("video call clicked");
  },
  middleGroup: {
    gap: 5,
    children: []
  },
  leftGroup: {
    gap: 5,
    children: []
  },
  rightGroup: {
    gap: 5,
    children: []
  },
  dir: "rtl",
  declineCallButtonProps: {
    id: _Constants.Ids.DeclineCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.DeclineCall,
    imageIconProps: {
      src: _Icons.CallRejectButtonBase64,
      styles: {
        image: {
          height: "18px",
          width: "18px"
        }
      }
    },
    iconSize: 20
  },
  audioCallButtonProps: {
    id: _Constants.Ids.AudioCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.AudioCall,
    imageIconProps: {
      src: _Icons.CallAcceptButtonBase64,
      styles: {
        image: {
          height: "18px",
          width: "18px"
        }
      }
    },
    iconSize: 20
  },
  videoCallButtonProps: {
    id: _Constants.Ids.VideoCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.VideoCall,
    imageIconProps: {
      src: _Icons.VideoCallAcceptButtonIconBase64,
      styles: {
        image: {
          height: "18px",
          width: "18px"
        }
      }
    },
    iconSize: 20
  },
  incomingCallTitle: {
    id: _Constants.Ids.IncomingCallTitleId,
    text: _Constants.Texts.IncomingCallTitle
  }
};
exports.defaultIncomingCallControlPropsRtl = defaultIncomingCallControlPropsRtl;
