// @microsoft/omnichannel-chat-components/lib/cjs/components/callingcontainer/subcomponents/IncomingCall/common/defaultProps/defaultIncomingCallControlProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultIncomingCallControlProps = void 0;

var _Constants = require("../../../../../../common/Constants");

const defaultIncomingCallControlProps = {
  id: _Constants.Ids.DefaultIncomingCallId,
  dir: "ltr",
  ariaLabel: _Constants.AriaLabels.IncomingCallArea,
  hideAudioCall: false,
  hideVideoCall: false,
  hideDeclineCall: false,
  hideIncomingCallTitle: false,
  onDeclineCallClick: function () {
    console.log("decline call clicked");
  },
  onAudioCallClick: function () {
    console.log("audio call clicked");
  },
  onVideoCallClick: function () {
    console.log("video call clicked");
  },
  middleGroup: {
    gap: 5,
    children: []
  },
  leftGroup: {
    gap: 5,
    children: []
  },
  rightGroup: {
    gap: 5,
    children: []
  },
  declineCallButtonProps: {
    id: _Constants.Ids.DeclineCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.RejectCall,
    iconName: _Constants.IconNames.DeclineCall,
    iconSize: 20
  },
  audioCallButtonProps: {
    id: _Constants.Ids.AudioCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.AcceptVoiceCall,
    iconName: _Constants.IconNames.IncomingCall,
    iconSize: 20
  },
  videoCallButtonProps: {
    id: _Constants.Ids.VideoCallButtonId,
    type: _Constants.ButtonTypes.Icon,
    ariaLabel: _Constants.AriaLabels.AcceptVideoCall,
    iconName: _Constants.IconNames.Video,
    iconSize: 20
  },
  incomingCallTitle: {
    id: _Constants.Ids.IncomingCallTitleId,
    text: _Constants.Texts.IncomingCallTitle
  }
};
exports.defaultIncomingCallControlProps = defaultIncomingCallControlProps;
