// @microsoft/ocsdk/lib/SDKProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _Constants = _interopRequireDefault(require("./Common/Constants"));

var _OCSDKLogger = _interopRequireDefault(require("./Common/OCSDKLogger"));

var _SDK = _interopRequireDefault(require("./SDK"));

var SDKProvider = /*#__PURE__*/function () {
  function SDKProvider() {
    (0, _classCallCheck2["default"])(this, SDKProvider);
  }

  (0, _createClass2["default"])(SDKProvider, null, [{
    key: "getSDK",
    value: function getSDK(omnichannelConfiguration, configuration, logger) {
      var ilogger = new _OCSDKLogger["default"](logger);
      return new _SDK["default"]({
        channelId: omnichannelConfiguration.channelId || _Constants["default"].defaultChannelId,
        orgId: omnichannelConfiguration.orgId,
        orgUrl: omnichannelConfiguration.orgUrl,
        widgetId: omnichannelConfiguration.widgetId
      }, configuration || {}, ilogger);
    }
  }]);
  return SDKProvider;
}();

exports["default"] = SDKProvider;
