// @microsoft/omnichannel-chat-widget/lib/cjs/components/livechatwidget/common/getGeneralStylesForButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getGeneralStylesForButton = void 0;
var _componentController = require("../../../controller/componentController");
var _defaultLiveChatWidgetGeneralStyles = require("./defaultStyles/defaultLiveChatWidgetGeneralStyles");
const getGeneralStylesForButton = state => {
  let generalStylesForButton = _defaultLiveChatWidgetGeneralStyles.defaultLiveChatWidgetGeneralStyles;
  if (!(0, _componentController.shouldShowChatButton)(state) && !(0, _componentController.shouldShowProactiveChatPane)(state)) {
    generalStylesForButton = Object.assign({}, _defaultLiveChatWidgetGeneralStyles.defaultLiveChatWidgetGeneralStyles, {
      boxShadow: "0px 0px 2px rgba(0, 0, 0, 0.16), 0px 4px 8px rgba(0, 0, 0, 0.12)",
      MozBoxShadow: "0px 0px 2px rgba(0, 0, 0, 0.16), 0px 4px 8px rgba(0, 0, 0, 0.12)",
      WebkitBoxShadow: "0px 0px 2px rgba(0, 0, 0, 0.16), 0px 4px 8px rgba(0, 0, 0, 0.12)",
      borderRadius: "4px",
      padding: "0.5"
    });
  }
  return generalStylesForButton;
};
exports.getGeneralStylesForButton = getGeneralStylesForButton;
