// @microsoft/omnichannel-chat-widget/lib/cjs/components/livechatwidget/common/setPostChatContextAndLoadSurvey.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setPostChatContextAndLoadSurvey = void 0;
var _TelemetryConstants = require("../../../common/telemetry/TelemetryConstants");
var _omnichannelChatComponents = require("@microsoft/omnichannel-chat-components");
var _ConversationState = require("../../../contexts/common/ConversationState");
var _LiveChatWidgetActionType = require("../../../contexts/common/LiveChatWidgetActionType");
var _TelemetryHelper = require("../../../common/telemetry/TelemetryHelper");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const setPostChatContextAndLoadSurvey = async (chatSDK, dispatch, persistedChat) => {
  try {
    if (!persistedChat) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const context = await chatSDK.getPostChatSurveyContext();
      _TelemetryHelper.TelemetryHelper.logSDKEvent(_TelemetryConstants.LogLevel.INFO, {
        Event: _TelemetryConstants.TelemetryEvent.PostChatContextCallSucceed,
        Description: "Postchat context call succeed."
      });
      dispatch({
        type: _LiveChatWidgetActionType.LiveChatWidgetActionType.SET_POST_CHAT_CONTEXT,
        payload: context
      });
    }
  } catch (ex) {
    _TelemetryHelper.TelemetryHelper.logSDKEvent(_TelemetryConstants.LogLevel.ERROR, {
      Event: _TelemetryConstants.TelemetryEvent.PostChatContextCallFailed,
      ExceptionDetails: {
        exception: ex
      }
    });
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _omnichannelChatComponents.BroadcastService.getMessageByEventName("LoadPostChatSurvey").subscribe(msg => {
    dispatch({
      type: _LiveChatWidgetActionType.LiveChatWidgetActionType.SET_CONVERSATION_STATE,
      payload: _ConversationState.ConversationState.Postchat
    });
  });
};
exports.setPostChatContextAndLoadSurvey = setPostChatContextAndLoadSurvey;
