// @microsoft/omnichannel-chat-widget/lib/esm/components/postchatloadingpanestateful/PostChatLoadingPaneStateful.js
import React, { useEffect } from "react";
import { LoadingPane } from "@microsoft/omnichannel-chat-components";
import { defaultGeneralPostChatLoadingPaneStyleProps } from "./common/defaultgeneralPostChatLoadingPaneStyleProps";
import { findAllFocusableElement } from "../../common/utils";
import useChatContextStore from "../../hooks/useChatContextStore";
import { TelemetryHelper } from "../../common/telemetry/TelemetryHelper";
import { LogLevel, TelemetryEvent } from "../../common/telemetry/TelemetryConstants";
export const PostChatLoadingPaneStateful = props => {
  var _props$styleProps;
  const [state] = useChatContextStore();
  const generalStyleProps = Object.assign({}, defaultGeneralPostChatLoadingPaneStyleProps, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps);
  const styleProps = {
    ...props.styleProps,
    generalStyleProps: generalStyleProps
  };
  const controlProps = {
    id: "oc-lcw-postchatloading-pane",
    dir: state.domainStates.globalDir,
    hideIcon: true,
    hideTitle: true,
    hideSpinner: true,
    hideSpinnerText: true,
    subtitleText: "Please take a moment to give us feedback about your chat experience. We are loading the survey for you now.",
    ...props.controlProps
  };

  // Move focus to the first button
  useEffect(() => {
    const firstElement = findAllFocusableElement(`#${state.domainStates.widgetElementId}`);
    if (firstElement && firstElement[0]) {
      firstElement[0].focus();
    }
    TelemetryHelper.logLoadingEvent(LogLevel.INFO, {
      Event: TelemetryEvent.PostChatSurveyLoadingPaneLoaded
    });
  }, []);
  return /*#__PURE__*/React.createElement(LoadingPane, {
    componentOverrides: props.componentOverrides,
    controlProps: controlProps,
    styleProps: styleProps
  });
};
export default PostChatLoadingPaneStateful;
