// botframework-webchat-component/lib/Styles/StyleSet/CarouselFilmStripAttachment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CarouselFilmStripAttachment;

var _mirrorStyle = _interopRequireDefault(require("../mirrorStyle"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function CarouselFilmStripAttachment(_ref) {
  var avatarSize = _ref.avatarSize,
      bubbleMaxWidth = _ref.bubbleMaxWidth,
      bubbleMinWidth = _ref.bubbleMinWidth,
      paddingRegular = _ref.paddingRegular,
      transcriptVisualKeyboardIndicatorColor = _ref.transcriptVisualKeyboardIndicatorColor,
      transcriptVisualKeyboardIndicatorStyle = _ref.transcriptVisualKeyboardIndicatorStyle,
      transcriptVisualKeyboardIndicatorWidth = _ref.transcriptVisualKeyboardIndicatorWidth,
      transitionDuration = _ref.transitionDuration;
  return _objectSpread({
    '&.webchat__carousel-filmstrip-attachment': {
      minWidth: bubbleMinWidth,
      maxWidth: bubbleMaxWidth,
      transitionDuration: transitionDuration,
      transitionProperty: 'max-width, min-width',
      '&:focus': {
        outline: 0
      },
      '&:focus .webchat__carousel-filmstrip-attachment--focus': {
        borderColor: transcriptVisualKeyboardIndicatorColor,
        borderStyle: transcriptVisualKeyboardIndicatorStyle,
        borderWidth: transcriptVisualKeyboardIndicatorWidth,
        boxSizing: 'border-box',
        height: "calc(100% - ".concat(transcriptVisualKeyboardIndicatorWidth, "px)"),
        left: 0,
        pointerEvents: 'none',
        position: 'absolute',
        top: 0,
        width: "calc(100% - ".concat(transcriptVisualKeyboardIndicatorWidth, "px)")
      }
    }
  }, (0, _mirrorStyle.default)('&.webchat__carousel-filmstrip-attachment--rtl', {
    paddingLeft: paddingRegular,
    '&.webchat__carousel-filmstrip-attachment--hide-avatar, &.webchat__carousel-filmstrip-attachment--show-avatar': {
      '&:first-child': {
        paddingLeft: avatarSize + paddingRegular * 2
      }
    },
    '&.webchat__carousel-filmstrip-attachment--hide-nub, &.webchat__carousel-filmstrip-attachment--show-nub': {
      '&:not(.webchat__carousel-filmstrip-attachment--hide-avatar.webchat__carousel-filmstrip-attachment--show-avatar):first-child': {
        paddingLeft: paddingRegular * 2
      }
    }
  }));
}
