// botframework-webchat-component/lib/Styles/StyleSet/TextContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createTextContentStyle;

/* eslint no-magic-numbers: "off" */
function createTextContentStyle(_ref) {
  var bubbleMaxWidth = _ref.bubbleMaxWidth,
      bubbleMinHeight = _ref.bubbleMinHeight,
      markdownExternalLinkIconImage = _ref.markdownExternalLinkIconImage,
      primaryFont = _ref.primaryFont,
      paddingRegular = _ref.paddingRegular;
  return {
    fontFamily: primaryFont,
    margin: 0,
    minHeight: bubbleMinHeight - paddingRegular * 2,
    padding: paddingRegular,
    '& > :first-child': {
      marginTop: 0
    },
    '& > :last-child': {
      marginBottom: 0
    },
    '&.markdown': {
      '& img:not(.webchat__markdown__external-link-icon)': {
        maxWidth: bubbleMaxWidth,
        width: '100%'
      },
      '& img.webchat__markdown__external-link-icon': {
        backgroundImage: markdownExternalLinkIconImage,
        height: '.75em',
        marginLeft: '.25em'
      },
      '& pre': {
        overflow: 'hidden'
      }
    }
  };
}
