// botframework-webchat-core/lib/sagas/postActivitySaga.js
"use strict";

var _regeneratorRuntime2 = require("@babel/runtime/regenerator");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = postActivitySaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _effects = require("redux-saga/effects");

var _incomingActivity = require("../actions/incomingActivity");

var _postActivity = require("../actions/postActivity");

var _dateToLocaleISOString = _interopRequireDefault(require("../utils/dateToLocaleISOString"));

var _deleteKey = _interopRequireDefault(require("../utils/deleteKey"));

var _language = _interopRequireDefault(require("../selectors/language"));

var _observeOnce = _interopRequireDefault(require("./effects/observeOnce"));

var _sendTimeout = _interopRequireDefault(require("../selectors/sendTimeout"));

var _sleep = _interopRequireDefault(require("../utils/sleep"));

var _uniqueID = _interopRequireDefault(require("../utils/uniqueID"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regeneratorRuntime2.mark(postActivity),
    _marked2 = /*#__PURE__*/_regeneratorRuntime2.mark(postActivitySaga);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

// After 5 minutes, the saga will stop from listening for echo backs and consider the outgoing message as permanently undeliverable.
// This value must be equals to or larger than the user-defined `styleOptions.sendTimeout`.
var HARD_SEND_TIMEOUT = 300000;

function postActivity(directLine, userID, username, numActivitiesPosted, _ref, ponyfill) {
  var method, activity, attachments, clientActivityID, locale, localTimeZone, now, outgoingActivity, meta, echoed, echoBackCall, sendTimeout, _yield$race, echoBack;

  return _regenerator["default"].wrap(function postActivity$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          method = _ref.meta.method, activity = _ref.payload.activity;
          attachments = activity.type === 'message' && activity.attachments || [];
          clientActivityID = (0, _uniqueID["default"])();
          _context3.next = 5;
          return (0, _effects.select)(_language["default"]);

        case 5:
          locale = _context3.sent;
          localTimeZone = typeof window.Intl === 'undefined' ? undefined : new Intl.DateTimeFormat().resolvedOptions().timeZone;
          now = new ponyfill.Date(); // Currently, we allow untyped outgoing activity as long as the chat adapter can deliver.
          // In the future, we should warn if the outgoing activity is not matching the type.

          outgoingActivity = _objectSpread(_objectSpread({}, (0, _deleteKey["default"])(activity, 'id')), {}, {
            channelData: _objectSpread(_objectSpread({}, (0, _deleteKey["default"])(activity.channelData, 'state', 'webchat:send-status')), {}, {
              clientActivityID: clientActivityID
            }),
            channelId: 'webchat',
            from: {
              id: userID,
              name: username,
              role: 'user'
            },
            locale: locale,
            localTimestamp: (0, _dateToLocaleISOString["default"])(now),
            localTimezone: localTimeZone
          }, activity.type === 'message' ? {
            attachments: attachments && attachments.map(function (_ref2) {
              var contentType = _ref2.contentType,
                  contentUrl = _ref2.contentUrl,
                  name = _ref2.name,
                  thumbnailUrl = _ref2.thumbnailUrl;
              return {
                contentType: contentType,
                contentUrl: contentUrl,
                name: name,
                thumbnailUrl: thumbnailUrl
              };
            }),
            text: activity.text
          } : activity.type === 'event' ? {
            name: activity.name,
            value: activity.value
          } : {});

          if (!numActivitiesPosted) {
            outgoingActivity.entities = [].concat((0, _toConsumableArray2["default"])(outgoingActivity.entities || []), [{
              // TODO: [P4] Currently in v3, we send the capabilities although the client might not actually have them
              //       We need to understand why we need to send these, and only send capabilities the client have
              requiresBotState: true,
              supportsListening: true,
              supportsTts: true,
              type: 'ClientCapabilities'
            }]);
          }

          meta = {
            clientActivityID: clientActivityID,
            method: method
          };
          _context3.next = 13;
          return (0, _effects.put)({
            type: _postActivity.POST_ACTIVITY_PENDING,
            meta: meta,
            payload: {
              activity: outgoingActivity
            }
          });

        case 13:
          _context3.prev = 13;
          // Quirks: We might receive INCOMING_ACTIVITY before the postActivity call completed
          //         So, we setup expectation first, then postActivity afterward
          echoBackCall = (0, _effects.call)( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
            var _activity$channelData, _yield$take, _activity;

            return _regenerator["default"].wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.next = 2;
                    return (0, _effects.take)(_incomingActivity.INCOMING_ACTIVITY);

                  case 2:
                    _yield$take = _context.sent;
                    _activity = _yield$take.payload.activity;

                    if (!(((_activity$channelData = _activity.channelData) === null || _activity$channelData === void 0 ? void 0 : _activity$channelData.clientActivityID) === clientActivityID && _activity.id)) {
                      _context.next = 7;
                      break;
                    }

                    echoed = true;
                    return _context.abrupt("return", _activity);

                  case 7:
                    _context.next = 0;
                    break;

                  case 9:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          })); // Timeout could be due to either:
          // - Post activity call may take too long time to complete
          //   - Direct Line service only respond on HTTP after bot respond to Direct Line
          // - Activity may take too long time to echo back

          _context3.next = 17;
          return (0, _effects.select)(_sendTimeout["default"]);

        case 17:
          sendTimeout = _context3.sent;
          _context3.next = 20;
          return (0, _effects.race)({
            send: (0, _effects.all)({
              echoBack: echoBackCall,
              postActivity: (0, _observeOnce["default"])(directLine.postActivity(outgoingActivity))
            }),
            timeout: (0, _effects.call)( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
              return _regenerator["default"].wrap(function _callee2$(_context2) {
                while (1) {
                  switch (_context2.prev = _context2.next) {
                    case 0:
                      _context2.next = 2;
                      return (0, _effects.call)(_sleep["default"], sendTimeout, ponyfill);

                    case 2:
                      _context2.next = 4;
                      return (0, _effects.put)({
                        type: _postActivity.POST_ACTIVITY_IMPEDED,
                        meta: meta,
                        payload: {
                          activity: outgoingActivity
                        }
                      });

                    case 4:
                      _context2.next = 6;
                      return (0, _effects.call)(_sleep["default"], HARD_SEND_TIMEOUT - sendTimeout, ponyfill);

                    case 6:
                      throw !echoed ? new Error('timed out while waiting for outgoing message to echo back') : new Error('timed out while waiting for postActivity to return any values');

                    case 7:
                    case "end":
                      return _context2.stop();
                  }
                }
              }, _callee2);
            }))
          });

        case 20:
          _yield$race = _context3.sent;
          echoBack = _yield$race.send.echoBack;
          _context3.next = 24;
          return (0, _effects.put)({
            type: _postActivity.POST_ACTIVITY_FULFILLED,
            meta: meta,
            payload: {
              activity: echoBack
            }
          });

        case 24:
          _context3.next = 31;
          break;

        case 26:
          _context3.prev = 26;
          _context3.t0 = _context3["catch"](13);
          console.error('botframework-webchat: Failed to post activity to chat adapter.', _context3.t0);
          _context3.next = 31;
          return (0, _effects.put)({
            type: _postActivity.POST_ACTIVITY_REJECTED,
            error: true,
            meta: meta,
            payload: _context3.t0
          });

        case 31:
          _context3.prev = 31;
          _context3.next = 34;
          return (0, _effects.cancelled)();

        case 34:
          if (!_context3.sent) {
            _context3.next = 37;
            break;
          }

          _context3.next = 37;
          return (0, _effects.put)({
            type: _postActivity.POST_ACTIVITY_REJECTED,
            error: true,
            meta: meta,
            payload: new Error('cancelled')
          });

        case 37:
          return _context3.finish(31);

        case 38:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked, null, [[13, 26, 31, 38]]);
}

function postActivitySaga(ponyfill) {
  return _regenerator["default"].wrap(function postActivitySaga$(_context6) {
    while (1) {
      switch (_context6.prev = _context6.next) {
        case 0:
          _context6.next = 2;
          return (0, _whileConnected["default"])( /*#__PURE__*/_regenerator["default"].mark(function postActivityWhileConnected(_ref3) {
            var directLine, userID, username, numActivitiesPosted;
            return _regenerator["default"].wrap(function postActivityWhileConnected$(_context5) {
              while (1) {
                switch (_context5.prev = _context5.next) {
                  case 0:
                    directLine = _ref3.directLine, userID = _ref3.userID, username = _ref3.username;
                    numActivitiesPosted = 0;
                    _context5.next = 4;
                    return (0, _effects.takeEvery)(_postActivity.POST_ACTIVITY, /*#__PURE__*/_regenerator["default"].mark(function postActivityWrapper(action) {
                      return _regenerator["default"].wrap(function postActivityWrapper$(_context4) {
                        while (1) {
                          switch (_context4.prev = _context4.next) {
                            case 0:
                              return _context4.delegateYield(postActivity(directLine, userID, username, numActivitiesPosted++, action, ponyfill), "t0", 1);

                            case 1:
                            case "end":
                              return _context4.stop();
                          }
                        }
                      }, postActivityWrapper);
                    }));

                  case 4:
                  case "end":
                    return _context5.stop();
                }
              }
            }, postActivityWhileConnected);
          }));

        case 2:
        case "end":
          return _context6.stop();
      }
    }
  }, _marked2);
}
