// botframework-webchat-core/lib/sagas/sendFilesToPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = sendFilesToPostActivitySaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _mime = _interopRequireDefault(require("mime"));

var _sendFiles = require("../actions/sendFiles");

var _postActivity = _interopRequireDefault(require("../actions/postActivity"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regenerator["default"].mark(postActivityWithFiles),
    _marked2 = /*#__PURE__*/_regenerator["default"].mark(sendFilesToPostActivity),
    _marked3 = /*#__PURE__*/_regenerator["default"].mark(sendFilesToPostActivitySaga);

var getType = _mime["default"].getType.bind(_mime["default"]);

function postActivityWithFiles(_ref) {
  var files;
  return _regenerator["default"].wrap(function postActivityWithFiles$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          files = _ref.payload.files;
          _context.next = 3;
          return (0, _effects.put)((0, _postActivity["default"])({
            attachments: [].map.call(files, function (_ref2) {
              var name = _ref2.name,
                  thumbnail = _ref2.thumbnail,
                  url = _ref2.url;
              return {
                contentType: getType(name) || 'application/octet-stream',
                contentUrl: url,
                name: name,
                thumbnailUrl: thumbnail
              };
            }),
            channelData: {
              attachmentSizes: [].map.call(files, function (_ref3) {
                var size = _ref3.size;
                return size;
              })
            },
            type: 'message'
          }));

        case 3:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function sendFilesToPostActivity() {
  return _regenerator["default"].wrap(function sendFilesToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref4) {
            var payload = _ref4.payload,
                type = _ref4.type;
            return type === _sendFiles.SEND_FILES && payload.files.length;
          }, postActivityWithFiles);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function sendFilesToPostActivitySaga() {
  return _regenerator["default"].wrap(function sendFilesToPostActivitySaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(sendFilesToPostActivity);

        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
