// botframework-webchat-core/lib/sagas/stopDictateOnCardActionSaga.js
"use strict";

var _regeneratorRuntime2 = require("@babel/runtime/regenerator");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = stopDictateOnCardActionSaga;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _postActivity = require("../actions/postActivity");

var _stopDictate = _interopRequireDefault(require("../actions/stopDictate"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked = /*#__PURE__*/_regeneratorRuntime2.mark(stopDictateOnCardAction),
    _marked2 = /*#__PURE__*/_regeneratorRuntime2.mark(stopDictateOnCardActionSaga);

function stopDictateOnCardAction() {
  return _regenerator["default"].wrap(function stopDictateOnCardAction$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)( // Currently, there are no actions that are related to card input
          // For now, we are using POST_ACTIVITY of a "message" activity
          // In the future, if we have an action for card input, we should use that instead
          function (_ref) {
            var payload = _ref.payload,
                type = _ref.type;
            return type === _postActivity.POST_ACTIVITY_PENDING && payload.activity.type === 'message';
          }, /*#__PURE__*/_regenerator["default"].mark(function putStopDictate() {
            return _regenerator["default"].wrap(function putStopDictate$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.next = 2;
                    return (0, _effects.put)((0, _stopDictate["default"])());

                  case 2:
                  case "end":
                    return _context.stop();
                }
              }
            }, putStopDictate);
          }));

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked);
}

function stopDictateOnCardActionSaga() {
  return _regenerator["default"].wrap(function stopDictateOnCardActionSaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(stopDictateOnCardAction);

        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked2);
}
