// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/presetOne/presetOneStyles/presetOneProactiveChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.presetOneProactiveChatPaneStyles = void 0;

var _presetOneProactiveChatPaneBodyTitleStyles = require("./presetOneProactiveChatPaneBodyTitleStyles");

var _presetOneProactiveChatPaneCloseButtonStyles = require("./presetOneProactiveChatPaneCloseButtonStyles");

var _presetOneProactiveChatPaneStartButtonStyles = require("./presetOneProactiveChatPaneStartButtonStyles");

var _presetOneProactiveChatPaneStartButtonHoveredStyles = require("./presetOneProactiveChatPaneStartButtonHoveredStyles");

var _presetOneProactiveChatPaneCloseButtonHoveredStyles = require("./presetOneProactiveChatPaneCloseButtonHoveredStyles");

const presetOneProactiveChatPaneStyles = {
  closeButtonStyleProps: _presetOneProactiveChatPaneCloseButtonStyles.presetOneProactiveChatPaneCloseButtonStyles,
  closeButtonHoveredStyleProps: _presetOneProactiveChatPaneCloseButtonHoveredStyles.presetOneProactiveChatPaneCloseButtonHoveredStyles,
  bodyTitleStyleProps: _presetOneProactiveChatPaneBodyTitleStyles.presetOneProactiveChatPaneBodyTitleStyles,
  startButtonStyleProps: _presetOneProactiveChatPaneStartButtonStyles.presetOneProactiveChatPaneStartButtonStyles,
  startButtonHoveredStyleProps: _presetOneProactiveChatPaneStartButtonHoveredStyles.presetOneProactiveChatPaneStartButtonHoveredStyles
};
exports.presetOneProactiveChatPaneStyles = presetOneProactiveChatPaneStyles;
