// @microsoft/omnichannel-chat-components/lib/cjs/components/proactivechatpane/common/default/defaultStyles/defaultProactiveChatPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultProactiveChatPaneStyles = void 0;

var _defaultProactiveChatPaneSubtitleStyles = require("./defaultProactiveChatPaneSubtitleStyles");

var _defaultProactiveChatPaneBodyContainerStyles = require("./defaultProactiveChatPaneBodyContainerStyles");

var _defaultProactiveChatPaneBodyTitleStyles = require("./defaultProactiveChatPaneBodyTitleStyles");

var _defaultProactiveChatPaneGeneralStyles = require("./defaultProactiveChatPaneGeneralStyles");

var _defaultProactiveChatPaneTitleStyles = require("./defaultProactiveChatPaneTitleStyles");

var _defaultProactiveChatPaneCloseButtonStyles = require("./defaultProactiveChatPaneCloseButtonStyles");

var _defaultProactiveChatPaneHeaderContainerStyles = require("./defaultProactiveChatPaneHeaderContainerStyles");

var _defaultProactiveChatPaneTextContainerStyles = require("./defaultProactiveChatPaneTextContainerStyles");

var _defaultProactiveChatPaneStartButtonStyles = require("./defaultProactiveChatPaneStartButtonStyles");

var _defaultProactiveChatPaneStartButtonHoveredStyles = require("./defaultProactiveChatPaneStartButtonHoveredStyles");

var _defaultProactiveChatPaneCloseButtonHoveredStyles = require("./defaultProactiveChatPaneCloseButtonHoveredStyles");

const defaultProactiveChatPaneStyles = {
  generalStyleProps: _defaultProactiveChatPaneGeneralStyles.defaultProactiveChatPaneGeneralStyles,
  headerContainerStyleProps: _defaultProactiveChatPaneHeaderContainerStyles.defaultProactiveChatPaneHeaderContainerStyles,
  textContainerStyleProps: _defaultProactiveChatPaneTextContainerStyles.defaultProactiveChatPaneTextContainerStyles,
  titleStyleProps: _defaultProactiveChatPaneTitleStyles.defaultProactiveChatPaneTitleStyles,
  subtitleStyleProps: _defaultProactiveChatPaneSubtitleStyles.defaultProactiveChatPaneSubtitleStyles,
  closeButtonStyleProps: _defaultProactiveChatPaneCloseButtonStyles.defaultProactiveChatPaneCloseButtonStyles,
  closeButtonHoveredStyleProps: _defaultProactiveChatPaneCloseButtonHoveredStyles.defaultProactiveChatPaneCloseButtonHoveredStyles,
  bodyContainerStyleProps: _defaultProactiveChatPaneBodyContainerStyles.defaultProactiveChatPaneBodyContainerStyles,
  bodyTitleStyleProps: _defaultProactiveChatPaneBodyTitleStyles.defaultProactiveChatPaneBodyTitleStyles,
  startButtonStyleProps: _defaultProactiveChatPaneStartButtonStyles.defaultProactiveChatPaneStartButtonStyles,
  startButtonHoveredStyleProps: _defaultProactiveChatPaneStartButtonHoveredStyles.defaultProactiveChatPaneStartButtonHoveredStyles
};
exports.defaultProactiveChatPaneStyles = defaultProactiveChatPaneStyles;
