// @microsoft/omnichannel-chat-components/lib/cjs/components/inputvalidationpane/common/default/defaultStyles/defaultInputValidationPaneStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultInputValidationPaneStyles = void 0;

var _defaultInputValidationPaneSubtitleStyles = require("./defaultInputValidationPaneSubtitleStyles");

var _defaultInputValidationPaneButtonGroupStyles = require("./defaultInputValidationPaneButtonGroupStyles");

var _defaultInputValidationPaneGeneralStyles = require("./defaultInputValidationPaneGeneralStyles");

var _defaultInputValidationPaneTitleStyles = require("./defaultInputValidationPaneTitleStyles");

var _defaultInputValidationPaneInputStyles = require("./defaultInputValidationPaneInputStyles");

var _defaultInputValidationPaneSendButtonStyles = require("./defaultInputValidationPaneSendButtonStyles");

var _defaultInputValidationPaneSendButtonHoveredStyles = require("./defaultInputValidationPaneSendButtonHoveredStyles");

var _defaultInputValidationPaneCancelButtonStyles = require("./defaultInputValidationPaneCancelButtonStyles");

var _defaultInputValidationPaneCancelButtonHoveredStyles = require("./defaultInputValidationPaneCancelButtonHoveredStyles");

var _defaultInputValidationPaneHeaderGroupStyles = require("./defaultInputValidationPaneHeaderGroupStyles");

var _defaultInputValidationPaneInvalidInputErrorMessageStyles = require("./defaultInputValidationPaneInvalidInputErrorMessageStyles");

const defaultInputValidationPaneStyles = {
  generalStyleProps: _defaultInputValidationPaneGeneralStyles.defaultInputValidationPaneGeneralStyles,
  headerGroupStyleProps: _defaultInputValidationPaneHeaderGroupStyles.defaultInputValidationPaneHeaderGroupStyles,
  titleStyleProps: _defaultInputValidationPaneTitleStyles.defaultInputValidationPaneTitleStyles,
  subtitleStyleProps: _defaultInputValidationPaneSubtitleStyles.defaultInputValidationPaneSubtitleStyles,
  inputStyleProps: _defaultInputValidationPaneInputStyles.defaultInputValidationPaneInputStyles,
  invalidInputErrorMessageStyleProps: _defaultInputValidationPaneInvalidInputErrorMessageStyles.defaultInputValidationPaneInvalidInputErrorMessageStyles,
  buttonGroupStyleProps: _defaultInputValidationPaneButtonGroupStyles.defaultInputValidationPaneButtonGroupStyles,
  sendButtonStyleProps: _defaultInputValidationPaneSendButtonStyles.defaultInputValidationPaneSendButtonStyles,
  sendButtonHoveredStyleProps: _defaultInputValidationPaneSendButtonHoveredStyles.defaultInputValidationPaneSendButtonHoveredStyles,
  cancelButtonStyleProps: _defaultInputValidationPaneCancelButtonStyles.defaultInputValidationPaneCancelButtonStyles,
  cancelButtonHoveredStyleProps: _defaultInputValidationPaneCancelButtonHoveredStyles.defaultInputValidationPaneCancelButtonHoveredStyles
};
exports.defaultInputValidationPaneStyles = defaultInputValidationPaneStyles;
