// @microsoft/ocsdk/lib/Common/Enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SDKError = exports.LiveChatVersion = exports.OCSDKTelemetryEvent = exports.ChannelId = exports.OperatingSystem = exports.DeviceType = exports.Browser = exports.BrowserVendor = void 0;
var BrowserVendor;
exports.BrowserVendor = BrowserVendor;

(function (BrowserVendor) {
  BrowserVendor["IE"] = "IE";
  BrowserVendor["CHROME"] = "Chrome";
  BrowserVendor["FIREFOX"] = "Firefox";
  BrowserVendor["SAFARI"] = "Safari";
  BrowserVendor["EDGE"] = "Edge";
  BrowserVendor["EDGE_CHROMIUM"] = "Edg";
  BrowserVendor["OPERA"] = "Opera";
  BrowserVendor["UNKNOWN"] = "Unknown";
})(BrowserVendor || (exports.BrowserVendor = BrowserVendor = {}));

var Browser;
exports.Browser = Browser;

(function (Browser) {
  Browser[Browser["Chrome"] = 0] = "Chrome";
  Browser[Browser["Firefox"] = 1] = "Firefox";
  Browser[Browser["Ie"] = 2] = "Ie";
  Browser[Browser["Edge"] = 3] = "Edge";
  Browser[Browser["Safari"] = 4] = "Safari";
  Browser[Browser["None"] = 5] = "None";
})(Browser || (exports.Browser = Browser = {}));

var DeviceType;
exports.DeviceType = DeviceType;

(function (DeviceType) {
  DeviceType["Desktop"] = "Desktop";
  DeviceType["Mobile"] = "Mobile";
  DeviceType["Tablet"] = "Tablet";
})(DeviceType || (exports.DeviceType = DeviceType = {}));

var OperatingSystem;
exports.OperatingSystem = OperatingSystem;

(function (OperatingSystem) {
  OperatingSystem["WINDOWS"] = "Windows";
  OperatingSystem["MACOSX"] = "Mac OS X";
  OperatingSystem["WINDOWS_PHONE"] = "Windows Phone";
  OperatingSystem["WINDOWS_RT"] = "Windows RT";
  OperatingSystem["IOS"] = "iOS";
  OperatingSystem["ANDROID"] = "Android";
  OperatingSystem["LINUX"] = "Linux";
  OperatingSystem["UNKNOWN"] = "Unknown";
})(OperatingSystem || (exports.OperatingSystem = OperatingSystem = {}));

var ChannelId;
exports.ChannelId = ChannelId;

(function (ChannelId) {
  ChannelId["LCW"] = "lcw";
  ChannelId["SMS"] = "sms";
  ChannelId["FACEBOOK"] = "facebook";
  ChannelId["WECHAT"] = "wechat";
})(ChannelId || (exports.ChannelId = ChannelId = {}));

var OCSDKTelemetryEvent;
exports.OCSDKTelemetryEvent = OCSDKTelemetryEvent;

(function (OCSDKTelemetryEvent) {
  OCSDKTelemetryEvent["GETCHATCONFIG"] = "GetChatConfig";
  OCSDKTelemetryEvent["GETCHATCONFIGSUCCESS"] = "GetChatConfigSucceeded";
  OCSDKTelemetryEvent["GETRECONNECTABLECHATS"] = "GetReconnnectableChats";
  OCSDKTelemetryEvent["GETRECONNECTAVAILABILITY"] = "GetReconnnectableAvailability";
  OCSDKTelemetryEvent["GETCHATTOKENSTARTED"] = "GetChatTokenStarted";
  OCSDKTelemetryEvent["GETCHATTOKENSUCCEEDED"] = "GetChatTokenSucceeded";
  OCSDKTelemetryEvent["GETCHATTOKENFAILED"] = "GetChatTokenFailed";
  OCSDKTelemetryEvent["GETLWISTATUSSTARTED"] = "GetLwiStatusStarted";
  OCSDKTelemetryEvent["GETLWISTATUSSUCCEEDED"] = "GetLwiStatusSucceeded";
  OCSDKTelemetryEvent["GETLWISTATUSFAILED"] = "GetLwiStatusFailed";
  OCSDKTelemetryEvent["SESSIONINITSTARTED"] = "SessionInitStarted";
  OCSDKTelemetryEvent["SESSIONINITSUCCEEDED"] = "SessionInitSucceeded";
  OCSDKTelemetryEvent["SESSIONINITFAILED"] = "SessionInitFailed";
  OCSDKTelemetryEvent["SESSIONCLOSESTARTED"] = "SessionCloseStarted";
  OCSDKTelemetryEvent["SESSIONCLOSESUCCEEDED"] = "SessionCloseSucceeded";
  OCSDKTelemetryEvent["SESSIONCLOSEFAILED"] = "SessionCloseFailed";
  OCSDKTelemetryEvent["SUBMITPOSTCHATSTARTED"] = "SubmitPostChatStarted";
  OCSDKTelemetryEvent["SUBMITPOSTCHATSUCCEEDED"] = "SubmitPostChatSucceeded";
  OCSDKTelemetryEvent["SUBMITPOSTCHATFAILED"] = "SubmitPostChatFailed";
  OCSDKTelemetryEvent["GETCHATTRANSCRIPTSTARTED"] = "GetChatTranscriptStarted";
  OCSDKTelemetryEvent["GETCHATTRANSCRIPTSUCCEEDED"] = "GetChatTranscriptSucceeded";
  OCSDKTelemetryEvent["GETCHATTRANSCRIPTFAILED"] = "GetChatTranscriptFailed";
  OCSDKTelemetryEvent["EMAILTRANSCRIPTSTARTED"] = "EmailTranscriptStarted";
  OCSDKTelemetryEvent["EMAILTRANSCRIPTSUCCEEDED"] = "EmailTranscriptSucceeded";
  OCSDKTelemetryEvent["EMAILTRANSCRIPTFAILED"] = "EmailTranscriptFailed";
  OCSDKTelemetryEvent["FETCHDATAMASKINGSTARTED"] = "FetchDataMaskingStarted";
  OCSDKTelemetryEvent["FETCHDATAMASKINGSUCCEEDED"] = "FetchDataMaskingSucceeded";
  OCSDKTelemetryEvent["FETCHDATAMASKINGFAILED"] = "FetchDataMaskingFailed";
  OCSDKTelemetryEvent["SECONDARYCHANNELEVENTREQUESTSTARTED"] = "SecondaryChannelEventRequestStarted";
  OCSDKTelemetryEvent["SECONDARYCHANNELEVENTREQUESTSUCCEEDED"] = "SecondaryChannelEventRequestSucceeded";
  OCSDKTelemetryEvent["SECONDARYCHANNELEVENTREQUESTFAILED"] = "SecondaryChannelEventRequestFailed";
  OCSDKTelemetryEvent["GETSURVEYINVITELINKSTARTED"] = "GetSurveyInviteLinkStarted";
  OCSDKTelemetryEvent["GETSURVEYINVITELINKSUCCEEDED"] = "GetSurveyInviteLinkSucceeded";
  OCSDKTelemetryEvent["GETSURVEYINVITELINKFAILED"] = "GetSurveyInviteLinkFailed";
  OCSDKTelemetryEvent["VALIDATEAUTHCHATRECORDSTARTED"] = "ValidateAuthChatRecordStarted";
  OCSDKTelemetryEvent["VALIDATEAUTHCHATRECORDSUCCEEDED"] = "ValidateAuthChatRecordSucceeded";
  OCSDKTelemetryEvent["VALIDATEAUTHCHATRECORDFAILED"] = "ValidateAuthChatRecordFailed";
  OCSDKTelemetryEvent["GETAGENTAVAILABILITYSTARTED"] = "GetAgentAvailabilityStarted";
  OCSDKTelemetryEvent["GETAGENTAVAILABILITYSUCCEEDED"] = "GetAgentAvailabilitySucceeded";
  OCSDKTelemetryEvent["GETAGENTAVAILABILITYFAILED"] = "GetAgentAvailabilityFailed";
  OCSDKTelemetryEvent["SENDTYPINGINDICATORFAILED"] = "SendTypingIndicatorFailed";
  OCSDKTelemetryEvent["SENDTYPINGINDICATORSUCCEEDED"] = "SendTypingIndicatorSucceeded";
})(OCSDKTelemetryEvent || (exports.OCSDKTelemetryEvent = OCSDKTelemetryEvent = {}));

var LiveChatVersion;
exports.LiveChatVersion = LiveChatVersion;

(function (LiveChatVersion) {
  LiveChatVersion[LiveChatVersion["V1"] = 1] = "V1";
  LiveChatVersion[LiveChatVersion["V2"] = 2] = "V2";
  LiveChatVersion[LiveChatVersion["V3"] = 3] = "V3";
})(LiveChatVersion || (exports.LiveChatVersion = LiveChatVersion = {}));

var SDKError;
exports.SDKError = SDKError;

(function (SDKError) {
  SDKError["ClientHTTPTimeoutErrorName"] = "ClientHTTPTimeoutError";
  SDKError["ClientHTTPTimeoutErrorMessage"] = "Server took too long to respond";
})(SDKError || (exports.SDKError = SDKError = {}));
