// @microsoft/ocsdk/lib/Common/OCSDKLogger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _TelemetryHelper = _interopRequireDefault(require("../Utils/TelemetryHelper"));

var OCSDKLogger = /*#__PURE__*/function () {
  function OCSDKLogger(logger) {
    (0, _classCallCheck2["default"])(this, OCSDKLogger);
    (0, _defineProperty2["default"])(this, "logger", void 0);
    this.logger = logger;
  }

  (0, _createClass2["default"])(OCSDKLogger, [{
    key: "log",
    value: function log(loglevel, telemetryEvent, customData, description) {
      // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
      if (this.isLoggingEnabled()) {
        var logData = _TelemetryHelper["default"].GETTELEMETRYEVENTDATA(telemetryEvent, customData, description);

        this.logEvent(loglevel, logData);
      }
    }
  }, {
    key: "logEvent",
    value: function logEvent(logLevel, logData) {
      if (this.isLoggingEnabled()) {
        setTimeout(this.logger.logClientSdkTelemetryEvent.bind(this.logger), 0, logLevel, logData);
      }
    }
  }, {
    key: "isLoggingEnabled",
    value: function isLoggingEnabled() {
      return !this.isNullOrUndefined(this.logger);
    }
  }, {
    key: "isNullOrUndefined",
    value: function isNullOrUndefined(obj) {
      return obj === null || obj === undefined;
    }
  }]);
  return OCSDKLogger;
}();

exports["default"] = OCSDKLogger;
