// @microsoft/ocsdk/lib/Utils/TelemetryHelper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
var TelemetryHelper = /*#__PURE__*/function () {
  function TelemetryHelper() {
    (0, _classCallCheck2["default"])(this, TelemetryHelper);
  }

  (0, _createClass2["default"])(TelemetryHelper, null, [{
    key: "GETTELEMETRYEVENTDATA",
    value: function GETTELEMETRYEVENTDATA(telemetryEvent, customData, description) {
      var logData = {
        Description: description,
        ElapsedTimeInMilliseconds: customData.ElapsedTimeInMilliseconds,
        Event: telemetryEvent,
        ExceptionDetails: customData.ExceptionDetails,
        Region: customData.Region,
        RequestId: customData.RequestId,
        TransactionId: customData.TransactionId,
        RequestPayload: customData.RequestPayload,
        RequestPath: customData.RequestPath,
        RequestMethod: customData.RequestMethod,
        ResponseStatusCode: customData.ResponseStatusCode
      };
      return logData;
    }
  }]);
  return TelemetryHelper;
}();

exports["default"] = TelemetryHelper;
