// botframework-webchat-core/lib/actions/setNotification.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SET_NOTIFICATION = void 0;
exports["default"] = setNotification;

var _uniqueID = _interopRequireDefault(require("../utils/uniqueID"));

var SET_NOTIFICATION = 'WEB_CHAT/SET_NOTIFICATION';
exports.SET_NOTIFICATION = SET_NOTIFICATION;

function setNotification(_ref) {
  var alt = _ref.alt,
      data = _ref.data,
      id = _ref.id,
      level = _ref.level,
      message = _ref.message;

  if (!id || typeof id !== 'string') {
    console.warn('botframework-webchat: The "id" argument passed to "setNotification" must be a string; defaulting to a random value.');
    id = (0, _uniqueID["default"])();
  }

  if (!level || typeof level !== 'string') {
    console.warn('botframework-webchat: The "level" argument passed to "setNotification" must be a string; defaulting to "info".');
    level = 'info';
  }

  return {
    type: SET_NOTIFICATION,
    payload: {
      alt: alt,
      data: data,
      id: id,
      level: level,
      message: message
    }
  };
}
