// @microsoft/omnichannel-chat-components/lib/cjs/components/callingcontainer/CallingContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _react2 = require("@fluentui/react");

var _CurrentCall = _interopRequireDefault(require("./subcomponents/CurrentCall/CurrentCall"));

var _IncomingCall = _interopRequireDefault(require("./subcomponents/IncomingCall/IncomingCall"));

var _defaultCallingContainerProps = require("./common/defaultProps/defaultCallingContainerProps");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function CallingContainer(props) {
  var _props$controlProps, _defaultCallingContai, _defaultCallingContai2, _props$styleProps, _props$controlProps2, _defaultCallingContai3, _props$styleProps2, _props$controlProps3, _defaultCallingContai4, _props$controlProps4, _props$styleProps3, _props$controlProps5, _props$styleProps4;

  const callingContainerId = ((_props$controlProps = props.controlProps) === null || _props$controlProps === void 0 ? void 0 : _props$controlProps.id) ?? ((_defaultCallingContai = _defaultCallingContainerProps.defaultCallingContainerProps.controlProps) === null || _defaultCallingContai === void 0 ? void 0 : _defaultCallingContai.id);
  let stackStyles = {
    root: Object.assign({}, (_defaultCallingContai2 = _defaultCallingContainerProps.defaultCallingContainerProps.styleProps) === null || _defaultCallingContai2 === void 0 ? void 0 : _defaultCallingContai2.generalStyleProps, (_props$styleProps = props.styleProps) === null || _props$styleProps === void 0 ? void 0 : _props$styleProps.generalStyleProps)
  };
  const hideCallingContainer = ((_props$controlProps2 = props.controlProps) === null || _props$controlProps2 === void 0 ? void 0 : _props$controlProps2.hideCallingContainer) ?? ((_defaultCallingContai3 = _defaultCallingContainerProps.defaultCallingContainerProps.controlProps) === null || _defaultCallingContai3 === void 0 ? void 0 : _defaultCallingContai3.hideCallingContainer);

  if (hideCallingContainer) {
    stackStyles = {
      root: Object.assign({}, stackStyles.root, {
        display: "none"
      })
    };
  }

  return /*#__PURE__*/React.createElement(_react2.Stack, {
    id: callingContainerId,
    className: (_props$styleProps2 = props.styleProps) === null || _props$styleProps2 === void 0 ? void 0 : _props$styleProps2.className,
    horizontalAlign: "space-between",
    dir: ((_props$controlProps3 = props.controlProps) === null || _props$controlProps3 === void 0 ? void 0 : _props$controlProps3.dir) ?? ((_defaultCallingContai4 = _defaultCallingContainerProps.defaultCallingContainerProps.controlProps) === null || _defaultCallingContai4 === void 0 ? void 0 : _defaultCallingContai4.dir),
    styles: stackStyles
  }, ((_props$controlProps4 = props.controlProps) === null || _props$controlProps4 === void 0 ? void 0 : _props$controlProps4.isIncomingCall) && /*#__PURE__*/React.createElement(_IncomingCall.default, {
    controlProps: props.controlProps.incomingCallControlProps,
    styleProps: (_props$styleProps3 = props.styleProps) === null || _props$styleProps3 === void 0 ? void 0 : _props$styleProps3.incomingCallStyleProps
  }), ((_props$controlProps5 = props.controlProps) === null || _props$controlProps5 === void 0 ? void 0 : _props$controlProps5.isIncomingCall) === false && /*#__PURE__*/React.createElement(_CurrentCall.default, {
    controlProps: props.controlProps.currentCallControlProps,
    styleProps: (_props$styleProps4 = props.styleProps) === null || _props$styleProps4 === void 0 ? void 0 : _props$styleProps4.currentCallStyleProps
  }));
}

var _default = CallingContainer;
exports.default = _default;
